---
title: "PTB_FinalMainAnalysis"
author: "Triet Thai"
date: "2024-03-29"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# PTB Transcriptomics Project

## 1. Import libraries and data

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	message = FALSE,
	warning = FALSE)

#library(parallel)
#mcore<-1 # for Window PC

#Load libraries
library(dplyr)
library(plyr)
library(readxl)
library(writexl)
library(survival)
library(glmnet)
library(stabs)
library(tidyverse)
library(magrittr)
library(Hmisc)
library(gtsummary)
library(WGCNA)
library(ggcorrplot)
library(scales)
library(gridExtra)
library(ggcorrplot)
library(broom)
library(rms)
library(mboost)
library(DESeq2)
library(CEMiTool)
library(igraph)
library(flextable)
library(pheatmap)
library(ggplot2)
library(ggplotify)
library(stringr)
library(ComplexHeatmap)
library(circlize)
library(rstatix)
library(ggpubr)
#library(colorRamps)
library(Metrics)
library(ggpmisc)
library(ggpubr)
library(cowplot)
library(ggplot2)
library(Metrics)
path_input <- paste0(str_replace_all(Sys.getenv("USERPROFILE"), '\\\\', '/'),
                     "/OneDrive - Oxford University Clinical Research Unit/#OUCRU-TB/#oucru-tb/PTB_project/")
path_output<-paste0(str_replace_all(Sys.getenv("USERPROFILE"), '\\\\', '/'),
                     "/OneDrive - Oxford University Clinical Research Unit/#OUCRU-TB/#oucru-tb/PTB_project/FinalResults")
```

```{r}
#move to bayesprism notebook later!!

# bp.res_ptb <- get(load(file=paste0(path_input,"Data/deconvolution_results/bp.res_ptb_newman_01_01.rdata"))) #bayesprism
# bp.res_healthy <- get(load(file=paste0(path_input,"Data/deconvolution_results/bp.res_healthy_newman_01_01.rdata"))) #instaprism
# bp.res_tbm <- get(load(file=paste0(path_input,"Data/deconvolution_results/bp.res_tbm_newman_01_01.rdata"))) #instaprism
# 
# ########### PTB ##################
# ptb.theta <- as.data.frame(get.fraction(bp=bp.res_ptb, which.theta="final",state.or.type="type"))
# #ptb.theta.cv<- bp.res_ptb@posterior.theta_f@theta.cv
# ptb.z <- get.exp(bp=bp.res_ptb,state.or.type="type")
# save(ptb.theta,ptb.z,file = paste0(path_input,"Data/deconvolution_results/bp.res_ptb_SCnewman_01_01.RData"))
# 
# healthy.theta <- as.data.frame(t(bp.res_healthy@theta))
# healthy.z <- get_Z_array(bp.res_healthy, resolution = 'ct', n.core = 7)
# save(healthy.theta, healthy.z, file = paste0(path_input,"Data/deconvolution_results/bp.res_healthy_SCnewman_01_01.RData"))
# 
# tbm.theta <- as.data.frame(t(bp.res_tbm@theta))
# tbm.z <- get_Z_array(bp.res_tbm, resolution = 'ct', n.core = 7)
# save(tbm.theta, tbm.z, file = paste0(path_input,"Data/deconvolution_results/bp.res_tbm_SCnewman_01_01.RData"))
```

```{r}
#clinical data
load(paste0(path_input,'/Data/clinical_data_HC_CNS_IF_PTB_TBM.RData'))
load(paste0(path_input,"/Raw_data/vst_Gensymbol_allPTB_295Sample_18Feb22.RData")) #vst_data 295 31425
load(paste0(path_input,"/Raw_data/vst_Gensymbol_allTBM_281Sample_12Jan24.RData")) #tbm_vst 281 31381
rownames(tbm_vst) <- tbm_vst$LIMS_ID
load(paste0(path_input,"/Data/Fluidigm_PTB_data.Rdata"))
load(paste0(path_input,"/Data/Most_variance_RNA_genes.Rdata"))
load(file=paste0(path_input,"Raw_data/count_data_adjusted_ComBat.Rdata"))
load(paste0(path_input, "Raw_data/Data_RNAseq_RUN1andRUN2_15Feb2023.RData"))
All_sample_RNAseq <- readxl::read_excel(paste0(path_input,"Raw_data/All_sample_RNAseq.xlsx"),
                                        sheet = "Run1and2") %>% as.data.frame()
rownames(All_sample_RNAseq) <- All_sample_RNAseq$LIMS_ID

load(file=paste0(path_input,"Raw_data/Data_26_27_CT_CNIH4.Rdata.RData"))
Fludigm_CNIH4_PTB <- get(load(paste0(path_input,'Raw_data/Fludigm_CNIH4_PTB.Rdata')))

ModuleHubgeneCT <- read.csv(paste0(path_input,"Data/Modules-Hubgenes_CT_value.csv"))
load(file=paste0(path_input,"Data/Gene Significance and Module Membership_clinical_trait.Rdata"))
load(file = paste0(path_input,"Data/PTB-01-dataInput.RData"))
load(file = paste0(path_input,"Data/PTB-02-networkConstruction-auto.RData"))
load(file = paste0(path_input,"Data/Module_discovery.Rdata"))
load(file=paste0(path_input,"Raw_data/PTB_TOM_10000-block.1.RData"))
TOM1 <- as.matrix(TOM)
colnames(TOM1) <- colnames(multiExpr[[1]]$data)
rownames(TOM1) <- colnames(multiExpr[[1]]$data)

clinical_2_discovery <- filter(clinical_2, cohort=='Discovery')
rownames(clinical_2_discovery) <- clinical_2_discovery$LIMS_ID

###------ImputeCIBERSORT------###
impute_df <-  read.csv(paste0(path_input,'/Raw_data/CIBERSORTx_Adjusted_CPM_LM22.txt'),sep='\t') %>% filter(P.value<0.05)
rownames(impute_df) <- impute_df$Mixture
impute_df$Mixture <- NULL

###------BayesPrism------###
load(paste0(path_output,"/deconvolution/bp.res_ptb2_PBMC_01_01.RData"))
load(paste0(path_output,"/deconvolution/bp.res_healthy2_PBMC_01_01.RData"))
load(paste0(path_output,"/deconvolution/bp.res_tbm2_PBMC_01_01.RData"))
gt <- data.frame(LIMS_ID = clinical_data_HC_CNS_IF_PTB_TBM$LIMS_ID,
           Neutrophil = clinical_data_HC_CNS_IF_PTB_TBM$NEUTLE,
           Lymphocyte = clinical_data_HC_CNS_IF_PTB_TBM$LYMLE,
           Monocyte =  clinical_data_HC_CNS_IF_PTB_TBM$MONOLE)

gt <- na.omit(gt)
rownames(gt) <- gt$LIMS_ID
gt$LIMS_ID <- NULL
```

```{r}
load(paste0(path_output,"/deconvolution/bp.res_healthy_PBMC_01_01.RData"))
healthy.theta <- ptb.theta
healthy.z <- ptb.z
save(healthy.theta,healthy.z,file=paste0(path_output,"/deconvolution/bp.res_healthy_PBMC_01_01.RData"))

load(paste0(path_output,"/deconvolution/bp.res_tbm_PBMC_01_01.RData"))
tbm.theta <- ptb.theta
tbm.z <- ptb.z
save(tbm.theta,tbm.z,file=paste0(path_output,"/deconvolution/bp.res_tbm_PBMC_01_01.RData"))
```

Split samples by low and high CNIH4 (refactor later)

```{r}
vst_ptb_tbm <- rbind.fill(vst_data,tbm_vst)
vst_ptb_tbm %<>% mutate(Type=ifelse(LIMS_ID %in% vst_data$LIMS_ID,"PTB","TBM"))

healthy_ComBat <- count_data_adjusted_ComBat %>% filter(LIMS_ID %in% filter(Cli_All,trial_number %in% c('HV','43TB'))$LIMS_ID)
qt <- quantile(healthy_ComBat$CNIH4,c(0.75)) #Q3 CNIH4 of healthy control

PTB_TBM_ComBat <- count_data_adjusted_ComBat %>% filter(LIMS_ID %in% filter(Cli_All,(Timepoint=='D0') & (trial_number %in% c('MDR_PTB','DS_PTB','26TB','27TB')))$LIMS_ID)
PTB_TBM_ComBat %<>% mutate(CNIH4level=ifelse((CNIH4<qt[1]),1,2))

vst_ptb_tbm <- merge(vst_ptb_tbm,PTB_TBM_ComBat %>% select('LIMS_ID','CNIH4level'),by="LIMS_ID")
rownames(vst_ptb_tbm) <- vst_ptb_tbm$LIMS_ID
```

```{r}
healthy_df <- count_data_adjusted_ComBat %>% filter(LIMS_ID %in% filter(Cli_All,trial_number=='HV'|trial_number=='43TB')$LIMS_ID)
qt<-quantile(healthy_df$CNIH4,c(0.75)) #Q3 CNIH4 of healthy control

PTB_df <- count_data_adjusted_ComBat %>% filter(LIMS_ID%in%filter(Cli_All,trial_number=='MDR_PTB'|trial_number=='DS_PTB')$LIMS_ID)

# tmp <- PTB_df %>% select(CNIH4) %>% mutate(level=ifelse((CNIH4<qt[1]),"LOW",'HIGH'))

tmp <- PTB_df %>% select(CNIH4) %>% mutate(level=ifelse((CNIH4<qt[1]),1,2))

rownames(tmp) <- PTB_df$LIMS_ID
table(tmp$level) # 244 52
vst_data2<-merge(vst_data,tmp[,2, drop=FALSE],by=0)
rownames(vst_data2) <- vst_data2$LIMS_ID
dim(vst_data2) # 295 31427
```

## 2. Helper functions

```{r}
plot_regs <- function(theta1,gt1,show_cutoff = TRUE,name='data') {
    cell_list = c('Lymphocyte','Monocyte',"Neutrophil",'Myeloid')
    p <- list()
    
    for (i in 1:length(cell_list)) {
      c <- cell_list[i]
      df <- data.frame(predicted=theta1[,c],groundtruth=gt1[,c]/100)
      
      rmse <- mse(gt1[,c]/100, theta1[,c])
    
      p1<-ggplot(data=df, aes(x=groundtruth,y=predicted))+
          geom_point(alpha=0.8,col='salmon')+
          stat_poly_line(se=T,col='#4164a9') +
          stat_cor(cor.coef.name = "r",size = 7,method = 'pearson')+
          ylab(paste0(c," estimated"))+
          xlab(paste0(c," true"))+
          theme_bw()+
          ggtitle(paste0("MSE = ",round(rmse,4)))+
          theme(plot.title = element_text(size = 15),
                legend.position = "none",
                axis.text = element_text(size = 15),
                axis.text.x=element_text(color="black",size=13,face="bold"),
                axis.title.x = element_text(size = 15,color="black",face="bold"),
                axis.text.y=element_text(color="black",size=15,face="bold"),
                axis.title.y = element_text(size = 15,color="black",face="bold"),
                legend.text = element_text(size = 15),
                legend.title = element_text( colour="black", size=12),
                legend.key = element_rect(colour = NA),
                strip.text=element_text(face = "bold",size=12))
      p[[i]] <-ggdraw(p1)
    }
    
    regs <- ggpubr::ggarrange(
    p[[1]],
    p[[2]],
    p[[3]],
    p[[4]],
    
    ncol = 4, nrow = 1)
  if(show_cutoff) {
      regs <- annotate_figure(regs, top = text_grob(paste0(name, ", pval.max=", pval, ", lfc.min=",lfc), 
                 color = "darkblue", face = "bold", size = 14))
  }
  else {
      regs <- annotate_figure(regs, top = text_grob(paste0(name), 
                 color = "darkblue", face = "bold", size = 14))
  }
  return(regs)
}

posterior_predict_hurdle_gaussian <- function(i, prep, ...) {
  mu <- brms::get_dpar(prep, "mu", i = i)
  sigma <- brms::get_dpar(prep, "sigma", i = i)
  theta <- brms::get_dpar(prep, "hu", i = i)
     
  hu <- runif(prep$ndraws, 0, 1)
  ifelse(hu < theta, 0, rnorm(prep$ndraws, mu,sigma))
}
   
posterior_epred_hurdle_gaussian <- function(prep) {
  with(prep$dpars, mu * (1 - hu))
}

cor.pval.test <- function(x1, x2, ...) {
  r <- ncol(x1)
  n <- ncol(x2)
  p.mat<- matrix(NA, r, n)
  for (i in 1:r) {
    for (j in 1:n) {
      tmp <- cor.test(x1[, i], x2[, j], ...)
      p.mat[i, j] <- tmp$p.value
    }
  }
  rownames(p.mat) <-  colnames(x1)
  colnames(p.mat) <- colnames(x2)
  return(p.mat)
}

anno_gene_symbol <- function(data,data_anno,export, name=NULL){
  anno_DESeq2 <- data_anno[,c("gene_id", "gene_symbol")]
  anno_DESeq2 <- subset(anno_DESeq2, gene_id %in% data$gene_id)
  data
  return(data)
}

get_genelist_pathway_kegg <- function(hsa_ID){
  library("KEGGREST")
  names <- keggGet(hsa_ID)[[1]]$GENE
  namesodd <-  names[seq(0,length(names),2)]
  namestrue <- gsub("\\;.*","",namesodd)
}

draw_colnames_45 <- function (coln, gaps, ...) {
    coord <- pheatmap:::find_coordinates(length(coln), gaps)
    x     <-  coord$coord - 0.5 * coord$size
    res   <- grid::textGrob(
      coln, x = x, y = unit(1, "npc") - unit(3,"bigpts"),
      vjust = 0.75, hjust = 1, rot = 45, gp = grid::gpar(...)
    )
    return(res)
}

assignInNamespace(
  x = "draw_colnames",
  value = "draw_colnames_45",
  ns = asNamespace("pheatmap")
)

# Modify ordering of the clusters using clustering callback option
callback <- function(hc, mat){
    sv = svd(t(mat))$v[,1]
    dend = reorder(as.dendrogram(hc), wts = sv)
    as.hclust(dend)
}

make_face_names <- function(mat, rc_fun, rc_names_b = NA, 
                            rc_names_i = NA) {
  f_names <- rc_fun(mat)
  ids_b <- rc_names_b %>% match(rc_fun(mat))
  ids_i <- rc_names_i %>% match(rc_fun(mat))
  
  ids_b %>%
    walk(
      function(i)
        f_names[i] <<-
        bquote(bold(.(rc_fun(mat)[i]))) %>%
        as.expression()
    )
  ids_i %>%
    walk(
      function(i)
        f_names[i] <<-
        bquote(italic(.(rc_fun(mat)[i]))) %>%
        as.expression()
    )
  
  f_names
}

color.bar <- function(lut, min, max=-min, nticks=11, ticks=seq(min, max, len=nticks), title='') {
  scale = (length(lut)-1)/(max-min)
  plot(c(0,10), c(min,max), type='n', bty='n', xaxt='n', xlab='', yaxt='n', ylab='', main=title)
  axis(2, ticks, las=1)
  for (i in 1:(length(lut)-1)) {
    print(lut[i])
    y = (i-1)/scale + min
    rect(0,y,10,y+1/scale, col=lut[i], border=NA)
  }
}

get_theta <- function(bp) {
  theta <- get.fraction (bp=bp,
            which.theta="final",
            state.or.type="type")
  theta <- as.data.frame(theta)
  row.names(theta) <- gsub("X", "", row.names(theta))
  row.names(theta) <- gsub("_PBMC", "", row.names(theta))
  theta <- theta[ order(row.names(theta)), ]
  theta$Lymphocyte <- theta$B + theta$CD4T + theta$CD8T + theta$NK
  return(theta)
}
```

## 3. Main Analysis

### 3.1. Correlation Heatmap CNIH4 vs 27 ICs

```{r}
IC_names<-c('CD200', 'CD2','CD58','TNFSF4', 'CD200R1', 'IDO1','TNFRSF18', 'TNFSF18',
             'CD28','BTLA', 'CD27', 'HAVCR2', 'CD274', 'TNFSF14', 'TNFRSF4',
             'CD276', 'TNFRSF14', 'PDCD1', 'CD86', 'TNFSF9', 'TIGIT','CD80',
             'ICOS', 'TNFRSF9', 'CTLA4', 'LAG3','LGALS9',
            'TMED9','TGFA','GLI1','CNIH4')

ICvsCNIH4_df <- vst_data %>% select(IC_names)
# ICvsCNIH4_df <- ICvsCNIH4_df[vapply(ICvsCNIH4_df, 
#                                     function(x) length(unique(x)) > 1, 
#                                     logical(1L))] %>% dplyr::rename('HAVCR2/TIM-3'='HAVCR2',
#                                                                     'CD274/PDL1'='CD274',
#                                                                     'LGALS9/Galactin-9'='LGALS9',
#                                                                     'PDCD1/PD1'='PDCD1')
cor_method = 'spearman'
ICvsCNIH4_mat <- cor(ICvsCNIH4_df,
                      use='pairwise.complete.obs',
                      method = cor_method)

p.ICvsCNIH4_mat <- cor.pval.test(ICvsCNIH4_df, 
                                  ICvsCNIH4_df, 
                                  method=cor_method, 
                                  use='pairwise.complete.obs')

#col=c("#001164", 'white', "#FF681E")
#col=c("#5E81AC", '#EBCB8B', "#BF616A")
col=c("#395b85", 'white', "#FF681E")
#col=c("#243b51","#4e80b1",'#b5c8e2', 'white', '#feab63',"#fe9250","#fe7a29")
# CNIH4 vs 25 ICs
mat1 <- ICvsCNIH4_mat
p.mat1 <- p.ICvsCNIH4_mat
labels1 <-round(mat1,2)
vals1 <- labels1
vals1[p.mat1 > 0.05] <- 0
labels1[p.mat1 > 0.05] <- ''

myBreaks <- c(seq(-1, 0, length.out=ceiling(1000/2) + 1),
              seq(1/1000, 1, length.out=floor(1000/2)))

p01 <- pheatmap(vals1,
               color = colorRampPalette(col)(1000),
               breaks=myBreaks,
               display_numbers=labels1,
               show_colnames = T,legend = T)
p01
row.order <- p01$tree_row$order
col.order <- p01$tree_col$order
p1 <- as.ggplot(p01)

library(NatParksPalettes)
names(NatParksPalettes)
```

Complex Heatmap

```{r}

genes<-c('STAT1','CNIH4', 'CD274','AIM2','ATF3','CT_mean')

genes<-c('CD200', 'CD2','CD58','TNFSF4', 'CD200R1', 'IDO1','TNFRSF18', 'TNFSF18',
             'CD28','BTLA', 'CD27', 'HAVCR2', 'CD274', 'TNFSF14', 'TNFRSF4',
             'CD276', 'TNFRSF14', 'PDCD1', 'CD86', 'TNFSF9', 'TIGIT','CD80',
             'ICOS', 'TNFRSF9', 'CTLA4', 'LAG3','LGALS9',
            'TMED9','TGFA','GLI1','CNIH4')

df <- vst_data %>% select(genes)
cor_method = 'spearman'
mat <- cor(df, use='pairwise.complete.obs',method = cor_method)
p.mat <- cor.pval.test(df, df, method=cor_method, use='pairwise.complete.obs')
mat[p.mat>0.05] = 0

col=c("#395b85", 'white', "#FF681E")
col_fun = colorRamp2(seq(-1, 1, length = 3), col)
Heatmap(mat,
        name = 'corr.',
        col = col_fun,
        heatmap_legend_param = list(legend_height = unit(4, "cm")),
        rect_gp = gpar(col = "black", lwd = 1),
        cell_fun = function(j, i, x, y, width, height, fill) {
          if(i==j) grid.text(sprintf("%.f", mat[i, j]), x, y, gp = gpar(fontsize = 10))
          if(i!=j & p.mat[i, j]<=0.05) grid.text(sprintf("%.2f", mat[i, j]), x, y, gp = gpar(fontsize = 10))},
        column_names_rot = 45)
```

```{r}
library(ggthemes)
library(scales)

palettes <- ggthemes_data[["tableau"]][["color-palettes"]][["regular"]]
for (palname in names(palettes)) {
  pal <- tableau_color_pal(palname)
  max_n <- attr(pal, "max_n")
  show_col(pal(max_n))
  title(main = palname)
}
```

### 3.2.0. Custom

```{r}

df <- vst_data %>%select('CNIH4','TGFA')

p1<-ggplot(data=df, aes(x=CNIH4,y=TGFA))+
  geom_point(alpha=0.2,col='red')+
  theme_bw()+
  stat_poly_line(se=T) +
  stat_cor(cor.coef.name = "r",size = 7,method = 'spearman')+
  theme_bw()+ 
  ylab(paste0(as.name(colnames(df)[i])," expression"))+
  xlab("CNIH4 expression")+
  theme(plot.title = element_text(size = 15),
        legend.position = "none",
        axis.text = element_text(size = 15),
        axis.text.x=element_text(color="black",size=13,face="bold"),
        axis.title.x = element_text(size = 15,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=15),
        axis.title.y = element_text(size = 15,color="black",face="bold"),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=12),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=12))
p1
```

### 3.2. Scatter plot CNIH4 vs 27 ICs

```{r}
library(ggpmisc)
library(ggpubr)
library(cowplot)

IC_names<-c('CD200', 'CD2','CD58','TNFSF4', 'CD200R1', 'IDO1','TNFRSF18', 'TNFSF18',
             'CD28','BTLA', 'CD27', 'HAVCR2', 'CD274', 'TNFSF14', 'TNFRSF4',
             'CD276', 'TNFRSF14', 'PDCD1', 'CD86', 'TNFSF9', 'TIGIT','CD80',
             'ICOS', 'TNFRSF9', 'CTLA4', 'LAG3','LGALS9','CNIH4')

df <- vst_data %>%select(IC_names) %>% dplyr::rename('HAVCR2/TIM-3'='HAVCR2',
         'CD274/PDL1'='CD274',
         'LGALS9/Galactin-9'='LGALS9',
         'PDCD1/PD1'='PDCD1'
         ) #295 patients

p<-list()

for(i in 1:27){
p1<-ggplot(data=df, aes_string(x='CNIH4',y=as.name(colnames(df)[i])))+
  geom_point(alpha=0.2,col='red')+
  theme_bw()+
  stat_poly_line(se=T) +
  stat_cor(cor.coef.name = "r",size = 7,method = 'spearman')+
  theme_bw()+ 
  ylab(paste0(as.name(colnames(df)[i])," expression"))+
  xlab("CNIH4 expression")+
  theme(plot.title = element_text(size = 15),
        legend.position = "none",
        axis.text = element_text(size = 15),
        axis.text.x=element_text(color="black",size=13,face="bold"),
        axis.title.x = element_text(size = 15,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=15),
        axis.title.y = element_text(size = 15,color="black",face="bold"),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=12),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=12))
  p[[i]]<-ggdraw(p1)
  ggdraw(p1)
}

#png(file=paste0(path_output,"/Figure4B_scatterplot_CNIH4vsICs_all_PTB.png"),res=600, units = 'in', width=45, height =30)#30:18
ggpubr::ggarrange(
p[[1]],p[[2]],
p[[3]],p[[4]],
p[[5]],p[[6]],
p[[7]],p[[8]],
p[[9]],p[[10]],
p[[11]],p[[12]],
p[[13]],p[[14]],
p[[15]],p[[16]],
p[[17]],p[[18]],
p[[19]],p[[20]],
p[[21]],p[[22]],
p[[23]],p[[24]],
p[[25]],p[[26]],
p[[27]],
ncol =6,nrow=5)
#dev.off()
```

### 3.3. Hurdles Model PTB vs TBM vs PCR

```{r}
library(brms)
library(cmdstanr)
library(posterior)
library(bayesplot)
library(tidybayes)
library(broom)
library(dplyr)
library(ggplot2)
library(broom.mixed)

options(mc.cores = 1, brms.backend = "cmdstanr")

CHAINS <- 4
ITER <- 2000
WARMUP <- 1000
BAYES_SEED <- 1234

# Create a custom family that is logit if y = 0, normal/gaussian if not
hurdle_gaussian <- custom_family("hurdle_gaussian", 
                dpars = c("mu", "sigma", "hu"),
                links = c("identity", "log", "logit"),
                lb = c(NA, 0, NA),
                type = "real")

# Stan code
stan_funs <- "
  real hurdle_gaussian_lpdf(real y, real mu, real sigma, real hu) { 
    if (y == 0) { 
      return bernoulli_lpmf(1 | hu); 
    } else { 
      return bernoulli_lpmf(0 | hu) +  
             normal_lpdf(y | mu, sigma); 
    } 
  }
"
# Prepare Stan code for use in brm()
stanvars <- stanvar(scode = stan_funs, block = "functions")
#set_cmdstan_path('C:/Users/triettm/AppData/Local/R/win-library/4.3/.cmdstan/cmdstan-2.33.1')
set_cmdstan_path('C:/Users/triettm/OneDrive - Oxford University Clinical Research Unit/Documents/.cmdstan/cmdstan-2.34.0')
```

#### 3.3.1. CNIH4 vs Ct value in PTB patients

#### 3.3.2. CNIH4 vs Ct value in TBM patients

#### 3.3.3. CNIH4 vs Ct value in PCR patients

### 3.4. Deconvolution Analysis using CIBERSORT

#### 3.4.1. CIBERSORT Cell pops

```{r}
library(ggpubr)
library(rstatix)

healthy_df <- count_data_adjusted_ComBat %>% filter(LIMS_ID %in% filter(Cli_All,trial_number=='HV'|trial_number=='43TB')$LIMS_ID)
rownames(healthy_df) <- healthy_df$LIMS_ID

PTB_df <- count_data_adjusted_ComBat %>% filter(LIMS_ID%in%filter(Cli_All,trial_number=='MDR_PTB'|trial_number=='DS_PTB')$LIMS_ID)
rownames(PTB_df) <- PTB_df$LIMS_ID

TBM_df <- count_data_adjusted_ComBat %>% filter(LIMS_ID%in%filter(Cli_All,(trial_number=='26TB'|trial_number=='27TB')&(Timepoint=='D0'))$LIMS_ID)
rownames(TBM_df) <- TBM_df$LIMS_ID

tmp1 <- merge(impute_df,healthy_df,by=0) %>% select(B.cells.naive:Neutrophils)
dim(tmp1)
tmp1 <- pivot_longer(tmp1,B.cells.naive:Neutrophils)
tmp1$group <- rep('Healthy',each=dim(tmp1)[1])

tmp2 <- merge(impute_df,PTB_df,by=0) %>% select(B.cells.naive:Neutrophils)
dim(tmp2)
tmp2 <- pivot_longer(tmp2,B.cells.naive:Neutrophils)
tmp2$group <- rep('PTB',each=dim(tmp2)[1])

tmp3 <- merge(impute_df,TBM_df,by=0) %>% select(B.cells.naive:Neutrophils)
dim(tmp3)
tmp3 <- pivot_longer(tmp3,B.cells.naive:Neutrophils)
tmp3$group <- rep('TBM',each=dim(tmp3)[1])



tmp<-rbind(tmp1,tmp2,tmp3)

stat.test <- tmp %>%
  group_by(name) %>% 
  t_test(value ~ group) %>% 
  add_significance("p") %>%
  add_xy_position(x = "name")

p <- ggplot(data=tmp,
       aes(x=name,y=value, color=factor(group, level = c('Healthy','PTB','TBM'))))+
  geom_boxplot() +
  geom_point(position=position_jitterdodge(jitter.width =0.3),
             alpha=0.1)+
  scale_color_manual(values=c("#81A1C1","#B48EAD","#A3BE8C"))+
  labs(color='Group')+ 
  xlab('Cell type')+
  ylab('% Cell population')+
  theme_bw()+
  theme(plot.title = element_text(size = 25),
        legend.position = "right",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_text(color="black",size=13,face="bold",angle = 45,vjust = 1, hjust=1),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold")) +
  stat_pvalue_manual(stat.test, label = 'p.signif', tip.length = 0)

# png(file=paste0(path_output,"/Figure_CellPop_HealthyvsPTBvsTBM.png"),res=900, units = 'in', width=17, height = 9)
# p
# dev.off()
```

#### 3.4.2. CIBERSORT CNIH4

```{r}
impute_df
```

```{r}
df <- filter(vst_ptb_tbm,Type=='PTB') %>% select(c('LIMS_ID','CNIH4level'))

tmp <- merge(impute_df,df,by=0) %>% select(c(B.cells.naive:Neutrophils,CNIH4level))
tmp <- pivot_longer(tmp,B.cells.naive:Neutrophils)

# p<-ggplot(data=tmp,
#        aes(x=name,y=value))+
#   geom_boxplot(aes(color=name)) +
#   geom_jitter(height=0,aes(color=name),alpha=0.1)+
#   xlab('Cell type')+
#   ylab('% Cell population')+
#   theme_bw()+
#   theme(plot.title = element_text(size = 25),
#         legend.position = "",
#         legend.text = element_text(size = 20) ,
#         legend.title  = element_text(size = 20,color="black",face="bold"),
#         axis.text = element_text(size = 25),
#         axis.text.x=element_text(color="black",size=13,face="bold",angle = 45,vjust = 1, hjust=1),
#         axis.title.x = element_text(size = 20,color="black",face="bold"),
#         axis.text.y=element_text(color="black",size=20),
#         axis.title.y = element_text(size = 20,color="black",face="bold"),
#         axis.title= element_text(size = 40,color="black",face="bold"))
#   
# png(file=paste0(path_output,"/Figure_CellPop_All.png"),res=900, units = 'in', width=14, height = 9)
# p
# dev.off()

stat.test <- tmp %>%
  group_by(name) %>% 
  t_test(value ~ CNIH4level) %>% 
  add_significance("p") %>%
  add_xy_position(x = "name")

p <- ggplot(data=tmp,
       aes(x=name,y=value, color=factor(CNIH4level, level = c('LOW','HIGH'))))+
  geom_boxplot() +
  geom_point(position=position_jitterdodge(jitter.width =0.3),
             alpha=0.1)+
  scale_color_manual(values=c("#81A1C1","#B48EAD"))+
  labs(color='CNIH4 Level')+ 
  xlab('Cell type')+
  ylab('% Cell population')+
  theme_bw()+
  theme(plot.title = element_text(size = 25),
        legend.position = "right",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_text(color="black",size=13,face="bold",angle = 45,vjust = 1, hjust=1),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold")) +
  stat_pvalue_manual(stat.test, label = 'p.signif', tip.length = 0)
p
```

### 3.5. Devolution Analysis (BayesPrism)

#### Prepared data for visualization

```{r}
# Get cell proportion for each patient
celltypes <- colnames(ptb.theta)
id26TB <- All_sample_RNAseq %>% subset(Study %in% c("26TB") & timepoint=='D0') %>% rownames()
id27TB <- All_sample_RNAseq %>% subset(Study %in% c("27TB") & timepoint=='D0') %>% rownames()

tbm26.theta <- tbm.theta[id26TB,]
tbm27.theta <- tbm.theta[id27TB,]
tbm26.z <- tbm.z[id26TB,,]
tbm27.z <- tbm.z[id27TB,,]

l = list(healthy.theta,ptb.theta,tbm26.theta,tbm27.theta)

for( i in seq_along(l)){
  l[[i]]$Neutrophil <- l[[i]]$Neutro
  l[[i]]$Lymphocyte <- l[[i]]$`B naive` + l[[i]]$`B memory` + l[[i]]$`CD8 Naive`+l[[i]]$`CD8 TM` + l[[i]]$MAIT + l[[i]]$NK + l[[i]]$`CD4 TM`+l[[i]]$`CD4 Naive`+l[[i]]$`CD4 CTL`+l[[i]]$OtherT +l[[i]]$Treg + l[[i]]$Plasmablast
  l[[i]]$Monocyte <- l[[i]]$`CD14 Mono`+l[[i]]$`CD16 Mono`
  l[[i]]$Myeloid <- l[[i]]$Monocyte+l[[i]]$Neutrophil
  }

healthy.theta <- l[[1]]
ptb.theta <- l[[2]]
tbm26.theta <- l[[3]]
tbm27.theta <- l[[4]]



ptb_high_id <-  (vst_data[,c('LIMS_ID','GenXpert')] %>% filter(GenXpert=='MTB Detected High'))$LIMS_ID
ptb_medium_id <-  (vst_data[,c('LIMS_ID','GenXpert')] %>% filter(GenXpert=='MTB Detected Medium'))$LIMS_ID
ptb_low_id <-  (vst_data[,c('LIMS_ID','GenXpert')] %>% filter(GenXpert %in% c('MTB Detected Low',
                                                                              'MTB Detected Very Low',
                                                                              'MTB Not Detected')))$LIMS_ID
ptb_low.z <- ptb.z[ptb_low_id,,]
ptb_medium.z <- ptb.z[ptb_medium_id,,]
ptb_high.z <- ptb.z[ptb_high_id,,]

# WBC_df <- clinical_data_HC_CNS_IF_PTB_TBM[,c('LIMS_ID','WBC')]
# row.names(WBC_df) <- WBC_df$LIMS_ID
# healthy.thetacount <- healthy.theta*WBC_df[rownames(healthy.theta),]$WBC
# ptb.thetacount <- ptb.theta*WBC_df[rownames(ptb.theta),]$WBC
# tbm.thetacount <- tbm.theta*WBC_df[rownames(tbm.theta),]$WBC
# tbm26.thetacount <- tbm26.theta*WBC_df[rownames(tbm26.theta),]$WBC
# tbm27.thetacount <- tbm27.theta*WBC_df[rownames(tbm27.theta),]$WBC
# 
# thetacount_long <- rbind(pivot_longer(healthy.thetacount,colnames(healthy.thetacount)) %>% mutate(group = 'Healthy'),
#                     pivot_longer(ptb.thetacount,colnames(ptb.thetacount)) %>% mutate(group = 'PTB'),
#                     pivot_longer(tbm27.thetacount,colnames(tbm27.thetacount)) %>% mutate(group = 'TBM (HIV-)'),
#                     pivot_longer(tbm26.thetacount,colnames(tbm26.thetacount)) %>% mutate(group = 'TBM (HIV+)'))
# 
# ptb_vst.z <- list()
# healthy_vst.z <- list()
# tbm_vst.z <- list()
# tbm26_vst.z <- list()
# tbm27_vst.z <- list()
# 
# for(c in celltypes) {
#   ptb_vst.z[[c]] = vst(round(t(ptb.z[,,c])), nsub=nrow(ptb.z[,,c]), blind=T, fitType='parametric')
#   healthy_vst.z[[c]] = vst(round(t(healthy.z[,,c])), nsub=nrow(healthy.z[,,c]), blind=T, fitType='parametric')
#   tbm_vst.z[[c]] = vst(round(t(tbm.z[,,c])), nsub=nrow(tbm.z[,,c]), blind=T, fitType='parametric')
#   tbm26_vst.z[[c]] = vst(round(t(tbm26.z[,,c])), nsub=nrow(tbm26.z[,,c]), blind=T, fitType='parametric')
#   tbm27_vst.z[[c]] = vst(round(t(tbm27.z[,,c])), nsub=nrow(tbm27.z[,,c]), blind=T, fitType='parametric')
# }
# 
# ptb_vst.z <- simplify2array(ptb_vst.z)
# healthy_vst.z <- simplify2array(healthy_vst.z)
# tbm_vst.z <- simplify2array(tbm_vst.z)
# tbm26_vst.z <- simplify2array(tbm26_vst.z)
# tbm27_vst.z <- simplify2array(tbm27_vst.z)
# 
# CNIH4.z_long <- rbind(pivot_longer(as.data.frame(healthy_vst.z['CNIH4',,]), celltypes) %>% mutate(group = 'Healthy'),
#                     pivot_longer(as.data.frame(ptb_vst.z['CNIH4',,]), celltypes) %>% mutate(group = 'PTB'),
#                     pivot_longer(as.data.frame(tbm27_vst.z['CNIH4',,]), celltypes) %>% mutate(group = 'TBM (HIV-)'),
#                     pivot_longer(as.data.frame(tbm26_vst.z['CNIH4',,]), celltypes) %>% mutate(group = 'TBM (HIV+)'))
# 
# CD274.z_long <- rbind(pivot_longer(as.data.frame(healthy_vst.z['CD274',,]), celltypes) %>% mutate(group = 'Healthy'),
#                     pivot_longer(as.data.frame(ptb_vst.z['CD274',,]), celltypes) %>% mutate(group = 'PTB'),
#                     pivot_longer(as.data.frame(tbm27_vst.z['CD274',,]), celltypes) %>% mutate(group = 'TBM (HIV-)'),
#                     pivot_longer(as.data.frame(tbm26_vst.z['CD274',,]), celltypes) %>% mutate(group = 'TBM (HIV+)'))
```

```{r}
#theta <- as.data.frame(healthy.theta)
#theta <- as.data.frame(ptb.theta)
#theta <- as.data.frame(tbm26.theta)
#theta <- as.data.frame(tbm27.theta)

theta <- as.data.frame(healthy.theta)
theta <- theta[ order(row.names(theta)),]

theta$Neutrophil <- theta$Neutro
theta$Lymphocyte <- theta$`B naive` + theta$`B memory` + theta$`CD8 Naive`+theta$`CD8 TM` + theta$MAIT + theta$NK + theta$`CD4 TM`+theta$`CD4 Naive`+theta$`CD4 CTL`+theta$OtherT +theta$Treg #+ theta$Plasmablast
theta$Monocyte <- theta$`CD14 Mono`+theta$`CD16 Mono`
theta$Myeloid <- theta$Monocyte+theta$Neutrophil

#theta$Lymphocyte <- theta$B + theta$CD8T +theta$CD4T +theta$NK
gt['HOA7812A12','Monocyte'] <- 4.5
theta1 <- theta[row.names(gt),]
theta1 <- na.omit(theta1)
gt1 <- gt[row.names(theta1),]

gt1$Myeloid <- gt1$Monocyte+gt1$Neutrophil

regs <- plot_regs(theta1,gt1,name='PTB',show_cutoff = FALSE)
regs
```

#### Cell types Proportions

```{r}
library(ggsci)

celltypes <- colnames(ptb.theta)[18:20] #level1
celltypes <- colnames(ptb.theta)[1:17] #level2

theta_long <- rbind(pivot_longer(healthy.theta,celltypes) %>% mutate(group = 'Healthy'),
                    pivot_longer(ptb.theta,celltypes) %>% mutate(group = 'PTB'),
                    pivot_longer(tbm27.theta,celltypes) %>% mutate(group = 'TBM (HIV-)'),
                    pivot_longer(tbm26.theta,celltypes) %>% mutate(group = 'TBM (HIV+)'))

stat.test <- theta_long %>%
  group_by(name) %>%
  wilcox_test(value ~ group) %>%
  add_significance("p") %>%
  filter(p.signif !="ns") %>%
  add_y_position(step.increase = 0.06)

p <- ggplot(data=theta_long,
       aes(x = factor(group, level = c('Healthy','PTB','TBM (HIV-)','TBM (HIV+)')),
           color=factor(group, level = c('Healthy','PTB','TBM (HIV-)', 'TBM (HIV+)')),
           y=value,))+
  facet_wrap(~factor(name, level=celltypes), ncol = 6,scale='free')+
  geom_boxplot() +
  geom_point(position=position_jitterdodge(jitter.width =0.3),alpha=0.1)+
  # stat_pvalue_manual(stat.test,
  #                    label = 'p.signif',
  #                    tip.length = 0,vjust=0.2) +

  scale_color_manual(values=c("#00A087B2","#81A1C1","#CD534C","#631840"))+
  labs(color='Group')+ 
  xlab('Cell type')+
  ylab('% Cell population')+
  theme_bw()+
  theme(plot.title = element_text(size = 25),
        legend.position = "right",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_blank(),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold"),
        strip.text = element_text(face="bold",size=14))


#png(file=paste0(path_output,"/BayesPrism_CellPops.png"),res=900, units = 'in', width=17, height = 7)
p
#dev.off()
```

#### CNIH4 estimation

```{r}

```

#### Heatmap TCR signaling pathways

```{r}
library(dcanr)

all.z <- abind(healthy.z[,common,], ptb.z[,common,], tbm27.z[,common,], tbm26.z[,common,],along = 1)

Neutrophil <- c('Neutro')
Monocyte <- c('CD14 Mono','CD16 Mono')
Lymphocyte <- c('B naive', 'B memory', 'CD8 Naive', 'CD8 TM', 'MAIT', 'NK', 'CD4 TM', 'CD4 Naive','CD4 CTL','OtherT' ,'Treg')

a <- all.z[,,Neutrophil]
row.names(a) <- paste0(rownames(a), "_Neutrophil")
b <- rowSums(all.z[,,Monocyte], dims = 2)
row.names(b) <- paste0(rownames(b), "_Monocyte")
c <- rowSums(all.z[,,Lymphocyte], dims = 2)
row.names(c) <- paste0(rownames(c), "_Lymphocyte")
all.z <-rbind(a,b,c)
all.z <- round(t(all.z))
dim(all.z)


group = rep(c('Healthy','PTB','TBM27','TBM26'),times=c(30,296,207,74))
group = as.factor(rep(group,each=3))
celltype = as.factor(rep(c('Neutrophil','Monocyte','Lymphocyte'),each=607))
coldata = data.frame(group,celltype)

dds <- DESeqDataSetFromMatrix(countData = all.z,
                              colData = coldata,
                              design = ~ celltype+group)

dds$group <- relevel(dds$group, ref = "Healthy")
dds$celltype <- relevel(dds$celltype, ref = "Neutrophil")
design(dds) <- formula(~ celltype + group)

all.z_vst <- assay(vst(dds, blind=FALSE))
dim(all.z_vst)
#######################
#vst all pops all celltypes
all.z_vst <- vst(all.z,blind=FALSE)
dim(all.z_vst)
#######################
#vst each pop all celltypes
# tmp = all.z
# a = colnames(all.z)[c(1:30, 608:(608+30-1),1215:(1215+30-1))]#healthy
# b = colnames(all.z)[c(31:326, 638:(638+296-1),1245:(1245+296-1))]#ptb
# c = colnames(all.z)[c(327:533, 934:(934+207-1),1541:(1541+207-1))]#tbm27
# d = colnames(all.z)[c(534:607, 1141:(1141+74-1),1748:(1748+74-1))]#tbm26
# tmp[,a] <- vst(tmp[,a],blind=FALSE)
# tmp[,b] <- vst(tmp[,b],blind=FALSE)
# tmp[,c] <- vst(tmp[,c],blind=FALSE)
# tmp[,d] <- vst(tmp[,d],blind=FALSE)
# all.z_vst<- tmp
# dim(all.z_vst)
#######################
#m = median(all.z_vst[,1:30])


#all.z_vst_scale = t(scale(t(all.z_vst)))
# mapply(FUN = `/`, all.z_vst[1:3,1:6], m[1:3], SIMPLIFY = FALSE)
# 
# View(all.z_vst/m)

#m = apply(all.z_vst[,1:30], 1, median, na.rm=TRUE)

#all.z_vst_scale = all.z_vst-m

#all.z_vst_scale = all.z_vst_scale[complete.cases(all.z_vst_scale), ]
cell <- 'Neutrophil'
cell <- 'Monocyte'
cell <- 'Lymphocyte'
geneset <- intersect(rownames(all.z_vst_scale),TCR_signaling)#IFN_gamma$geneSymbol
sample_col <- to_vec(for(i in colnames(all.z_vst)) if(strsplit(i,"_",fixed=T)[[1]][2]==cell) i )

a = all.z_vst[geneset,sample_col]
m = apply(a[,1:30], 1, median, na.rm=TRUE)
a = a-m
#a = all.z_vst[intersect(rownames(all.z_vst_scale),geneset),]
dim(a)

## vst after
# cell <- 'Neutrophil'
# sample_col <- to_vec(for(i in colnames(all.z_vst)) if(strsplit(i,"_",fixed=T)[[1]][2]==cell) i )
# a <- all.z[,sample_col]
# a <- vst(a,blind=FALSE)
# geneset <- TCR_signaling #IFN_gamma$geneSymbol
# a = a[intersect(rownames(a),geneset),]
# a = t(scale(t(a)))
# a = a[complete.cases(a), ]
# dim(a)
##

#col = rev(brewer.pal(11, "Spectral"))
#col_fun = colorRamp2(c(0, 0.1, 0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1.0), col)

col = rev(brewer.pal(11, "Spectral"))
#col_fun = colorRamp2(c(-2.2,-1.7,-1.5,-1,-0.5,0,0.5,1,1.5,1.7,2.2), col)
col_fun = colorRamp2(c(-2.2,-1.7,-1.5,-1,-0.5,0,0.5,1,1.5,1.7,2.2), col)

col_fun = colorRamp2(c(-4,-1.2,-0.9,-0.6,-0.3,0,0.3,0.6,0.9,1.2,4), col)
col_fun = colorRamp2(c(-2.2,-1.6,-0.8,-0.4,-0.2,0,0.2,0.4,0.8,1.6,2), col)
#col_fun = colorRamp2(c(-5,-4,-3,-2,-1,0,1,2,3,4,5), col)
#col = rev(brewer.pal(5, "Spectral"))

#col_fun = colorRamp2(c(-2,-1,0,1,2), col)

# hm1 <- Heatmap(a[,1:30],
#         name = 'expression',
#         col=col_fun,
#         cluster_rows = TRUE,
#         cluster_columns = TRUE,
#         show_row_names = FALSE,
#         show_column_names = FALSE,use_raster = FALSE,       
#         column_title = "Control", 
#     column_title_side = "bottom")
# 
# hm2 <- Heatmap(a[,31:(31+87-1)],
#         name = 'expression',
#         col=col_fun,
#         cluster_rows = TRUE,
#         cluster_columns = TRUE,
#         show_row_names = FALSE,
#         show_column_names = FALSE,use_raster = FALSE,       
#         column_title = "PTB Low", 
#     column_title_side = "bottom")
# hm3 <- Heatmap(a[,118:(118+132-1)],
#         name = 'expression',
#         col=col_fun,
#         cluster_rows = TRUE,
#         cluster_columns = TRUE,
#         show_row_names = FALSE,
#         show_column_names = FALSE,use_raster = FALSE,       
#         column_title = "PTB Medium", 
#     column_title_side = "bottom")
# 
# hm4 <- Heatmap(a[,250:(250+75-1)],
#         name = 'expression',
#         col=col_fun,
#         cluster_rows = TRUE,
#         cluster_columns = TRUE,
#         show_row_names = FALSE,
#         show_column_names = FALSE,use_raster = FALSE,       
#         column_title = "PTB High", 
#     column_title_side = "bottom")
# 
# hm5 <- Heatmap(a[,325:(325+207-1)],
#         name = 'expression',
#         col=col_fun,
#         cluster_rows = TRUE,
#         cluster_columns = TRUE,
#         show_row_names = FALSE,
#         show_column_names = FALSE,use_raster = FALSE,       
#         column_title = "TBM (HIV-)", 
#     column_title_side = "bottom")
# 
# hm6 <- Heatmap(a[,532:(532+74-1)],
#         name = 'expression',
#         col=col_fun,
#         cluster_rows = TRUE,
#         cluster_columns = TRUE,
#         show_row_names = FALSE,
#         show_column_names = FALSE,use_raster = FALSE,       
#         column_title = "TBM (HIV+)", 
#     column_title_side = "bottom")
# 
# hm1+hm2+hm3+hm4+hm5+hm6

#############################################
hm1 <- Heatmap(a[,1:30],
        name = 'expression',
        col=col_fun,
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,use_raster = FALSE,       
        column_title = "Healthy", 
    column_title_side = "bottom")

hm2 <- Heatmap(a[,31:326],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,
                column_title = "PTB", 
    column_title_side = "bottom")
hm3 <- Heatmap(a[,327:533],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,        
        column_title = "TBM (HIV-)", 
    column_title_side = "bottom")

hm4 <- Heatmap(a[,534:607],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns =TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,
        column_title = "TBM (HIV+)", 
    column_title_side = "bottom")
lgd = Legend(col_fun = col_fun, title = "foo", direction = "horizontal")
hm = hm1+hm2+hm3+hm4
ht = draw(hm, column_title = cell, column_title_gp = gpar(fontsize = 16))

###################
rownames(a)[row_order(ht)]


a1 = all.z_vst_scale[,sample_col]
a1 = all.z_vst[,sample_col]
g = a1['MAPK8',1:326]

group = rep(c('Healthy','PTB'),times=c(30,296))

df = data.frame(g,group)

ggplot(data=df,
       aes(x = group,y=g))+
  geom_boxplot()

head(df)
dim(vst_data%>%select(any_of(TCR_signaling)))
```

```{r}
library(ReactomeContentService4R) # Extract data from the three pathways
library(abind)
library(comprehenr)

IFN_alpha_beta<-event2Ids(event.id = "R-HSA-909733.9")
IFN_gamma<-event2Ids(event.id = "R-HSA-877300.7")
PD_L1<-get_genelist_pathway_kegg(hsa_ID = "hsa05235") 
TCR_signaling<-get_genelist_pathway_kegg(hsa_ID = "hsa04660")

common <- Reduce(intersect, list(colnames(healthy.z[,,1]),
                                 colnames(ptb.z[,,1]),
                                 colnames(tbm.z[,,1])))

scale_row <- function(x) {
  new_range <- c(0, 1)  # You can change this to your desired range (e.g., -1, 1)
  #return((x - x.mean(axis=1, keepdims=True)) / x.std(axis=1, keepdims=True))
  return((x - min(x)) / (max(x) - min(x)) * diff(new_range) + min(new_range))
}






all.z <- abind(healthy.z[,common,],ptb.z[,common,],tbm27.z[,common,],tbm26.z[,common,],along = 1)

# 
Neutrophil <- c('Neutro')
Monocyte <- c('CD14 Mono','CD16 Mono')
Lymphocyte <- c('B naive', 'B memory', 'CD8 Naive', 'CD8 TM', 'MAIT', 'NK', 'CD4 TM', 'CD4 Naive','CD4 CTL','OtherT' ,'Treg')
# 
# a <- all.z[,,Neutrophil]
# row.names(a) <- paste0(rownames(a), "_Neutrophil")
# b <- rowSums(all.z[,,Monocyte], dims = 2)
# row.names(b) <- paste0(rownames(b), "_Monocyte")
# c <- rowSums(all.z[,,Lymphocyte], dims = 2)
# row.names(c) <- paste0(rownames(c), "_Lymphocyte")
# all.z <-rbind(a,b,c)
# all.z <- round(t(all.z))
# dim(all.z)




all_neutro.z <- all.z[,,'Neutro']
all_neutro.z <- round(t(all_neutro.z))
all_neutro.z_vst <- vst(all_neutro.z,blind=FALSE)

dim(all_neutro.z_vst)

geneset <- TCR_signaling #IFN_gamma$geneSymbol

a = all_neutro.z_vst[intersect(rownames(all_neutro.z_vst),geneset),]
dim(a)
#a = scale(a,center=TRUE,scale=FALSE)
#hist(a)
a = as.data.frame(t(apply(a, 1, scale_row)))
a = a[complete.cases(a), ]

a = a[rowSums(a[])>100,]

hist(rowSums(a[]))

col = rev(brewer.pal(11, "Spectral"))
col_fun = colorRamp2(c(0, 0.1, 0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1.0), col)


library(RColorBrewer)

hm1 <- Heatmap(a[,1:30],
        name = 'expression',
        col=col_fun,
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,use_raster = FALSE,       
        column_title = "Control", 
    column_title_side = "bottom")

hm2 <- Heatmap(a[,31:326],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,
                column_title = "PTB", 
    column_title_side = "bottom")

hm3 <- Heatmap(a[,327:533],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,        
        column_title = "TBM (HIV-)", 
    column_title_side = "bottom")

hm4 <- Heatmap(a[,534:607],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns =TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,
        column_title = "TBM (HIV+)", 
    column_title_side = "bottom")

lgd = Legend(col_fun = col_fun, title = "foo", direction = "horizontal")
hm = hm1+hm2+hm3+hm4
draw(hm, column_title = cell, column_title_gp = gpar(fontsize = 16))


#####################################################################################

hm1

hm2 <- Heatmap(all_neutro.z_vst[intersect(rownames(all_neutro.z_vst),geneset),31:326],
               name = 'expression',
        #col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,
                column_title = "PTB", 
    column_title_side = "bottom")
hm1+hm2



CNIH4 = all_neutro.z_vst['CD274',]
group = rep(c('Healthy','PTB'),times=c(30,296))

df = data.frame(CNIH4,group)

ggplot(data=df,
       aes(x = group,y=CNIH4))+
  geom_boxplot()
head(df)

ggplot(df,)


boxplot(all_neutro.z_vst['CNIH4',1:30])
boxplot(all_neutro.z_vst['CNIH4',31:326])
```

```{r}
library(ReactomeContentService4R) # Extract data from the three pathways
library(abind)
library(comprehenr)

IFN_alpha_beta<-event2Ids(event.id = "R-HSA-909733.9")
IFN_gamma<-event2Ids(event.id = "R-HSA-877300.7")
PD_L1<-get_genelist_pathway_kegg(hsa_ID = "hsa05235") 
TCR_signaling<-get_genelist_pathway_kegg(hsa_ID = "hsa04660")

common <- Reduce(intersect, list(colnames(healthy.z[,,1]),
                                 colnames(ptb.z[,,1]),
                                 colnames(tbm.z[,,1])))

celltypes <- colnames(healthy.z[,1,])
all.z <- abind(healthy.z[,common,],ptb.z[,common,],tbm27.z[,common,],tbm26.z[,common,],along = 1)
dim(all.z)

Neutrophil <- c('Neutro')
Monocyte <- c('CD14 Mono','CD16 Mono')
Lymphocyte <- c('B naive', 'B memory', 'CD8 Naive', 'CD8 TM', 'MAIT', 'NK', 'CD4 TM', 'CD4 Naive','CD4 CTL','OtherT' ,'Treg')

a <- all.z[,,Neutrophil]
row.names(a) <- paste0(rownames(a), "_Neutrophil")
b <- rowSums(all.z[,,Monocyte], dims = 2)
row.names(b) <- paste0(rownames(b), "_Monocyte")
c <- rowSums(all.z[,,Lymphocyte], dims = 2)
row.names(c) <- paste0(rownames(c), "_Lymphocyte")
all.z <-rbind(a,b,c)
all.z <- round(t(all.z))
dim(all.z)


View(all.z)
group = as.factor(rep(c('Healthy','PTB','TBM27','TBM26'),times=c(30,296,207,74)))
celltype = as.factor(rep(c('Neutrophil','Monocyte','Lymphocyte'),times=c(607,607,607)))
factor = as.factor(paste0(group,celltype))
coldata <- data.frame(group,celltype,factor)

dds <- DESeqDataSetFromMatrix(countData = all.z,
                              colData = coldata,
                              design= ~ factor)

all.z_vst <- vst(dds,blind=FALSE,fitType = "parametric")

dim(all.z_vst)

cell <- 'Neutrophil'
sample_col <- to_vec(for(i in colnames(all.z_vst)) if(strsplit(i,"_",fixed=T)[[1]][2]==cell) i )


geneset <- IFN_gamma$geneSymbol
tmp_vst <- assay(all.z_vst[intersect(rownames(all.z_vst),geneset),sample_col])
dim(tmp_vst)

col=c("blue",'yellow',"red")
col_fun = colorRamp2(c(6,8,10), col)

# hm1 <- Heatmap(tmp_vst,
#         name = 'expression',
#         col=col_fun,
#         cluster_rows = TRUE,
#         cluster_columns = FALSE,
#         show_row_names = FALSE,
#         show_column_names = FALSE,#use_raster = TRUE,       
#         column_title = "Control", 
#     column_title_side = "bottom")
# 
# hm1
# 
# dim(tmp_vst)

hm1 <- Heatmap(tmp_vst[,1:30],
        name = 'expression',
        col=col_fun,
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,use_raster = FALSE,       
        column_title = "Control", 
    column_title_side = "bottom")

hm2 <- Heatmap(tmp_vst[,31:326],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,
                column_title = "PTB", 
    column_title_side = "bottom")

hm3 <- Heatmap(tmp_vst[,327:533],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,        
        column_title = "TBM (HIV-)", 
    column_title_side = "bottom")

hm4 <- Heatmap(tmp_vst[,534:607],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,
        column_title = "TBM (HIV+)", 
    column_title_side = "bottom")

lgd = Legend(col_fun = col_fun, title = "foo", direction = "horizontal")
hm = hm1+hm2+hm3+hm4
draw(hm, column_title = cell, column_title_gp = gpar(fontsize = 16))

dim(ptb.z)

################################################################################################

cell <- "Neutro"
tmp <- all.z[,,'Neutro']
tmp <- round(t(tmp))
tmp_vst <- vst(tmp,blind=FALSE,fitType = "parametric")

geneset <- TCR_signaling#IFN_gamma$geneSymbol
tmp_vst0 <- tmp_vst[intersect(rownames(tmp_vst),geneset),]

hm1 <- Heatmap(tmp_vst,
        name = 'expression',
        col=col_fun,
        cluster_rows = TRUE,
        cluster_columns = FALSE,
        show_row_names = FALSE,
        show_column_names = FALSE,use_raster = TRUE,       
        column_title = "Control", 
    column_title_side = "bottom")
hm1


```

```{r}


level1 <- c('Neutrophil','Monocyte','Lymphocyte')
level2 <- names(all.z[1,1,])
level<- level1

Monocyte <- c('CD14 Mono','CD16 Mono')
Lymphocyte <- c('B naive', 'B memory', 'CD8 Naive', 'CD8 TM', 'MAIT', 'NK', 'CD4 TM', 'CD4 Naive','CD4 CTL','OtherT' ,'Treg', 'Plasmablast')
Neutrophil <- c('Neutro')

a <- all.z[,,Neutrophil]
row.names(a) <- paste0(rownames(a), "_Neutrophil")
b <- rowSums(all.z[,,Monocyte], dims = 2)
row.names(b) <- paste0(rownames(b), "_Monocyte")
c <- rowSums(all.z[,,Lymphocyte], dims = 2)
row.names(c) <- paste0(rownames(c), "_Lymphocyte")

all.z <-rbind(a,b,c)
all.z0 <- round(t(all.z))
dim(all.z0)


cell <- 'Neutrophil'
sample_col <- to_vec(for(i in colnames(all.z0)) if(strsplit(i,"_",fixed=T)[[1]][2]==cell) i )
all.z0 <- all.z0[,sample_col]
geneset <- TCR_signaling#IFN_gamma$geneSymbol
all.z0 <- all.z0[intersect(rownames(all.z0),geneset),]

dim(all.z0)

all.z0_vst <- vst(all.z0,nsub = 98)

all.z_vst1 <- varianceStabilizingTransformation(all.z0)




dds$factor <- factor(dds$factor, levels = unique(coldata$factor))
# dds$group <- factor(dds$group, levels = c('Healthy','PTB','TBM27','TBM26'))
# dds$celltype <- factor(dds$celltype, levels = celltypes)
all.z_vst <- vst(dds, nsub=98,blind=TRUE)
hist(assay(all.z_vst))


cell <- 'Monocyte'
sample_col <- to_vec(for(i in colnames(all.z_vst)) if(strsplit(i,"_",fixed=T)[[1]][2]==cell) i )
all.z_vst1 <- all.z_vst[,sample_col]
all.z_vst1 <- assay(all.z_vst1)
dim(all.z_vst1)

# all.z0 <- round(t(a))
# all.z_vst1 <- vst(all.z0 ,blind=FALSE,fitType = 'parametric')
# all.z_vst1 <- all.z_vst1[intersect(rownames(all.z_vst1),geneset),]

col=c("blue",'yellow',"red")
col_fun = colorRamp2(c(4,8,12), col)
# hm1 <- Heatmap(assay(all.z_vst),
#         col=col_fun,
#         cluster_rows = TRUE,
#         cluster_columns = FALSE,
#         show_row_names = FALSE,
#         show_column_names = FALSE,
#         use_raster = FALSE,
#         column_title = "Control",
#     column_title_side = "bottom")
# hm1


hm1 <- Heatmap(all.z_vst1[,1:30],
        name = 'expression',
        col=col_fun,
        cluster_rows = TRUE,
        cluster_columns = FALSE,
        show_row_names = FALSE,
        show_column_names = FALSE,use_raster = FALSE,       
        column_title = "Control", 
    column_title_side = "bottom")

hm2 <- Heatmap(all.z_vst1[,31:326],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = FALSE,
        show_row_names = FALSE,
        show_column_names = FALSE,
                column_title = "PTB", 
    column_title_side = "bottom")

hm3 <- Heatmap(all.z_vst1[,327:533],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = FALSE,
        show_row_names = FALSE,
        show_column_names = FALSE,        
        column_title = "TBM (HIV-)", 
    column_title_side = "bottom")

hm4 <- Heatmap(all.z_vst1[,534:607],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = FALSE,
        show_row_names = FALSE,
        show_column_names = FALSE,
        column_title = "TBM (HIV+)", 
    column_title_side = "bottom")

lgd = Legend(col_fun = col_fun, title = "foo", direction = "horizontal")
hm = hm1+hm2+hm3+hm4
draw(hm, column_title = "Neutrophil", column_title_gp = gpar(fontsize = 16))
```

```{r}
library(ReactomeContentService4R) # Extract data from the three pathways
library(abind)

IFN_alpha_beta<-event2Ids(event.id = "R-HSA-909733.9")
IFN_gamma<-event2Ids(event.id = "R-HSA-877300.7")
PD_L1<-get_genelist_pathway_kegg(hsa_ID = "hsa05235") 
TCR_signaling<-get_genelist_pathway_kegg(hsa_ID = "hsa04660")

common <- Reduce(intersect, list(colnames(healthy.z[,,1]),
                                 colnames(ptb.z[,,1]),
                                 colnames(tbm.z[,,1])))
celltypes <- colnames(healthy.z[,1,])
all.z <- abind(healthy.z[,common,],ptb.z[,common,],tbm.z[,common,],along = 1)

Monocyte <- c('CD14 Mono','CD16 Mono')
Lymphocyte <- c('B naive', 'B memory', 'CD8 Naive', 'CD8 TM', 'MAIT', 'NK', 'CD4 TM', 'CD4 Naive','CD4 CTL','OtherT' ,'Treg', 'Plasmablast')
Neutrophil <- c('Neutro')

a <- all.z[,,Neutrophil]
row.names(a) <- paste0(rownames(a), "_Neutrophil")
b <- rowSums(all.z[,,Monocyte], dims = 2)
row.names(b) <- paste0(rownames(b), "_Monocyte")
c <- rowSums(all.z[,,Lymphocyte], dims = 2)
row.names(c) <- paste0(rownames(c), "_Lymphocyte")
all.z <- rbind(a,b,c)
dim(all.z)

all.z0 <- round(t(all.z))
#all.z0 <- all.z0[rowSums(all.z0[])>0,] 
#all.z0 <- all.z0[rowSums(all.z0)>10319,]
dim(all.z0)
all.z_vst <- as.data.frame(vst(all.z0,
                         blind=F,
                         #nsub=500, 
                         fitType='parametric'))

library(comprehenr)
cell <- 'Lymphocyte'
geneset <- TCR_signaling #IFN_gamma$geneSymbol
sample_col <- to_vec(for(i in colnames(all.z_vst)) if(strsplit(i,"_",fixed=T)[[1]][2]==cell) i )
all.z_vst1 <- all.z_vst[,sample_col]
all.z_vst1 <- all.z_vst1[intersect(rownames(all.z_vst1),geneset),]
dim(all.z_vst1)

col=c("blue",'yellow',"red")
col_fun = colorRamp2(c(5,7,10), col)

hm1 <- Heatmap(all.z_vst1,
        name = 'expression',
        #col=col_fun,
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,use_raster = FALSE,       
        column_title = "Control", 
    column_title_side = "bottom")
hm1


hm1 <- Heatmap(all.z_vst1[,1:30],
        name = 'expression',
        col=col_fun,
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,use_raster = FALSE,       
        column_title = "Control", 
    column_title_side = "bottom")

hm2 <- Heatmap(all.z_vst1[,31:326],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = FALSE,
        show_row_names = FALSE,
        show_column_names = FALSE,
                column_title = "PTB", 
    column_title_side = "bottom")

hm3 <- Heatmap(all.z_vst1[,327:533],
               name = 'expression',
        col=col_fun,
        cluster_rows = TRUE,
        cluster_columns = FALSE,
        show_row_names = FALSE,
        show_column_names = FALSE,        
        column_title = "TBM (HIV-)", 
    column_title_side = "bottom")

hm4 <- Heatmap(all.z_vst1[,534:607],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = FALSE,
        show_row_names = FALSE,
        show_column_names = FALSE,
        column_title = "TBM (HIV+)", 
    column_title_side = "bottom")
lgd = Legend(col_fun = col_fun, title = "foo", direction = "horizontal")
hm = hm1+hm2+hm3+hm4
draw(hm, column_title = "Neutrophil", column_title_gp = gpar(fontsize = 16))
```

```{r}
library(ReactomeContentService4R) # Extract data from the three pathways

IFN_alpha_beta<-event2Ids(event.id = "R-HSA-909733.9")
IFN_gamma<-event2Ids(event.id = "R-HSA-877300.7")
PD_L1<-get_genelist_pathway_kegg(hsa_ID = "hsa05235") 
TCR_signaling<-get_genelist_pathway_kegg(hsa_ID = "hsa04660")
```

```{r}
#[1] "Neutro"      "CD14 Mono"   "CD4 TM"      "NK"          "CD8 TM"      "CD16 Mono"   "CD8 Naive"   "CD4 Naive"   "CD4 CTL"    
# [10] "B naive"     "MAIT"        "OtherT"      "B memory"    "pDC"         "Treg"        "cDC"         "Plasmablast"

geneset <- IFN_gamma$geneSymbol
#geneset <- TCR_signaling
cell <- 'Neutro'

# library(ReactomeContentService4R) # Extract data from the three pathways
# 
# IFN_alpha_beta<-event2Ids(event.id = "R-HSA-909733.9")
# IFN_gamma<-event2Ids(event.id = "R-HSA-877300.7")
# PD_L1<-get_genelist_pathway_kegg(hsa_ID = "hsa05235") 
# TCR_signaling<-get_genelist_pathway_kegg(hsa_ID = "hsa04660")
# 
# ptb_tcr <- as.data.frame(ptb.z[,,'Neutro']) %>% select(any_of(IFN_gamma$geneSymbol))
# ptb_tcr <- floor(t(ptb_tcr) + 0.5)
# ptb_tcr_vst <- as.matrix(vst(ptb_tcr, nsub=50, 
#                              blind=T, fitType='parametric'
#                              ))
# # tbm26_tcr <- as.data.frame(tbm26.z[,,'Neutro']) %>% select(any_of(TCR_signaling))
# # tbm26_tcr <- floor(t(tbm26_tcr) + 0.5)
# # tbm26_tcr_vst <- as.matrix(vst(tbm26_tcr, nsub=70, 
# #                              #blind=T, fitType='parametric'
# #                              ))
# 
# healthy_tcr <- as.data.frame(healthy.z[,,'Neutro']) %>% select(any_of(IFN_gamma$geneSymbol))
# healthy_tcr <- floor(t(healthy_tcr) + 0.5)
# healthy_tcr_vst <- as.matrix(vst(healthy_tcr,nsub=50,blind=T, fitType='parametric'))

##########################################################################################

ptb_tcr <- as.data.frame(ptb.z[,,cell]) %>% select(any_of(geneset))
ptb_tcr <- ceiling(t(ptb_tcr))
tbm26_tcr <- as.data.frame(tbm26.z[,,cell]) %>% select(any_of(geneset))
tbm26_tcr <- ceiling(t(tbm26_tcr))
tbm27_tcr <- as.data.frame(tbm27.z[,,cell]) %>% select(any_of(geneset))
tbm27_tcr <- ceiling(t(tbm27_tcr))
healthy_tcr <- as.data.frame(healthy.z[,,cell]) %>% select(any_of(geneset))
healthy_tcr <- ceiling(t(healthy_tcr))

tmp <- cbind(healthy_tcr, ptb_tcr, tbm27_tcr, tbm26_tcr)#+1
#tmp = tmp+1
tmp = tmp[rowSums(tmp)>630,]
#tmp = tmp[rowSums(tmp[1:30,])>30,]

dim(tmp)
tmp_vst <- as.matrix(vst(t(tmp),
                         blind=F,
                         nsub=600, 
                         fitType='local')) %>% t()

col=c("blue",'yellow',"red")
#col = c('#bae1ff','#ffffba','#ffb3ba')
col_fun = colorRamp2(c(6,9.5,11.5), col)
#col_fun = colorRamp2(c(3,1,10), col)
# hm <- Heatmap(tmp_vst,
#         name = 'corr.',
#         col=col_fun,
#         cluster_rows = FALSE,
#         cluster_columns = FALSE,
#         show_row_names = FALSE,
#         show_column_names = FALSE)
# hm

hm1 <- Heatmap(tmp_vst[,1:30],
        name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = FALSE,
        show_row_names = FALSE,
        show_column_names = FALSE,        
        column_title = "Control", 
    column_title_side = "bottom")

hm2 <- Heatmap(tmp_vst[,31:326],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = FALSE,
        show_row_names = FALSE,
        show_column_names = FALSE,
                column_title = "PTB", 
    column_title_side = "bottom")

hm3 <- Heatmap(tmp_vst[,327:533],
               name = 'expression',
        col=col_fun,
        cluster_rows = TRUE,
        cluster_columns = FALSE,
        show_row_names = FALSE,
        show_column_names = FALSE,        
        column_title = "TBM (HIV-)", 
    column_title_side = "bottom")

hm4 <- Heatmap(tmp_vst[,534:607],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = FALSE,
        show_row_names = FALSE,
        show_column_names = FALSE,
        column_title = "TBM (HIV+)", 
    column_title_side = "bottom")
lgd = Legend(col_fun = col_fun, title = "foo", direction = "horizontal")
hm = hm1+hm2+hm3+hm4
draw(hm, column_title = "Neutrophil", column_title_gp = gpar(fontsize = 16))
#draw(lgd, x = unit(1, "cm"), y = unit(1, "cm"), just = c("left", "bottom"))

# cor_method = 'spearman'
# mat <- cor(df, use='pairwise.complete.obs',method = cor_method)
# p.mat <- cor.pval.test(df, df, method=cor_method, use='pairwise.complete.obs')
# mat[p.mat>0.05] = 0
# library(nord)
# library(NatParksPalettes)
# nord_show_palette("baie_mouton")
# 
# col=c("blue",'yellow',"red")
# col_fun = colorRamp2(c(0,8,10), col)
# 
# hm1 <- Heatmap(ptb_tcr_vst,
#         name = 'corr.',
#         col=col_fun,
#         cluster_rows = FALSE,
#         cluster_columns = FALSE,
#         show_row_names = FALSE,
#         show_column_names = FALSE)
# 
# hm2 <- Heatmap(healthy_tcr_vst,
#         name = 'corr.',
#         col=col_fun,
#         cluster_rows = TRUE,
#         cluster_columns = FALSE,
#         show_row_names = FALSE,
#         show_column_names = FALSE)
# 
# # hm3 <- Heatmap(tbm26_tcr_vst,
# #         name = 'corr.',
# #         #col=col_fun,
# #         cluster_rows = FALSE,
# #         cluster_columns = FALSE,
# #         show_row_names = FALSE,
# #         show_column_names = FALSE)
# 
# hm2+hm1
```

```{r}
library(abind)

common <- Reduce(intersect, list(colnames(healthy.z[,,1]),
                                 colnames(ptb.z[,,1]),
                                 colnames(tbm.z[,,1])))
celltypes <- colnames(healthy.z[,1,])
all.z <- abind(healthy.z[,common,],ptb.z[,common,],tbm.z[,common,],along = 1)

samples <- rownames(all.z[,1,])
genes <- colnames(all.z[,,1])

dim(all.z) <- c(607*17 , 4854)

a <- samples
b <- celltypes
rownames(all.z) <- c(outer(a,b,FUN=paste,sep="."))
colnames(all.z) <- genes

dim(all.z)
########################################################################################

all.z0 <- round(t(all.z))
all.z0 <- all.z0[rowSums(all.z0[])>0,] 
#all.z0 <- all.z0[rowSums(all.z0)>10319,]
dim(all.z0)
all.z_vst <- as.data.frame(vst(all.z0,
                         blind=F,
                         #nsub=500, 
                         fitType='parametric'))

########################################################################################
hist(vst(round(t(ptb.z[,,'Neutro']))))
tmp <- vst(round(t(all.z)))
library("DESeq2")
dds <- DESeqDataSetFromMatrix(countData = all.z,)
rownames(all.z)
to_vec(for(i in samples) ifelse(i %in% ptb_id,'PTB','ha'))

healthy_id <-names(healthy.z[, 1, 1])
ptb_id <-names(ptb.z[, 1, 1])
tbm26_id <- id26TB
tbm27_id<-id27TB

########################################################################################
cell <- 'Neutro'

library(comprehenr)

geneset <- TCR_signaling #IFN_gamma$geneSymbol
sample_col <- to_vec(for(i in colnames(all.z_vst)) if(strsplit(i,".",fixed=T)[[1]][2]==cell) i )
all.z_vst1 <- all.z_vst[,sample_col]
all.z_vst1 <- all.z_vst1[intersect(rownames(all.z_vst1),geneset),]
dim(all.z_vst1)

col=c("blue",'yellow',"red")
col_fun = colorRamp2(c(-5,0,5), col)

hm1 <- Heatmap(all.z_vst1[,1:30],
        name = 'expression',
        col=col_fun,
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,use_raster = FALSE,       
        column_title = "Control", 
    column_title_side = "bottom")

hm2 <- Heatmap(all.z_vst1[,31:326],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,use_raster = FALSE,
                column_title = "PTB", 
    column_title_side = "bottom")

hm3 <- Heatmap(all.z_vst1[,327:533],
               name = 'expression',
        col=col_fun,
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,
        use_raster = FALSE,
        column_title = "TBM (HIV-)", 
    column_title_side = "bottom")

hm4 <- Heatmap(all.z_vst1[,534:607],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,
        column_title = "TBM (HIV+)",
        use_raster = FALSE,
    column_title_side = "bottom")
hm = hm1+hm2+hm3+hm4
draw(hm, column_title = cell, column_title_gp = gpar(fontsize = 16))
```

```{r}
cell <- 'NK'

common <- Reduce(intersect, list(colnames(healthy.z[,,cell]),
                       colnames(ptb.z[,,cell]),
                       colnames(tbm.z[,,cell])))

all.z <- rbind(healthy.z[,common,cell],ptb.z[,common,cell],tbm.z[,common,cell])
all.z <- ceiling(t(all.z))
all.z <- all.z[rowSums(all.z)>630,]
tmp_vst <- as.data.frame(vst(t(all.z),
                         blind=F,
                         nsub=500, 
                         fitType='parametric')) %>% t()

geneset <- IFN_gamma$geneSymbol
geneset <- TCR_signaling #IFN_gamma$geneSymbol
all.z_vst <- tmp_vst[intersect(rownames(tmp_vst),geneset),]

col=c("blue",'yellow',"red")
col_fun = colorRamp2(c(6,7,8), col)

hm1 <- Heatmap(all.z_vst[,1:30],
        name = 'expression',
        col=col_fun,
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,use_raster = FALSE,       
        column_title = "Control", 
    column_title_side = "bottom")

hm2 <- Heatmap(all.z_vst[,31:326],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,use_raster = FALSE,
                column_title = "PTB", 
    column_title_side = "bottom")

hm3 <- Heatmap(all.z_vst[,327:533],
               name = 'expression',
        col=col_fun,
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,
        use_raster = FALSE,
        column_title = "TBM (HIV-)", 
    column_title_side = "bottom")

hm4 <- Heatmap(all.z_vst[,534:607],
               name = 'expression',
        col=col_fun,
        cluster_rows = FALSE,
        cluster_columns = TRUE,
        show_row_names = FALSE,
        show_column_names = FALSE,
        column_title = "TBM (HIV+)",
        use_raster = FALSE,
    column_title_side = "bottom")
#lgd = Legend(col_fun = col_fun, title = "foo", direction = "horizontal")
hm = hm1+hm2+hm3+hm4
draw(hm, column_title = "Neutrophil", column_title_gp = gpar(fontsize = 16))
```

```{r}
library(abind)

common <- Reduce(intersect, list(colnames(healthy.z[,,1]),
                                 colnames(ptb.z[,,1]),
                                 colnames(tbm.z[,,1])))
celltypes <- colnames(healthy.z[,1,])
all.z <- abind(healthy.z[,common,],ptb.z[,common,],tbm.z[,common,],along = 1)

samples <- rownames(all.z[,1,])
genes <- colnames(all.z[,,1])

dim(all.z) <- c(607*17 , 4854)

a <- samples
b <- celltypes
rownames(all.z) <- c(outer(a,b,FUN=paste,sep="."))
colnames(all.z) <- genes

dim(all.z)
```

#### 3.5.2. Cell counts

```{r}
stat.test <- thetacount_long %>% 
  group_by(name) %>%
  wilcox_test(value ~ group) %>%
  add_significance("p") %>%
  #filter(p.signif !="ns") %>%
  add_y_position(step.increase = 0.2)

p <- ggplot(data=thetacount_long,
       aes(x = factor(group, level = c('Healthy','PTB','TBM (HIV-)','TBM (HIV+)')),
           color=factor(group, level = c('Healthy','PTB','TBM (HIV-)','TBM (HIV+)')),
           y=value,))+
  facet_wrap(~factor(name, 
                     level=c('B','CD4T','CD8T','NK','Monocyte','Neutrophil')), scale='free', 
             ncol = 6)+
  geom_boxplot() +
  geom_point(position=position_jitterdodge(jitter.width =0.3),alpha=0.1)+
  scale_color_manual(values=c("#A3BE8C","#81A1C1","#B48EAD","#D08770"))+
  labs(color='Group')+ 
  xlab('Cell type')+
  scale_y_continuous(trans='log10')+
  #ylab('Log10 cell count')+
  # stat_pvalue_manual(stat.test,
  #                    label = 'p.signif',
  #                    tip.length = 0,vjust=0.5)+
  theme_bw()+
  theme(plot.title = element_text(size = 25),
        legend.position = "right",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_blank(),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold"),
        strip.text = element_text(face="bold",size=14))


#png(file=paste0(path_output,"/BayesPrism_CellCounts.png"),res=900, units = 'in', width=17, height = 7)
p
#dev.off()
```

#### 3.5.3. CNIH4 estimation

```{r}
stat.test <- CNIH4.z_long %>% 
  group_by(name) %>%
  wilcox_test(value ~ group) %>%
  add_significance("p") %>%
  #filter(p.signif !="ns") %>%
  add_y_position(step.increase = 0.04)

p <- ggplot(data=CNIH4.z_long,
       aes(x = factor(group, level = c('Healthy',
                                         'PTB',
                                         'TBM (HIV-)',
                                         'TBM (HIV+)')),
           y=value,
           color=factor(group, level = c('Healthy',
                                         'PTB',
                                         'TBM (HIV-)',
                                         'TBM (HIV+)'
                                         ))))+
  facet_wrap(~factor(name, level=c('B','CD4T','CD8T','NK','Monocyte','Neutrophil')), 
             ncol = 6)+
  geom_boxplot() +
  geom_point(position=position_jitterdodge(jitter.width =0.3),
             alpha=0.1)+
  scale_color_manual(values=c("#A3BE8C",
                              "#81A1C1",
                              "#B48EAD",
                              "#D08770"
                              ))+
  labs(color='Group')+ 
  xlab('Cell type')+
  ylab('CNIH4 estimation')+
  theme_bw()+
  theme(plot.title = element_text(size = 25),
        legend.position = "right",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_blank(),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold"),
        strip.text = element_text(face="bold",size=14)) +
  stat_pvalue_manual(stat.test, 
                     label = 'p.signif',
                     tip.length = 0,vjust=1)

#png(file=paste0(path_output,"/BayesPrism_zCNIH4.png"),res=900, units = 'in', width=17, height = 7)
p
#dev.off()
```

#### 3.5.4. CD274/PD-L1 Estimation

```{r}
stat.test <- CD274.z_long %>% 
  group_by(name) %>%
  wilcox_test(value ~ group) %>%
  add_significance("p") %>%
  #filter(p.signif !="ns") %>%
  add_y_position(step.increase = 0.04)

p <- ggplot(data=CD274.z_long,
       aes(x = factor(group, level = c('Healthy',
                                         'PTB',
                                         'TBM (HIV-)',
                                         'TBM (HIV+)')),
           y=value,
           color=factor(group, level = c('Healthy',
                                         'PTB',
                                         'TBM (HIV-)',
                                         'TBM (HIV+)'
                                         ))))+
  facet_wrap(~factor(name, level=c('B','CD4T','CD8T','NK','Monocyte','Neutrophil')), 
             ncol = 6)+
  geom_boxplot() +
  geom_point(position=position_jitterdodge(jitter.width =0.3),
             alpha=0.1)+
  scale_color_manual(values=c("#A3BE8C",
                              "#81A1C1",
                              "#B48EAD",
                              "#D08770"
                              ))+
  labs(color='Group')+ 
  xlab('Cell type')+
  ylab('CD274 estimation')+
  theme_bw()+
  theme(plot.title = element_text(size = 25),
        legend.position = "right",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_blank(),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold"),
        strip.text = element_text(face="bold",size=14)) +
  stat_pvalue_manual(stat.test, 
                     label = 'p.signif',
                     tip.length = 0,vjust=1)

#png(file=paste0(path_output,"/BayesPrism_zCNIH4.png"),res=900, units = 'in', width=17, height = 7)
p
#dev.off()
```

### 3.5. Deconvolution Analysis using BayesPrism

```{r}
library(comprehenr)

vsd_adj <- get(load(file=paste0(path_input,"Raw_data/vsd_TBM_PTB_HV_batch_adjusted.Rdata")))
vsd_adj_study <- get(load(file=paste0(path_input,"Raw_data/vsd_TBM_PTB_HV_batch_adjusted_bySTUDY.Rdata")))
vsd_before<- get(load(file=paste0(path_input,"Raw_data/vsd_TBM_PTB_HV_batch_before.Rdata")))

# vsd_adj <- get(load(file=paste0(path_input,"Raw_data/vsd_26TB_batch_adjusted.Rdata")))
# vsd_before <- get(load(file=paste0(path_input,"Raw_data/vsd_26TB_batch_before.Rdata")))

c1 = c('HOA11355A060','HOA11355A330','HOA11355A334','HOA11355A373','HOA11355A374') #pp1
c2 = c('HOA7812A5','HOA7812A1','HOA7812A12','HOA7812A13','HOA7812A14') #pp2

tmp1 <- vsd_before[,c1]%>% as.data.frame()
tmp2 <- vsd_before[,c2]%>% as.data.frame()


tmp3 <- vsd_adj[,c1]%>% as.data.frame()
tmp4 <- vsd_adj[,c2]%>% as.data.frame()
to_vec(for(i in 1:5) sqrt(sum((tmp3[,c1[i]]-tmp4[,c2[i]])^2)))




tmp5 <- vsd_adj_study[,c1]%>% as.data.frame()
tmp6 <- vsd_adj_study[,c2]%>% as.data.frame()
to_vec(for(i in 1:5) sqrt(sum((tmp5[,c1[i]]-tmp6[,c2[i]])^2)))

#
to_vec(for(i in 1:5) sqrt(sum((tmp1[,c1[i]]-tmp3[,c1[i]])^2)))
to_vec(for(i in 1:5) sqrt(sum((tmp1[,c1[i]]-tmp5[,c1[i]])^2)))
to_vec(for(i in 1:5) sqrt(sum((tmp3[,c1[i]]-tmp5[,c1[i]])^2)))

to_vec(for(i in 1:5) sqrt(sum((tmp2[,c2[i]]-tmp4[,c2[i]])^2)))
to_vec(for(i in 1:5) sqrt(sum((tmp2[,c2[i]]-tmp6[,c2[i]])^2)))
to_vec(for(i in 1:5) sqrt(sum((tmp4[,c2[i]]-tmp6[,c2[i]])^2)))
```

#### 3.5.0. Extract data from BayesPrism results

```{r}

get_theta <- function(bp) {
  theta <- get.fraction (bp=bp,
            which.theta="final",
            state.or.type="type")
  
  theta <- as.data.frame(theta)
  row.names(theta) <- gsub("X", "", row.names(theta))
  row.names(theta) <- gsub("_PBMC", "", row.names(theta))
  theta <- theta[ order(row.names(theta)), ]
  theta$Lymphocyte <- theta$B + theta$CD4T + theta$CD8T + theta$NK
  return(theta)
} #bayesprism

get_theta1 <- function(bp) {
  theta <- as.data.frame(t(bp@theta))
  row.names(theta) <- gsub("X", "", row.names(theta))
  row.names(theta) <- gsub("_PBMC", "", row.names(theta))
  theta <- theta[ order(row.names(theta)), ]
  theta$Lymphocyte <- theta$B + theta$CD4T + theta$CD8T + theta$NK
  return(theta)
} #instaprism


#load BayesPrism result
# bp.res_healthy <- get(load(file=paste0(path_input,"Data/bp.res_healthy.rdata")))
# bp.res_ptb <- get(load(file=paste0(path_input,"Data/bp.res_ptb.rdata")))
# bp.res_tbm <- get(load(file=paste0(path_input,"Data/bp.res_tbm.rdata")))

bp.res_healthy <- get(load(file=paste0(path_input,"Data/bp.res_healthy_newman_01_01.rdata")))
bp.res_ptb <- get(load(file=paste0(path_input,"Data/bp.res_ptb_newman_01_01.rdata")))
bp.res_tbm <- get(load(file=paste0(path_input,"Data/bp.res_tbm_newman_01_01.rdata")))


# Get cell proportion for each patient
id26TB <- All_sample_RNAseq %>% subset(Study %in% c("26TB") & timepoint=='D0') %>% select('LIMS_ID')
id27TB <- All_sample_RNAseq %>% subset(Study %in% c("27TB") & timepoint=='D0') %>% select('LIMS_ID')

theta_ptb <- get_theta(bp.res_ptb)
theta_tbm <- get_theta1(bp.res_tbm)
theta_tbm26 <- theta_tbm[id26TB$LIMS_ID,]
theta_tbm27 <- theta_tbm[id27TB$LIMS_ID,]
theta_healthy <- get_theta1(bp.res_healthy)

theta_ptb_long <- pivot_longer(theta_ptb,CD8T:Neutrophil)
theta_ptb_long$group <- rep('PTB',each=dim(theta_ptb_long)[1])
theta_tbm_long <- pivot_longer(theta_tbm,CD8T:Neutrophil)
theta_tbm_long$group <- rep('TBM',each=dim(theta_tbm_long)[1])
theta_tbm26_long <- pivot_longer(theta_tbm26,CD8T:Neutrophil)
theta_tbm26_long$group <- rep('TBM (HIV+)',each=dim(theta_tbm26_long)[1])
theta_tbm27_long <- pivot_longer(theta_tbm27,CD8T:Neutrophil)
theta_tbm27_long$group <- rep('TBM (HIV-)',each=dim(theta_tbm27_long)[1])
theta_healthy_long <- pivot_longer(theta_healthy,CD8T:Neutrophil)
theta_healthy_long$group <- rep('Healthy',each=dim(theta_healthy_long)[1])
theta_long<-rbind(theta_ptb_long,theta_tbm27_long,theta_tbm26_long,theta_healthy_long)

# Get z estimation for each cell type for each group
z_ptb <- get.exp(bp=bp.res_ptb, state.or.type="type")
z_healthy <- get_Z_array(bp.res_healthy, resolution = 'ct', n.core = 7)
z_tbm <- get_Z_array(bp.res_tbm, resolution = 'ct', n.core = 7)
z_tbm26 <- z_tbm[id26TB$LIMS_ID,,]
z_tbm27 <- z_tbm[id27TB$LIMS_ID,,]

#z_healthy <- get.exp(bp=bp.res_healthy, state.or.type="type")
#z_tbm <- get.exp(bp=bp.res_tbm, state.or.type="type")
#z_tbm26 <- z_tbm[id26TB$LIMS_ID,,]
#z_tbm27 <- z_tbm[id27TB$LIMS_ID,,]

#vst transform
for (c in colnames(z_healthy[1,,])) {
  z_healthy[,,c] <-vst(round(t(z_healthy[,,c])),
                    blind=T,
                    nsub=dim(z_healthy[,,c])[1],
                    fitType='parametric') %>% t()
  
  z_ptb[,,c] <-vst(round(t(z_ptb[,,c])),
                  blind=T,
                  nsub=dim(z_ptb[,,c])[1],
                  fitType='parametric') %>% t()
  z_tbm[,,c] <-vst(round(t(z_tbm[,,c])),
                  blind=T,
                  nsub=dim(z_tbm[,,c])[1],
                  fitType='parametric') %>% t()
  z_tbm26[,,c] <-vst(round(t(z_tbm26[,,c])),
                  blind=T,
                  nsub=dim(z_tbm26[,,c])[1],
                  fitType='parametric') %>% t()
  z_tbm27[,,c] <-vst(round(t(z_tbm27[,,c])),
                blind=T,
                nsub=dim(z_tbm27[,,c])[1],
                fitType='parametric') %>% t()
}

gt <- data.frame(LIMS_ID = clinical_data_HC_CNS_IF_PTB_TBM$LIMS_ID,
           Neutrophil = clinical_data_HC_CNS_IF_PTB_TBM$NEUTLE,
           Lymphocyte = clinical_data_HC_CNS_IF_PTB_TBM$LYMLE,
           Monocyte =  clinical_data_HC_CNS_IF_PTB_TBM$MONOLE
           )
gt <- na.omit(gt)
rownames(gt) <- gt$LIMS_ID
gt$LIMS_ID <- NULL

```

#### 3.5.1.0. Cell Types Proportion

```{r}
library(rstatix)
library(ggpubr)

stat.test <- theta_long %>% 
  group_by(name) %>%
  wilcox_test(value ~ group) %>%
  add_significance("p") %>%
  #filter(p.signif !="ns") %>%
  add_y_position(step.increase = 0.06)

p <- ggplot(data=theta_long,
       aes(x = factor(group, level = c('Healthy','PTB','TBM (HIV+)','TBM (HIV-)')),
           color=factor(group, level = c('Healthy','PTB','TBM (HIV+)','TBM (HIV-)')),
           y=value,))+
  facet_wrap(~factor(name, level=c('B','CD4T','CD8T','NK','Monocyte','Neutrophil')), 
             ncol = 6)+
  geom_boxplot() +
  geom_point(position=position_jitterdodge(jitter.width =0.3),alpha=0.1)+
  scale_color_manual(values=c("#A3BE8C","#81A1C1","#B48EAD","#D08770"))+
  labs(color='Group')+ 
  xlab('Cell type')+
  ylab('% Cell population')+
  theme_bw()+
  theme(plot.title = element_text(size = 25),
        legend.position = "right",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_blank(),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold"),
        strip.text = element_text(face="bold",size=14))+ 
  stat_pvalue_manual(stat.test,
                     label = 'p.signif',
                     tip.length = 0,vjust=0.2)

#png(file=paste0(path_output,"/BayesPrism_CellPops.png"),res=900, units = 'in', width=17, height = 7)
p
#dev.off()
```

#### 3.5.1.1. Cell counts

```{r}
WBC_df <- clinical_data_HC_CNS_IF_PTB_TBM[,c('LIMS_ID','WBC')]
row.names(WBC_df) <- WBC_df$LIMS_ID

wbc <- WBC_df[rownames(theta_ptb),]$WBC
theta_ptb_count <- theta_ptb*wbc
wbc <- WBC_df[rownames(theta_tbm),]$WBC
theta_tbm_count <- theta_tbm*wbc
wbc <- WBC_df[rownames(theta_tbm26),]$WBC
theta_tbm26_count <- theta_tbm26*wbc
wbc <- WBC_df[rownames(theta_tbm27),]$WBC
theta_tbm27_count <- theta_tbm27*wbc
wbc <- WBC_df[rownames(theta_healthy),]$WBC
theta_healthy_count <- theta_tbm*wbc

# 
# 
# tmp <- merge(theta1,WBC_df,by=0)
# theta1_count <- data.frame(B = tmp$B*tmp$WBC,
#                            CD4T = tmp$CD4T*tmp$WBC,
#                            CD8T = tmp$CD8T*tmp$WBC,
#                            NK = tmp$NK*tmp$WBC,
#                            Monocyte = tmp$Monocyte*tmp$WBC,
#                            Neutrophil = tmp$Neutrophil*tmp$WBC)
# 
# tmp <- merge(theta2_1,WBC_df,by=0)
# theta2_1_count <- data.frame(B = tmp$B*tmp$WBC,
#                            CD4T = tmp$CD4T*tmp$WBC,
#                            CD8T = tmp$CD8T*tmp$WBC,
#                            NK = tmp$NK*tmp$WBC,
#                            Monocyte = tmp$Monocyte*tmp$WBC,
#                            Neutrophil = tmp$Neutrophil*tmp$WBC)
# 
# tmp <- merge(theta2_2,WBC_df,by=0)
# theta2_2_count <- data.frame(B = tmp$B*tmp$WBC,
#                            CD4T = tmp$CD4T*tmp$WBC,
#                            CD8T = tmp$CD8T*tmp$WBC,
#                            NK = tmp$NK*tmp$WBC,
#                            Monocyte = tmp$Monocyte*tmp$WBC,
#                            Neutrophil = tmp$Neutrophil*tmp$WBC)
# tmp <- merge(theta3,WBC_df,by=0)
# theta3_count <- data.frame(B = tmp$B*tmp$WBC,
#                            CD4T = tmp$CD4T*tmp$WBC,
#                            CD8T = tmp$CD8T*tmp$WBC,
#                            NK = tmp$NK*tmp$WBC,
#                            Monocyte = tmp$Monocyte*tmp$WBC,
#                            Neutrophil = tmp$Neutrophil*tmp$WBC)

tmp1 <- pivot_longer(theta_ptb_count,CD8T:Neutrophil)
tmp1$value <- log10(tmp1$value)
tmp1$group <- rep('PTB',each=dim(tmp1)[1])
tmp2_1<- pivot_longer(theta_tbm26_count,CD8T:Neutrophil)
tmp2_1$value <- log10(tmp2_1$value)
tmp2_1$group <- rep('TBM (HIV+)',each=dim(tmp2_1)[1])
tmp2_2<- pivot_longer(theta_tbm27_count,CD8T:Neutrophil)
tmp2_2$value <- log10(tmp2_2$value)
tmp2_2$group <- rep('TBM (HIV-)',each=dim(tmp2_2)[1])
tmp3 <- pivot_longer(theta_healthy_count,CD8T:Neutrophil)
tmp3$value <- log10(tmp3$value)
tmp3$group <- rep('Healthy',each=dim(tmp3)[1])
tmp<-rbind(tmp1,tmp2_1,tmp2_2,tmp3)

stat.test <- tmp %>% 
  group_by(name) %>%
  wilcox_test(value ~ group) %>%
  add_significance("p") %>%
  #filter(p.signif !="ns") %>%
  add_y_position(step.increase = 0.2)

p <- ggplot(data=tmp,
       aes(x = factor(group, level = c('Healthy','PTB','TBM (HIV-)','TBM (HIV+)')),
           color=factor(group, level = c('Healthy','PTB','TBM (HIV-)','TBM (HIV+)')),
           y=value,))+
  facet_wrap(~factor(name, 
                     level=c('B','CD4T','CD8T','NK','Monocyte','Neutrophil')), scale='free', 
             ncol = 6)+
  geom_boxplot() +
  geom_point(position=position_jitterdodge(jitter.width =0.3),alpha=0.1)+
  scale_color_manual(values=c("#A3BE8C","#81A1C1","#B48EAD","#D08770"))+
  labs(color='Group')+ 
  xlab('Cell type')+
  ylab('Log10 cell count')+
  theme_bw()+
  theme(plot.title = element_text(size = 25),
        legend.position = "right",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_blank(),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold"),
        strip.text = element_text(face="bold",size=14)) +
  stat_pvalue_manual(stat.test, 
                     label = 'p.signif',
                     tip.length = 0,vjust=0.5)

png(file=paste0(path_output,"/BayesPrism_CellCounts.png"),res=900, units = 'in', width=17, height = 7)
p
dev.off()
```

#### 3.5.2. CNIH4 Expression Estimation

```{r}
get_z_vst <- function(bp,gene) {
  z <- get.exp(bp=bp, state.or.type="type")
  tmp <- vst(round(t(z[,,'B'])),
             blind=T,
             nsub=dim(z[,,'B'])[1],
             fitType='parametric') %>% t() %>% as.data.frame()
  df <- data.frame(B = tmp[,gene])
  row.names(df) <- row.names(tmp)
  cell_list <- c('B','CD4T','CD8T','NK','Monocyte','Neutrophil')
  for(i in 1:length(cell_list)){
    c <- cell_list[i]
    print(c)
    tmp <- round(z[,,c])
    tmp <- vst(round(t(z[,,c])),blind=T,nsub=dim(z[,,c])[1]) %>% t() %>% as.data.frame()
    df[,c] <- tmp[,gene]
  }
  return(df)
}

zptb_cnih4 <- get_z_vst(bp.res_ptb,'CNIH4')
ztbm_cnih4 <- get_z_vst(bp.res_tbm,'CNIH4')
ztbm26_cnih4 <- ztbm_cnih4[id26TB$LIMS_ID,]
ztbm27_cnih4 <- ztbm_cnih4[id27TB$LIMS_ID,]
zhealthy_cnih4 <- get_z_vst(bp.res_healthy,'CNIH4')

which(ztbm26_cnih4<0,arr.ind = TRUE)
```

```{r}

tmp1 <- pivot_longer(zptb_cnih4,B:Neutrophil)
tmp1$group <- rep('PTB',each=dim(tmp1)[1])
tmp2<- pivot_longer(ztbm_cnih4,B:Neutrophil)
tmp2$group <- rep('TBM',each=dim(tmp2)[1])

tmp2_1<- pivot_longer(ztbm26_cnih4,B:Neutrophil)
tmp2_1$group <- rep('TBM (HIV+)',each=dim(tmp2_1)[1])
tmp2_2<- pivot_longer(ztbm27_cnih4,B:Neutrophil)
tmp2_2$group <- rep('TBM (HIV-)',each=dim(tmp2_2)[1])

tmp3 <- pivot_longer(zhealthy_cnih4,B:Neutrophil)
tmp3$group <- rep('Healthy',each=dim(tmp3)[1])
tmp<-rbind(tmp1,
           tmp2_1,
           tmp2_2,
           tmp3)

stat.test <- tmp %>% 
  group_by(name) %>%
  wilcox_test(value ~ group) %>%
  add_significance("p") %>%
  #filter(p.signif !="ns") %>%
  add_y_position(step.increase = 0.04)

p <- ggplot(data=tmp,
       aes(x = factor(group, level = c('Healthy',
                                         'PTB',
                                         'TBM (HIV+)',
                                         'TBM (HIV-)')),
           y=value,
           color=factor(group, level = c('Healthy',
                                         'PTB',
                                         'TBM (HIV+)',
                                         'TBM (HIV-)'
                                         ))))+
  facet_wrap(~factor(name, level=c('B','CD4T','CD8T','NK','Monocyte','Neutrophil')), 
             ncol = 6)+
  geom_boxplot() +
  geom_point(position=position_jitterdodge(jitter.width =0.3),
             alpha=0.1)+
  scale_color_manual(values=c("#A3BE8C",
                              "#81A1C1",
                              "#B48EAD",
                              "#D08770"
                              ))+
  labs(color='Group')+ 
  xlab('Cell type')+
  ylab('CNIH4 estimation (after VST)')+
  theme_bw()+
  theme(plot.title = element_text(size = 25),
        legend.position = "right",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_blank(),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold"),
        strip.text = element_text(face="bold",size=14)) +
  stat_pvalue_manual(stat.test, 
                     label = 'p.signif',
                     tip.length = 0,vjust=1)

png(file=paste0(path_output,"/BayesPrism_zCNIH4.png"),res=900, units = 'in', width=17, height = 7)
p
dev.off()

```

#### 3.5.3. PD-L1 Expression Estimation. Wait

```{r}
zptb_pdl1 <- get_z_vst(bp.res_ptb,'CD274')
tmp1 <- pivot_longer(zptb_pdl1,B:Neutrophil)

p <- ggplot(data=tmp1,
       aes(x = factor(name, level=c('B','CD4T','CD8T','NK','Monocyte','Neutrophil')),
           y=value,
           color=name))+
         geom_boxplot()+
geom_point(position=position_jitterdodge(jitter.width =0.3),
             alpha=0.1)+
  scale_color_manual(values=c("#A3BE8C",
                              "#81A1C1",
                              "#B48EAD",
                              "#BF616A","#D08770","#88C0D0"
                              ))+
  labs(color='Group')+ 
  xlab('Cell type')+
  ylab('PD-L1 estimation (after VST)')+
  theme_bw()+
  theme(plot.title = element_text(size = 25),
        legend.position = "none",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_text(color="black",size=13,face="bold",angle = 45,vjust = 1, hjust=1),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold"),
        strip.text = element_text(face="bold",size=14))
p
```

```{r}
z <- get.exp(bp=bp.res_ptb, state.or.type="type")
tmp <- vst(round(t(z[,,'Neutrophil'])),
           blind=T,
           nsub=dim(z[,,'Neutrophil'])[1],
           fitType='mean') %>% t() %>% as.data.frame()
df <- data.frame(Neutrophil = tmp[,'CD274'])
```

```{r}
zptb_cd274 <- get_z_vst(bp.res_ptb,'CD274')
ztbm_cd274 <- get_z_vst(bp.res_tbm,'CD274')
ztbm26_cnih4 <- ztbm_cnih4[id26TB$LIMS_ID,]
ztbm27_cnih4 <- ztbm_cnih4[id27TB$LIMS_ID,]
zhealthy_cd274 <- get_z_vst(bp.res_healthy,'CD274')
```

```{r}

tmp1 <- pivot_longer(zptb_cd274,Neutrophil:NK)
tmp1$group <- rep('PTB',each=dim(tmp1)[1])
tmp2<- pivot_longer(ztbm_cnih4,Neutrophil:NK)
tmp2$group <- rep('TBM',each=dim(tmp2)[1])
tmp3 <- pivot_longer(zh_cnih4,Neutrophil:NK)
tmp3$group <- rep('Healthy',each=dim(tmp3)[1])
tmp<-rbind(tmp1,tmp2,tmp3)

stat.test <- tmp %>%
  group_by(name) %>% 
  t_test(value ~ group) %>% 
  add_significance("p") %>%
  add_xy_position(x = "name")

p <- ggplot(data=tmp,
       aes(x=name,
           y=value,
           color=factor(group, level = c('Healthy','PTB','TBM'))))+
  geom_boxplot() +
  geom_point(position=position_jitterdodge(jitter.width =0.3),
             alpha=0.1)+
  scale_color_manual(values=c("#81A1C1","#B48EAD","#A3BE8C"))+
  labs(color='Group')+ 
  xlab('Cell type')+
  ylab('CNIH4 Z Estimation')+
  theme_bw()+
  theme(plot.title = element_text(size = 25),
        legend.position = "right",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_text(color="black",size=13,face="bold",angle = 45,vjust = 1, hjust=1),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold"))+
  stat_pvalue_manual(stat.test, label = 'p.signif', tip.length = 0)
```

#### 3.5.4. Heatmap genes vs samples

```{r}
z <- get.exp(bp=bp.res_ptb, state.or.type="type",cell.name = 'Neutrophil')
tmp <- vst(round(t(z[,])),
             blind=T,
             nsub=dim(z[,])[1],
             fitType='parametric') %>% t() %>% as.data.frame()
df <- data.frame(tmp)
df
library(NatParksPalettes)
natparks.pals(name="DeathValley",n=7,type="continuous")
```

#### 3.5.5. ssGSEA for group + celltype

```{r}
library(GSVA)
library(BiocParallel)
library(ReactomeContentService4R)
library(msigdbr)
library(ggpmisc)

get_genelist_pathway_kegg <- function(hsa_ID){
  library("KEGGREST")
  names <- keggGet(hsa_ID)[[1]]$GENE
  namesodd <-  names[seq(0,length(names),2)]
  namestrue <- gsub("\\;.*","",namesodd)
}

#T cell receptor signaling pathway - Homo sapiens (human)
TCR_signaling<-get_genelist_pathway_kegg(hsa_ID = "hsa04660")

get_ssgsea <- function(bp.res){
  num_samples <- dim(bp.res@prism@mixture)[1]
  TCR_signaling_df <- data.frame()
  TCR_signaling_df[1:num_samples,]=NA
  for (i in 1:6){
  cellname = bp.res@prism@map[[i]] #"Neutrophil"
  z <- get.exp(bp=bp.res,
               state.or.type="type",
               cell.name = cellname
               )
  
  z <- vst(round(t(z)),
           blind=T, 
           nsub=dim(z)[1],
           fitType='parametric') %>% t() %>% as.data.frame()
  
  
  GSVA_targeted <- gsva(t(z), gset.idx.list=list(TCR_signaling=TCR_signaling),method="ssgsea",mx.diff=T)
  GSVA_targeted<-as.data.frame(t(GSVA_targeted))
  TCR_signaling_df[,cellname] = GSVA_targeted$TCR_signaling
  }
  rownames(TCR_signaling_df) <- rownames(GSVA_targeted)
  TCR_signaling_df <- TCR_signaling_df[,c('B','CD4T','CD8T','NK','Neutrophil','Monocyte')]
  return(TCR_signaling_df)
}

TCRsig_df_ptb <- get_ssgsea(bp.res = bp.res_ptb)
TCRsig_df_healthy <- get_ssgsea(bp.res = bp.res_healthy)
TCRsig_df_tbm <- get_ssgsea(bp.res = bp.res_tbm)
TCRsig_df_tbm26 <- TCRsig_df_tbm[id26TB$LIMS_ID,]
TCRsig_df_tbm27 <- TCRsig_df_tbm[id27TB$LIMS_ID,]


TCRsig_df_long_ptb <- pivot_longer(TCRsig_df_ptb,B:Monocyte)
TCRsig_df_long_healthy <- pivot_longer(TCRsig_df_healthy,B:Monocyte)
TCRsig_df_long_tbm <- pivot_longer(TCRsig_df_tbm,B:Monocyte)
TCRsig_df_long_tbm26 <- pivot_longer(TCRsig_df_tbm26,B:Monocyte)
TCRsig_df_long_tbm27 <- pivot_longer(TCRsig_df_tbm27,B:Monocyte)

TCRsig_df_long_ptb$group <- rep('PTB',each=dim(TCRsig_df_long_ptb)[1])
TCRsig_df_long_healthy$group <- rep('Healthy',each=dim(TCRsig_df_long_healthy)[1])
TCRsig_df_long_tbm$group <- rep('TBM',each=dim(TCRsig_df_long_tbm)[1])
TCRsig_df_long_tbm26$group <- rep('TBM (HIV+)',each=dim(TCRsig_df_long_tbm26)[1])
TCRsig_df_long_tbm27$group <- rep('TBM (HIV-)',each=dim(TCRsig_df_long_tbm27)[1])


tmp<-rbind(TCRsig_df_long_healthy,
           TCRsig_df_long_ptb,
           TCRsig_df_long_tbm26,
           TCRsig_df_long_tbm27)

stat.test <- tmp %>% 
  group_by(name) %>%
  wilcox_test(value ~ group) %>%
  add_significance("p") %>%
  #filter(p.signif !="ns") %>%
  add_y_position(step.increase = 0.06)

p <- ggplot(data=tmp,
       aes(x = factor(group, level = c('Healthy','PTB','TBM (HIV+)','TBM (HIV-)')),
           color=factor(group, level = c('Healthy','PTB','TBM (HIV+)','TBM (HIV-)')),
           y=value,))+
  facet_wrap(~factor(name, level=c('B','CD4T','CD8T','NK','Monocyte','Neutrophil')), 
             ncol = 6)+
  geom_boxplot() +
  geom_point(position=position_jitterdodge(jitter.width =0.3),alpha=0.1)+
  scale_color_manual(values=c("#A3BE8C","#81A1C1","#B48EAD","#D08770"))+
  labs(color='Group')+ 
  xlab('Cell type')+
  ylab('Enrichment score of T-cell receptor signaling pathways')+
  theme_bw()+
  theme(plot.title = element_text(size = 25),
        legend.position = "right",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_blank(),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold"),
        strip.text = element_text(face="bold",size=14)) +
  stat_pvalue_manual(stat.test, 
                     label = 'p.signif',
                     tip.length = 0,vjust=0.8)
png(file=paste0(path_output,"/BayesPrism_ssGSEA_TCRsignal.png"),res=900, units = 'in', width=17, height =10)
p
dev.off()
```

#### 3.5.6. Regression plot

```{r}

library(ggpmisc)
library(ggpubr)
library(cowplot)
library(ggplot2)
library(Metrics)

theta <- theta_healthy #input theta_ptb, theta_tbm, theta_healthy

theta1 <- theta[row.names(gt),]
theta1 <- na.omit(theta1)
gt1 <- gt[row.names(theta1),]

cell_list = c('Lymphocyte','Monocyte',"Neutrophil")
p <- list()

for (i in 1:length(cell_list)) {
  c <- cell_list[i]
  
  df <- data.frame(predicted=theta1[,c]*100,groundtruth=gt1[,c])
  rmse <- rmse(theta1[,c]*100,gt1[,c])
  p1<-ggplot(data=df, aes(x=groundtruth,y=predicted))+
      geom_point(alpha=0.8,col='red')+
      theme_bw()+
      stat_poly_line(se=T) +
      stat_cor(cor.coef.name = "r",size = 7,method = 'pearson')+
      theme_bw()+
      ylab(paste0(c," predicted"))+
      xlab(paste0(c," groundtruth"))+
    ggtitle(paste0("PTB, ","RMSE = ",round(rmse,2)))+
      theme(plot.title = element_text(size = 15),
            legend.position = "none",
            axis.text = element_text(size = 15),
            axis.text.x=element_text(color="black",size=13,face="bold"),
            axis.title.x = element_text(size = 15,color="black",face="bold"),
            axis.text.y=element_text(color="black",size=15),
            axis.title.y = element_text(size = 15,color="black",face="bold"),
            legend.text = element_text(size = 15),
            legend.title = element_text( colour="black", size=12),
            legend.key = element_rect(colour = NA),
            strip.text=element_text(face = "bold",size=12))
  p[[i]] <-ggdraw(p1)
}

regs <- ggpubr::ggarrange(
p[[1]],
p[[2]],
p[[3]],

ncol =3,nrow=1)

#png(file="scatterplot_BayesPrism.png",res=600, units = 'in', width=45, height =30) #30:18
regs
#dev.off()

```

### 3.6. DESEQ2 Analysis

```{r}
library(EnhancedVolcano)
library(comprehenr)

df <- filter(vst_ptb_tbm,Type=='PTB') %>% select(c('LIMS_ID','CNIH4level'))
PTB_sample_RNAseq <- All_sample_RNAseq %>% subset(Study %in% c("28TB","29TB"))
count_data <-count_all[,rownames(PTB_sample_RNAseq)]
count_data<-count_data[ , !(names(count_data) %in% c('HOA7818A92'))] # drop 1 patients not in vst

df<- df[order(df$LIMS_ID),] #col
df$CNIH4level<-factor(df$CNIH4level,levels=c(1,2))
df$LIMS_ID<-NULL

count_data<-count_data[ , order(names(count_data))]
anno_symbol <- anno[,c("gene_id", "gene_symbol")] %>% filter(gene_id %in% rownames(count_data))
rownames(anno_symbol) <- anno_symbol$gene_id
anno_symbol <- anno_symbol[rownames(count_data),]
count_data <- cbind(genenames=anno_symbol$gene_symbol,count_data)
count_data <-count_data %>% filter(genenames %in% names(RNA_data) & genenames!='CNIH4')

count_data$genenames<-NULL
anno_symbol <- anno[,c("gene_id", "gene_symbol")] %>% filter(gene_id %in% rownames(count_data))

dup_genes_id <- anno_symbol[duplicated(anno_symbol$gene_symbol), ]$gene_id
count_data <- count_data[!(rownames(count_data) %in% dup_genes_id),]

dds <- DESeqDataSetFromMatrix(count_data, colData=df, ~CNIH4level)
dds <- DESeq(dds)

#results(dds, contrast=c("condition","C","B")) meaning genes with:
#logFC > 0 are overexpressed in C and underexpressed in B.
#logFC < 0 means underexpressed in C and overexpressed in B

res1 <- results(dds,contrast=c('CNIH4level',2,1),alpha=0.05)
#res2 <- lfcShrink(dds, coef= 'CNIH4level_2_vs_1', type="apeglm")
summary(res1)

labels <- mapvalues(rownames(res1), 
          from=anno_symbol$gene_id, 
          to=anno_symbol$gene_symbol)

EnhancedVolcano(res1,
    lab = labels,
    x = 'log2FoldChange',labSize = 2.0,
    pCutoff = 0.05,
    FCcutoff = log2(1.5),
    y = 'padj',)

get_DEGgenes <- function(df,reg){
    if(reg=='high'){
      key <- intersect(rownames(df)[which(df$log2FoldChange>=log2(1.5))], rownames(df)[which(df$padj<0.05)])
    }
    if(reg=='low'){
      key <- intersect(rownames(df)[which(df$log2FoldChange<=-log2(1.5))], rownames(df)[which(df$padj<0.05)])
    }
    results <- as.data.frame((df)[which(rownames(df) %in% key),])
    return(results)
}

deg_high <- rownames(get_DEGgenes(as.data.frame(res1),reg='high'))
deg_low <- rownames(get_DEGgenes(as.data.frame(res1),reg='low'))

deg_high <- to_vec(for(i in 1:length(deg_high)) filter(anno_symbol,gene_id==deg_high[i])$gene_symbol)
deg_low <- to_vec(for(i in 1:length(deg_low)) filter(anno_symbol,gene_id==deg_low[i])$gene_symbol)
```

### 3.7. Pathway Analysis

```{r}
ORA_bar_plot <- function(df,name,n=5){
  n <- min(nrow(df),n)
  df <- df[1:n,]
  #n_break <- as.integer(max(df$log_pvalue))
  p <- ggplot(aes(x= ID,y = log_pvalue),data =df) + 
  geom_col(position = "stack", 
           alpha = 0.5,fill="#4292c6") +
  geom_text(aes(x=ID,y=0.1,label =ID), 
            size=3,fontface='bold',hjust=0)+
  ggtitle(name)+
  #geom_text(aes(x=ID,y=pos,label =log_pvalue), hjust=0.5) +
  #scale_y_continuous(breaks = 1:n_break)+ 
  coord_flip()+
  labs(x="", y="Log10 P value") +
  #scale_fill_manual(values=c("#4292c6")) +
  theme_bw()+
  theme(plot.title = element_text(size = 20),
        legend.text = element_text(size = 15) ,
        legend.position="right",
        legend.title = element_text(size = 13,color="black",face="bold"),
        axis.text.x=element_text(color="black",size=13),
        axis.title.x = element_text(size = 13,color="black",face="bold"),
        axis.text.y=element_blank(),
        axis.ticks.y = element_blank(),
        axis.title.y = element_text(size = 13,color="black",face="bold"))
  
  #p
  return(p)
}
```

```{r}
library(msigdbr)
library(dplyr)
library(clusterProfiler)
# db <- msigdbr(species = "Homo sapiens", category = "H") %>%
#   dplyr::select(gs_name, human_gene_symbol)
#C5, C8

# db <- msigdbr(species = "Homo sapiens", category = "C5") %>%
#   filter(gs_subcat=="GO:BP") %>% #C5
#   dplyr::select(gs_name, human_gene_symbol)

db <- msigdbr(species = "Homo sapiens", category = "C8") %>%
  dplyr::select(gs_name, human_gene_symbol)

dim(db)

ORA_analysis <- function(gene_list){
  temp <- enricher(gene_list, TERM2GENE=db, pvalueCutoff = 1, qvalueCutoff = 1)
  temp <- temp@result
  if(!is.na(table(temp$p.adjust < padj)["TRUE"]) &  table(temp$p.adjust < padj)["TRUE"]>10){
    temp <- temp %>% filter(p.adjust<padj) %>%
    dplyr::arrange(pvalue)
    } 
  else{
    temp <- temp %>%
    dplyr::arrange(pvalue)
    temp <- temp[1:min(nrow(temp),10),]
    }
  # df <- temp %>%
  # dplyr::arrange(pvalue) %>%
  # dplyr::mutate(log_pvalue=-log10(pvalue),
  #               pos = ifelse(log_pvalue <4,log_pvalue + 1,log_pvalue + 0.5))
  # ORA_bar_plot(df)
  return(temp)
}

ora_low <- ORA_analysis(deg_low)
df1 <- ora_low %>%
  dplyr::arrange(pvalue) %>%
  dplyr::mutate(log_pvalue=-log10(pvalue),
                pos = ifelse(log_pvalue <4,log_pvalue + 1,log_pvalue + 0.5))
p1<-ORA_bar_plot(df1,'Genes down-regulated in high CNIH4')
p1
ora_high <- ORA_analysis(deg_high)
df2 <- ora_high %>%
  dplyr::arrange(pvalue) %>%
  dplyr::mutate(log_pvalue=-log10(pvalue),
                pos = ifelse(log_pvalue <4,log_pvalue + 1,log_pvalue + 0.5))
p2<-ORA_bar_plot(df1,'Genes up-regulated in high CNIH4')


n = 5
df1 <- df1[1:min(nrow(df1),n),] %>% mutate(group='downregulated genes') %>% arrange(ID)
df2 <- df2[1:min(nrow(df2),n),] %>% mutate(group='upregulated genes') %>% arrange(ID)
df <- rbind(df1,df2)
df$ID <- factor(df$ID, levels = df$ID)

#n_break <- as.integer(max(df$log_pvalue))
p <- ggplot(data =df,aes(x= ID,y = log_pvalue,fill=group)) + 
  geom_col(position = 'stack',
         alpha = 0.5,
         aes(x=ID,y = log_pvalue,
             fill = group)) +
scale_fill_manual( values = c('#5E81AC','#BF616A'))+
geom_text(aes(x=ID,y=0.1,label =ID), 
          size=3,fontface='bold',hjust=0)+
#geom_text(aes(x=ID,y=pos,label =log_pvalue), hjust=0.5) +
#scale_y_continuous(breaks = 1:n_break)+ 
coord_flip()+
labs(x="", y="Log10 P value") +
#scale_fill_manual(values=c("#4292c6")) +
theme_bw()+
theme(plot.title = element_text(size = 20),
      legend.text = element_text(size = 15) ,
      legend.position="right",
      legend.title = element_text(size = 13,color="black",face="bold"),
      axis.text.x=element_text(color="black",size=13),
      axis.title.x = element_text(size = 13,color="black",face="bold"),
      axis.text.y=element_blank(),
      axis.ticks.y = element_blank(),
      axis.title.y = element_text(size = 13,color="black",face="bold"))
p
```

### 3.8. Salmon Isoforms Correlation Analysis

```{r}
write.csv(salmons12,'ptb_CNIH4_CD274_salmon_vst.csv',row.names = TRUE)
```

```{r}
salmons <- read.csv(paste0(path_input,'/Data/ptb_CNIH4_CD274_salmon_counts.csv'))

salmons12 <- salmons %>% filter(Run=='1+2') %>% tibble::column_to_rownames('transcript_name') %>% select(colnames(salmons)[8:dim(salmons)[2]]) %>% t()

salmons12 <- as.data.frame(vst(round(salmons12),blind=T, nsub=295))
salmons12 <- merge(salmons12,vst_data,by=0) %>% select(c(colnames(salmons12),'CT_mean'))

```

```{r}
library(grid)
a <- ifelse(colnames(salmons12) %in% c('CNIH4-201','CNIH4-202','CNIH4-203','CNIH4-205', 'CD274-201','CD274-202'), "red", "black")
p01$gtable$grobs[[5]]$gp=gpar(col=a)
p01
```

```{r}
cor_method = 'pearson'
mat <- cor(salmons12,
                      use='pairwise.complete.obs',
                      method = cor_method)

p.mat <- cor.pval.test(salmons12, 
                                  salmons12, 
                                  method=cor_method, 
                                  use='pairwise.complete.obs')

#col=c("#001164", 'white', "#FF681E")
#col=c("#5E81AC", '#EBCB8B', "#BF616A")
col=c("#395b85", 'white', "#FF681E")
# CNIH4 vs 25 ICs
mat1 <- mat
p.mat1 <- p.mat
labels1 <-round(mat1,2)
vals1 <- labels1
vals1[p.mat1 > 0.05] <- 0
labels1[p.mat1 > 0.05] <- ''

myBreaks <- c(seq(-1, 0, length.out=ceiling(1000/2) + 1),
              seq(1/1000, 1, length.out=floor(1000/2)))

p01 <- pheatmap(vals1,
               color = colorRampPalette(col)(1000),
               breaks=myBreaks,
               display_numbers=labels1,
               show_colnames = F,legend = F)
p01
row.order <- p01$tree_row$order
col.order <- p01$tree_col$order
p1 <- as.ggplot(p01)
```

### 3.9. Association of CNIH4 with PD1/PD-L1 pathway (ssGSEA)

```{r}

```

### 3.10. Association of CNIH4 with IFN pathway from GO/Reactome database (ssGSEA)

### 3.11. Association of CNIH4 with pro-inflamatory and anti-inflamatory cytokines

### 3.12. AUC of TB vs Healthy using CNIH4, PD-L1, PD1

```{r}
library(pacman)
pacman::p_load(
  rio,         # importing data
  here,        # relative file pathways
  janitor,     # data cleaning and tables
  lubridate,   # working with dates
  tidyverse,   # data management and visualization
  dplyr,       # data manipulation
  naniar,      # package for "Replace with NA"
  purrr,       # iteration
  stringr,     # many functions for handling strings
  pROC,        # ROC curve and AUC
  ROCR,        # Visualizing the Performance of Scoring Classifiers
  mgcv,        # Gam model with cross validation
  splines,     # Regression Spline Functions and Classes
  rms,
  data.table  # XGBoost preparation data
)
```

```{r}
# genename <- 'CNIH4'
# 
# ptb_ct40 <- vst_data %>% select(c('LIMS_ID',genename,'CT_mean')) %>% filter(CT_mean==40)
# tbm_ct40 <- Data_26_27_CT %>% mutate(CT_mean=NAT) %>% select(c('LIMS_ID',genename,'CT_mean')) %>% filter(CT_mean==40)
# row.names(tbm_ct40) <- tbm_ct40$LIMS_ID
# ptb_tbm_ct40 <- rbind(ptb_ct40,tbm_ct40)
# healthy <- healthy_df %>% select(c('LIMS_ID',genename))
# 
# 

genename <- 'CNIH4'
ptb_ct40 <- vst_data %>% select(c('LIMS_ID',genename,'CT_mean')) %>% filter(CT_mean==40)
tbm_ct40 <- merge(Data_26_27_CT[,c('LIMS_ID','NAT')],TBM_df,by='LIMS_ID') %>% mutate(CT_mean=NAT) %>% select(c('LIMS_ID',genename,'CT_mean')) %>% filter(CT_mean==40)
row.names(tbm_ct40) <- tbm_ct40$LIMS_ID
ptb_tbm_ct40 <- rbind(ptb_ct40,tbm_ct40)
healthy <- healthy_df %>% select(c('LIMS_ID',genename))

#write.csv(ptb_tbm_ct40,'ptb_tbm_ct40_CNIH4.csv',row.names = FALSE)
#write.csv(healthy,'healthy_CNIH4.csv',row.names = FALSE)

data_TB <- ptb_tbm_ct40 %>% 
  dplyr::mutate(
    phenotype = 1
  ) %>% 
  dplyr::select(-c(CT_mean))

data_healthy <- healthy %>% 
  dplyr::mutate(
    phenotype = 0
  )

data_combine <- rbind(data_TB, data_healthy)
```

```{r}
par(pty = "s")

roc <- plot.roc(data_combine$phenotype, data_combine[,genename]) 

ci.sp <- ci.sp(roc, sensitivities=seq(0, 1, .01), boot.n=100)

p <- data_combine %>% 
  roc(phenotype, 
      predictor = c(get(genename)), 
      plot = TRUE, 
      legacy.axes = TRUE,
    percent = FALSE, xlab = "1 - Specificity", 
    ylab = "Sensitivity", col = "#377eb8", lwd = 5,
    print.auc = TRUE, print.auc.x = 0.55, ci = TRUE)

plot(ci.sp, type="shape", col="#377eb822")
title(main = 'PD1')
```

```{r}
dd <-  datadist (data_combine); options ( datadist = "dd")

fit <- lrm (phenotype ~ CNIH4 , data = data_combine, x = TRUE, y = TRUE )

set.seed(12345)
model_valid <- validate(fit, B = 10000)
model_valid
#  index.orig training    test optimism index.corrected     n
# Dxy           0.7102   0.7096  0.7102  -0.0006          0.7109 10000
# R2            0.3432   0.3474  0.3432   0.0042          0.3390 10000
# Intercept     0.0000   0.0000 -0.0310   0.0310         -0.0310 10000
# Slope         1.0000   1.0000  1.0032  -0.0032          1.0032 10000
# Emax          0.0000   0.0000  0.0078   0.0078          0.0078 10000
# D             0.2103   0.2140  0.2103   0.0038          0.2065 10000
# U            -0.0097  -0.0097 -0.0001  -0.0096         -0.0001 10000
# Q             0.2200   0.2237  0.2104   0.0134          0.2066 10000
# B             0.1011   0.0993  0.1025  -0.0032          0.1043 10000
# g             2.9300   3.0243  2.9300   0.0943          2.8357 10000
# gp            0.1734   0.1728  0.1734  -0.0006          0.1741 10000

auc_corrected <- (model_valid[1,5]/2) +0.5
auc_corrected #[1] 0.8554285
```

```{r}
fit_glm <- glm(phenotype ~ CNIH4 , data = data_combine, family=binomial)
prob1 = predict(fit_glm, type='response', data_combine)
pred1 = prediction(prob1, data_combine$phenotype)
performance(pred1,"auc")@y.values[[1]][1]
```

### 3.13. Mediation Analysis

#### 3.13.1. Mtb load --\> AIM2 --\> STAT1

```{r}
library(CMAverse)
library(ReactomeContentService4R) # Extract data from the three pathways
library(GSVA)

set.seed(1)

tmp <- vst_data %>% select(any_of(c('AIM2','CNIH4','STAT1')))
tmp$log2_CT <- log2(vst_data$CT_mean)

res_gformula <- cmest(data = tmp, model = "gformula", 
                      exposure = "log2_CT",mediator = "AIM2", outcome = "STAT1",
                      #basec =  c("Age"), 
                      EMint = FALSE,
                      mreg = list("linear"), yreg = "linear",multimp = TRUE,
                      astar = 0, a = 1, mval = list(1), 
                      estimation = "imputation", inference = "bootstrap", nboot = 3000)

summary(res_gformula)$summarydf

mediated_res<-summary(res_gformula)$summarydf[c("tnde","tnie","te","pm"),] 
mediated_res$Effect<-rownames(mediated_res)

mediated_res %<>%  mutate(Effect=case_when(Effect=="tnde"~"Natural direct impact",
                                           Effect=="tnie"~"Natural indirect impact",
                                           Effect=="te"~"Total impact",
                                           .default ="Proportion Mediated")) %>% 
                    mutate(Effect=factor(Effect,
                                         levels=c("Natural direct impact",
                                                  "Natural indirect impact",
                                                  "Total impact",
                                                  "Proportion Mediated")))
mediated_res
```

```{r}
library(CMAverse)
library(ReactomeContentService4R) # Extract data from the three pathways
library(GSVA)

IFN_alpha_beta<-event2Ids(event.id = "R-HSA-909733.9")
IFN_gamma<-event2Ids(event.id = "R-HSA-877300.7")

#PD-L1 expression and PD-1 checkpoint pathway in cancer - Homo sapiens (human) hsa05235
PD_L1<-get_genelist_pathway_kegg(hsa_ID = "hsa05235") 
#T cell receptor signaling pathway - Homo sapiens (human)
TCR_signaling<-get_genelist_pathway_kegg(hsa_ID = "hsa04660") 

tmp <- vst_data %>% select(any_of(c(TCR_signaling, 'CNIH4','STAT1')))
GSVA_targeted <- gsva(t(tmp), gset.idx.list=TCR_signaling,method="ssgsea",mx.diff=T,BPPARAM=MulticoreParam(12))

set.seed(1)
res_gformula <- cmest(data = tmp,
                      model = "gformula", 
                      outcome = colnames(tmp)[1:117],
                      exposure = "STAT1",
                      mediator = "CNIH4",
                      #basec =  c("Age"), 
                      EMint = FALSE,
                      mreg = list("linear"), yreg = "linear",multimp = TRUE,
                      astar = 0, a = 1, mval = list(1),
                      estimation = "imputation", inference = "bootstrap", nboot = 3000)
```

#### 3.13.2. STAT1 --\> (CNIH4, CD274) --\> T-cell receptor signaling pathways

```{r}
library(GSVA)
library(ReactomeContentService4R)

IFN_alpha_beta<-event2Ids(event.id = "R-HSA-909733.9")
IFN_gamma<-event2Ids(event.id = "R-HSA-877300.7")
PD_L1<-get_genelist_pathway_kegg(hsa_ID = "hsa05235") 
TCR_signaling<-get_genelist_pathway_kegg(hsa_ID = "hsa04660")
targeted_pathways<-list(IFN_alpha_beta=IFN_alpha_beta$geneSymbol,
                        IFN_gamma=IFN_gamma$geneSymbol,
                        PD_L1=PD_L1,
                        TCR_signaling=TCR_signaling)

tmp <- vst_data %>% select(any_of(c(TCR_signaling,
                                    IFN_alpha_beta$geneSymbol,
                                    IFN_gamma$geneSymbol,
                                    PD_L1))) %>% na.omit()

GSVA_targeted <- gsva(t(tmp), gset.idx.list = targeted_pathways, method="ssgsea", mx.diff=T) %>% t()
tmp <- merge(vst_data[,c('Age','STAT1','CNIH4','CD274')],GSVA_targeted,by='row.names')
rownames(tmp) <- tmp$Row.names
tmp$Row.names<-NULL

set.seed(1)
res_gformula <- cmest(data = tmp,
                      model = "gformula", 
                      outcome = 'TCR_signaling',
                      exposure = "STAT1",
                      mediator = c("CNIH4",'CD274'),
                      basec =  c("Age"), 
                      EMint = FALSE,
                      mreg = list("linear","linear"), yreg = "linear",multimp = TRUE,
                      astar = 0, a = 1, mval = list(1,2),
                      estimation = "imputation", inference = "bootstrap", nboot = 3000)

mediated_res<-summary(res_gformula)$summarydf[c("tnde","tnie","te","pm"),] 
mediated_res$Effect<-rownames(mediated_res)

mediated_res %<>%  mutate(Effect=case_when(Effect=="tnde"~"Natural direct impact",
                                           Effect=="tnie"~"Natural indirect impact",
                                           Effect=="te"~"Total impact",
                                           .default ="Proportion Mediated")) %>% 
                    mutate(Effect=factor(Effect,
                                         levels=c("Natural direct impact",
                                                  "Natural indirect impact",
                                                  "Total impact",
                                                  "Proportion Mediated")))
mediated_res
```

```{r}
# Extract data from the three pathways: 
library(ReactomeContentService4R)
library(GSVA)

IFN_alpha_beta<-event2Ids(event.id = "R-HSA-909733.9")
IFN_gamma<-event2Ids(event.id = "R-HSA-877300.7")

#PD-L1 expression and PD-1 checkpoint pathway in cancer - Homo sapiens (human) hsa05235
PD_L1<-get_genelist_pathway_kegg(hsa_ID = "hsa05235") 
#T cell receptor signaling pathway - Homo sapiens (human)
TCR_signaling<-get_genelist_pathway_kegg(hsa_ID = "hsa04660")
targeted_pathways<-list(IFN_alpha_beta=IFN_alpha_beta$geneSymbol,IFN_gamma=IFN_gamma$geneSymbol,PD_L1=PD_L1,TCR_signaling=TCR_signaling)

## Compute ssGSEA of these pathways

GSVA_targeted <- gsva(t(vst_data), gset.idx.list=targeted_pathways,method="ssgsea",mx.diff=T,BPPARAM=MulticoreParam(12))
GSVA_targeted<-as.data.frame(t(GSVA_targeted))
GSVA_targetedcombined<-merge(clinical_RNA,GSVA_targeted,by='row.names')
GSVA_targetedcombined %<>% dplyr::select(Row.names,"Age","log2_Lymphocyte_counts","log2_Neutrophil_counts","log2_CT","CNIH4","IFNG","CD274",colnames(GSVA_targeted))
```

# 4. Draft analysis

```{r}
library(ggrepel)
library(tidyverse)
library(ggpmisc)
library(ggpubr)
load(paste0(path_input,'Data/Module_discovery.Rdata'))
load(paste0(path_input,'Data/Gene Significance and Module Membership_log2_CT.Rdata'))
#load(paste0(path_input,"Data/PTB-02-networkConstruction-auto.RData"))
  
anno_coding <- subset(anno,class=="protein_coding")
# We only focus on the significant modules, with FDR<0.1 in both discovery and validation cohort
# Combine the p value from modules-traits association from two cohorts
moduleTraitPvalue_1<-as.data.frame(moduleTraitPvalue.adj[[1]])
moduleTraitPvalue_2<-as.data.frame(moduleTraitPvalue.adj[[2]])
colnames(moduleTraitPvalue_1)<-paste0(colnames(moduleTraitPvalue_1),"_discovery")
colnames(moduleTraitPvalue_2)<-paste0(colnames(moduleTraitPvalue_2),"_validation")
moduleTraitPvalue_wide<-cbind(moduleTraitPvalue_1,moduleTraitPvalue_2)
# Combine the correlation from modules-traits association from two cohorts
moduleTraitCor_1<-as.data.frame(moduleTraitCor[[1]])
moduleTraitCor_2<-as.data.frame(moduleTraitCor[[2]])
colnames(moduleTraitCor_1)<-paste0(colnames(moduleTraitCor_1),"_discovery")
colnames(moduleTraitCor_2)<-paste0(colnames(moduleTraitCor_2),"_validation")
moduleTraitCor_wide<-cbind(moduleTraitCor_1,moduleTraitCor_2)

selected_modules<-moduleTraitPvalue_wide%>% filter(log2_CT_discovery<0.1&log2_CT_validation<0.1) %>% rownames()
upregulated_modules<-selected_modules
p<-list()
output_hubs<-data.frame()
modNames = names(MEs[[1]]$data)

for(i in 1:length(upregulated_modules)){
  module<-upregulated_modules[i]
  modNames<-modNames
  column<-match(module,modNames);
  moduleGenes<-(moduleColors==module);

  selected_module<-data.frame(MM=geneModuleMembership[moduleGenes,column],
                           GA=-geneTraitAssociation_log2_CT[moduleGenes,1],
                           GS=geneTraitSignificant_log2_CT[moduleGenes,1])

  # Identify the hub genes as the genes above the 5% and 95% quantiles for two directional selection
  MM_cutoff <- quantile(selected_module$MM,prob=c(0.025,0.975))
  GA_cutoff <- quantile(selected_module$GA,prob=c(0.025,0.975))

  GAMM <- selected_module %>% mutate(gene=rownames(geneModuleMembership)[moduleGenes],
                                  module=module)

  selected_genes_module <- subset(GAMM,((MM < MM_cutoff[1])|(MM > MM_cutoff[2]))& ((GA < GA_cutoff[1])|(GA > GA_cutoff[2]))) %>% filter(gene %in% anno_coding$gene_symbol)
  ggplot(data=GAMM,aes(x=MM,y=GA))+
  geom_point(alpha=0.1,col="#313695",size=3)+
  theme_bw()+stat_poly_line(se=F) +
  xlab("Module Membership (PC1)")+
  ylab("Association between genes with bacterial load
")+
  geom_hline(yintercept=GA_cutoff,col="red",alpha=0.5,linetype="dashed")+
  geom_vline(xintercept=MM_cutoff,col="red",alpha=0.5,linetype="dashed")+
  geom_point(data=selected_genes_module,aes(x=MM,y=GA),col="#313695",size=1.5,alpha=0.9)+
  geom_text_repel(data=selected_genes_module,aes(x=MM,y=GA,label=gene),size=3.5,col="#A50026")+
  stat_poly_line(se=T) +
  stat_cor(size = 7,cor.coef.name = "r")+
  theme_bw()+
  theme(plot.title = element_text(size = 15),
        legend.position = "none",
        axis.text = element_text(size = 15),
        axis.text.x=element_text(color="black",size=13,face="bold"),
        axis.title.x = element_text(size = 15,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=15),
        axis.title.y = element_text(size = 15,color="black",face="bold"),
        legend.text = element_text(size = 15),
        legend.title = element_text(colour="black", size=12),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=12))




  }

selected_genes_module$gene
```

```{r}
common_gene <- c("GATAD2B","SMARCA5","ARNT","PRDM10","EED","EHMT2","ZNF146","TCF12","ATF4","RELB","YBX1","CREM","SP1","TRIM22","JUNB","BCLAF1","POLR2A","IRF9","ATF3","RAD51","NFRKB","NFIC","NKRF","RB1","USF2","ZNF592","MTA2","MEF2D","KLF5","BACH1","MAFG","ZNF207","HDAC2","EP300","SMAD5","EBF1","CEBPB","TARDBP","BCL11A","MTA3","ATF1","NR2F2","FOS","REST","ELF1","MEIS2","PKNOX1","KDM1A","HDGF","NFE2","ZNF24","ZNF217","ARID3A","BATF","ATF2","NR2F1","YY1","BHLHE40","HMBOX1","MAFF","MAX","NFE2L2","RFX5","IKZF1","RUNX3","SPI1","TAF1","SIN3A","ASH2L","JUND","MLLT1","TBL1XR1","TBP","EMSY","NBN","ZNF316","BCOR","TAF7","RCOR1","IKZF2","DPF2","MAFK","ZNF687")
```

```{r}
dim(selected_genes_module)
```
