---
title: "Transcriptome-wide bacterial burden association in patients with PTB"
author: "Triet Thai"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
#knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(
	echo = TRUE,
	message = FALSE,
	warning = FALSE)
library(parallel)
mcore<-1
# mcore<-1# for Window PC
#Load libraries
library(dplyr)
library(readxl)
library(writexl)
library(survival)
library(glmnet)
library(stabs)
library(tidyverse)
library(magrittr)
library(Hmisc)
library(gtsummary)
library(WGCNA)
library(ggcorrplot)
library(scales)
library(gridExtra)
library(ggcorrplot)
library(broom)
library(rms)
library(tidyverse)
library("stabs")
library("mboost")
library(DESeq2)
library(CEMiTool)
# path_output<-"/home/ubuntu/tb_volume/PTB_transcriptomics_project/Analysis/Results"
# ###------Import clinical data------###
# load("/home/ubuntu/tb_volume/PTB_transcriptomics_project/Raw_data/vst_Gensymbol_allPTB_295Sample_18Feb22.RData")
# clinical <-vst_data%>% select("Patient_ID","Age","SEX","BMI","Timika_score_CXR","Diabetes","Cavity_CXR","Lung_proportion_CXR","CT_mean","Timepoint","GeneXpert","HIV","trial_number","LIMS_ID") %>% rename("SEX"="Sex","trial_number"="Treatment_sensitive")

path_output<- paste0(dirname(getwd()),"/Triet_Results")
path_input <- paste0(str_replace_all(Sys.getenv("USERPROFILE"), '\\\\', '/'),
                     "/OneDrive - Oxford University Clinical Research Unit/#OUCRU-TB/#oucru-tb/PTB_project/")
###------Import clinical data------###
load(paste0(path_input,"/Raw_data/vst_Gensymbol_allPTB_295Sample_18Feb22.RData"))

clinical <-vst_data%>% select("Patient_ID","Age","SEX","BMI","Timika_score_CXR","Diabetes","Cavity_CXR","Lung_proportion_CXR","CT_mean","Timepoint","GeneXpert","HIV","trial_number","LIMS_ID") %>% rename("SEX"="Sex","trial_number"="Treatment_sensitive")
```

```{r include=FALSE}
# Import raw RNA_seq data, do the do the stabilizing variance transformation, and select the top 10000 with high variances genes
#path_rawRNA_seqdata<-"/home/ubuntu/tb_volume/Hai_Cpath_Analysis/Rstudio_sever/github_repos"
## load input data. This is the pooled data of all RNA_seq data
#lnames <- load(file=paste(path_rawRNA_seqdata,"RData/Data_RNAseq_RUN1andRUN2_15Feb2023.RData",sep="/"))
#CEMiTool
lnames <- load(file=paste0(path_input,"Raw_data/Data_RNAseq_RUN1andRUN2_15Feb2023.RData"))

## load clinical data from the meta data
#All_sample_RNAseq <- readxl::read_excel(paste(path_rawRNA_seqdata,"General_analysis/Infor/All_sample_RNAseq.xlsx",sep="/"),sheet = "Run1and2") %>% as.data.frame()
All_sample_RNAseq <- readxl::read_excel(paste0(path_input,"Raw_data/All_sample_RNAseq.xlsx"),sheet = "Run1and2") %>% as.data.frame()

rownames(All_sample_RNAseq) <- All_sample_RNAseq$LIMS_ID

## check clinical data and count data 
all(rownames(All_sample_RNAseq) %in% colnames(count_all))
dim(All_sample_RNAseq)## 1276 samples

## Extract only the data for PTB cohort (28,29TB)
PTB_sample_RNAseq <- All_sample_RNAseq %>% subset(Study %in% c("28TB","29TB"))
count_data <-count_all[,rownames(PTB_sample_RNAseq)]

## create anno data
anno_symbol <- anno[,c("gene_id", "gene_symbol")] %>% filter(gene_id %in% rownames(count_data))
rownames(anno_symbol) <- anno_symbol$gene_id
anno_symbol <- anno_symbol[rownames(count_data),]
## add gene symbol to count data
count_data <- cbind(genenames=anno_symbol$gene_symbol,count_data)
## Some genes have multiple segments that map to one gene. Thus, the counts of that gene is the sum of all the segments counts of that genes
count_deseq2 <- count_data %>%
                group_by(genenames) %>%
                summarise_all(sum) %>%
                data.frame()
rownames(count_deseq2) <- count_deseq2$genenames

count_deseq2 %<>% dplyr::select(-genenames)

## filter low expressed genes
filter_min_group <- ncol(count_deseq2)/5
## filter for gene expression of 10 counts in more than 20% sample
filter_gene <- apply(count_deseq2, 1, function(x) length(x[x>10]) >= filter_min_group) 
count_deseq2 <- count_deseq2[filter_gene,] %>% as.matrix()

## normalize vst by deseq2
PTB_sample_RNAseq %<>% mutate(Study=factor(Study))
dds <- DESeqDataSetFromMatrix(count_deseq2, colData=PTB_sample_RNAseq, ~ Study)
# rownames(dds)
# dds_anno <- anno_gene_symbol(data = dds,data_anno = anno,export = F,name = NULL)
vsd <- assay(vst(dds, blind=T)) # Variance stablized transform with log transform
vsd_t<-vsd %>% t() %>%
 as.data.frame() %>%
 tibble::rownames_to_column(var='LIMS_ID') %>%
 as.data.frame()

RNA_data<-vsd_t %>% select(-LIMS_ID)
rownames(RNA_data)<-vsd_t$LIMS_ID

## Method 1:  Using variance_based filtering method to filter gene most variance

RNA_data_transpose <- as.data.frame(t(RNA_data))
RNA_data_transpose_filtered <- filter_genes(RNA_data_transpose,pct =.75) #.75
# Selected genes
selected <- select_genes(RNA_data_transpose_filtered,filter_pval=0.35) #0.35

# ## Method 2:  Using the top 10000 genes as the most variance
# dv <- NULL
# dv$Symbols <- colnames(RNA_data)
# dv$variance <- matrixStats::colVars(as.matrix(RNA_data))
# dv <- as.data.frame(dv)
# ## sorting gene by var
# dv <- dv[order(dv$variance, decreasing = TRUE),]
# 
# 
# RNA_data$LIMS_ID<-vsd_t$LIMS_ID
# selected <- dv$Symbols[1:10000] #method 2 choose 10000 genes most variant

RNA_data <- RNA_data[,selected]
RNA_data$LIMS_ID<-vsd_t$LIMS_ID

expr_normalized_df <- data.frame(RNA_data) %>%
  pivot_longer(cols=RN7SL1:ZNF876P)
expr_normalized_df<-plyr::join_all(list(expr_normalized_df,PTB_sample_RNAseq))


expr_normalized_df %>% ggplot(., aes(x = Study, y = value)) +
  geom_violin() +
  geom_jitter(width = 0.1,alpha=0.1,size=1) +
  theme_bw() +
  theme(
    axis.text.x = element_text( angle = 90)
  ) +
  ylim(0, NA) +
  labs(
    title = "Normalized and 95 quantile Expression",
    x = "treatment",
    y = "normalized expression"
  )+ theme(plot.title = element_text(hjust = 0.5))->p

#png(file=paste0(path_output,"/Gene_expression_after_normalized_by_study.png"), res=400, units = 'in', width = 10, height =5)
p
#dev.off()


gene_names_full<-colnames(RNA_data)
gene_names_full<-gsub("-","_",gene_names_full)
colnames(RNA_data)<-gene_names_full

# Here we only combine the 295 patients who have good RNA_seq quality
clinical_RNA<-merge(RNA_data,clinical,by="LIMS_ID")
RNA_data<-clinical_RNA[,gene_names_full]
rownames(RNA_data)<-rownames(clinical_RNA)<-clinical_RNA$LIMS_ID
#save(RNA_data,clinical,file="W:/#oucru-tb/PTB_Project/PTB_transcriptomics_project/Data/Most_variance_RNA_genes.Rdata")
```

# Patient characteristics of discovery and validation cohort

```{r echo=FALSE}
load(file=paste0(path_input,"Data/Most_variance_RNA_genes.Rdata"))
# Here we only combine the 295 patients who have good RNA_seq quality
clinical_RNA<-merge(RNA_data,clinical,by="LIMS_ID")

rownames(clinical_RNA)<-rownames(RNA_data)
clinical_RNA%<>% as.data.frame() %>% mutate(log2_CT=log2(CT_mean))
datExpr0<-datExpr<-RNA_data %>% select(-LIMS_ID)
gene_names_full<-colnames(datExpr)

splitting_set_size <- 0.5
set.seed(1234) #4650
n<-nrow(clinical)
index_set_1 <- sample (1:n, floor(splitting_set_size * nrow(clinical)), replace=F)
index_set_2<-(1:n)[!(1:n%in%index_set_1)]

set_1_set  <- clinical[index_set_1, ] %>% select("Age","Sex","BMI","Treatment_sensitive","Diabetes","CT_mean","Cavity_CXR","Lung_proportion_CXR")
set_2_set <- clinical[index_set_2,]%>% select("Age","Sex","BMI","Treatment_sensitive","Diabetes","CT_mean","Cavity_CXR","Lung_proportion_CXR")

# Overview patient characteristics to see if the characteristics are distributed equally
set_1_set$Set <- 'Discovery set'
set_2_set$Set <- 'Validation set'

PTB_sets <- rbind(set_1_set, set_2_set) %>% 
  mutate(Set=factor(Set,levels=c('Discovery set','Validation set')))

table <- 
  tbl_summary(
    PTB_sets,
    by = Set, # split table by group
    missing = "ifany", # don't list missing data separately,
    statistic=list(all_continuous() ~ "{N_nonmiss}")
  ) %>%
  add_n() %>% # add column with total number of non-missing observations
  add_p() %>% # test for a difference between groups
  modify_header(label = "**Variable**") %>% # update the column header
  bold_labels()
table
```

## Overview:

-   We only focus on the top 10000 gene most variant. Those lower would be removed from the analysis. The results we chose this large cut-off because we dont want to missed signal from the small variant gene but huge impact on the bacterial burden. However, extreme small variant unliky to infuence on the bacterial burden rather than noise. And we also want to remove those noise.

-   GAMs fitted using GCV smoothness selection can suffer from under-smoothing. This can happen where the GCV profile is relative flat and random variation can lead to the algorithm converging at too wiggly a fit. Fitting via REML (use method = "REML" in the gam() call) or ML has been shown by Simon Wood and colleagues to be much more robust to under smoothing, but at computational expense. The above summaries are based on the descriptions in Simon Wood's rather excellent book on GAMs: Wood, S. N. (2006). Generalized Additive Models: An Introduction with R. Chapman and Hall/CRC

-   In this analysis, we defined the signal genes are those with top variances and associated with phenotype with p.value \<0.05 and the R_sq\>0.04

-   We're primarily interested in both linear and non-linear patterns of bacterial burden with respect to the gene expression levels. Therefore, we used the general form (non-linear univariate analysis) to detect the signal genes.

-   We splitted the data into two indenpendt cohorts: set_1 and set_2 sets: We peform the same procedure of ranking and then take the common top-hit genes.

-   We further investigate the network analysis (WGCNA analysis) to investigate the potential cluster of genes which regulates the host responses to control the bacterial burden.

## Pre-screeing signals from both cohorts

```{r eval=FALSE, include=FALSE}
## Step 1: Pre-screening signal genes (gene with p<0.05 and R2>0.04 ) regardless linear or non-linear associations using the whole data set.
## We use the method "REML" for a more robust smoothness selection.
signal_genes<-c()
# Null model
f0<-mgcv::gam(log2_CT~1,data=clinical_RNA[index_set_1,c(gene_names_full,"log2_CT")],family="gaussian",method="REML")
Rsq_adj<-df<-c()
for(i in gene_names_full){
  f<-formula(paste0("~ s(",i,")"))
  sum_f<-summary(update(f0,f))
  if((sum_f$s.table[,"p-value"]<0.05)&(sum_f$r.sq>0.04)){
    signal_genes<-c(signal_genes,i)
    Rsq_adj<-c(Rsq_adj,unname(sum_f$r.sq))
  }
}
R_sq<-data.frame(signal_genes=signal_genes,Rsq_adj=Rsq_adj) %>% arrange(desc(Rsq_adj))
# save(signal_genes,R_sq,file="/home/ubuntu/tb_volume/PTB_transcriptomics_project/Analysis/Data/signal_genes_discovery.Rdata")
```

```{r eval=FALSE, include=FALSE}
## Step 1: Pre-screening signal genes (gene with signal ) regardless linear or non-linear associations using the whole data set.
signal_genes<-c()
f0<-mgcv::gam(log2_CT~1,data=clinical_RNA[index_set_2,c(gene_names_full,"log2_CT")],family="gaussian",method="REML")
Rsq_adj<-df<-c()
for(i in gene_names_full){
  f<-formula(paste0("~ s(",i,")"))
  sum_f<-summary(update(f0,f))
  if((sum_f$s.table[,"p-value"]<0.05)&(sum_f$r.sq>0.04)){
    signal_genes<-c(signal_genes,i)
    Rsq_adj<-c(Rsq_adj,unname(sum_f$r.sq))
  }
}
R_sq<-data.frame(signal_genes=signal_genes,Rsq_adj=Rsq_adj) %>% arrange(desc(Rsq_adj))
# save(signal_genes,R_sq,file="/home/ubuntu/tb_volume/PTB_transcriptomics_project/Analysis/Data/signal_genes_validation.Rdata")
```

### Step 1: Ranking signal genes with criteria (150-1500) and selecting top-hit genes on set_1 set.

```{r message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
load(file=paste0(path_input,"Data/signal_genes_discovery.Rdata"))
set.seed(123)
B <- 2000
m<-ceiling(floor(splitting_set_size * nrow(clinical_RNA))*2/3)
gene_names<-signal_genes
branks<-as.data.frame(matrix(0,ncol=length(gene_names),nrow=B))
colnames(branks) <- gene_names
R2<-branks
```

```{r eval=FALSE, message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
library(matrixStats)
R2_compute<-function(data=data,gene_names=gene_names,variables_names=variables_names){
  #Parallel computing
  f0<-mgcv::gam(log2_CT~1,data,family="gaussian",method="REML")
  l<-mclapply(1:length(gene_names),function(i){
    summary(update(f0,formula(paste0("~ s(",gene_names[i],")"))))$r.sq
    },mc.cores = mcore)
  return(unlist(l))
}

for(k in 1:B) {
  index <- sample(index_set_1, m, replace=F)
  bR2<- R2_compute(data=clinical_RNA[index,],
                    gene_names=gene_names,
                    variables_names=c("log2_CT"))
  names(bR2)<-gene_names
  R2[k,]<-bR2
  branks[k,]<-order(sort(bR2,decreasing = T,index.return=T)$ix)
  print(k)
}
# file_names<-paste0("/home/ubuntu/tb_volume/PTB_transcriptomics_project/Analysis/correlation_set_1_set_",gsub("-","_", Sys.Date()),".Rdata")
# save(branks,R2,file=file_names)
```

#### Re-sampling ranking based on the estimate R_sq of genes on Gene-Xpert Ct value on discovery set.

```{r fig.height=10, fig.width=10, message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
load(file=paste0(path_input,"Data/correlation_set_1_set_2023_07_16.Rdata"))

gene_names<-colnames(branks)
branks<-as.matrix(branks)
branks_sum<-matrixStats::colQuantiles(branks,probs = c(0.025,0.975,0.5)) %>% as.data.frame()
branks_sum$means_ranks<-colMeans(branks)

branks_sum<-as.data.frame(branks_sum) %>% arrange(means_ranks)
names(branks_sum)[1:3]<-paste0("rank-",colnames(branks_sum[,1:3]))
branks_sum$gene_names<-rownames(branks_sum)

branks_sum %>%
  mutate(name = reorder(gene_names, dplyr::desc(`rank-97.5%`))) %>%
  ggplot( aes(x=name, y=`rank-50%`)) +
    geom_point(col="red") +
  geom_errorbar(aes(ymin = `rank-2.5%`, ymax = `rank-97.5%`,alpha=0.01))+
  theme_classic()+
  theme(axis.text.y = element_blank())+
  geom_hline(yintercept =c(50,500), linetype = 'dashed', color= 'red',size=1)+
    coord_flip() +
    xlab("") +
    theme_bw()+ theme(legend.position="none")+ylab("Rank of  genes")+theme(panel.grid.minor = element_blank(),axis.title.x = element_text(face="bold", colour="black", size=12),
        axis.title.y = element_text(face="bold", colour="black", size=12),
        axis.text.y = element_blank(),
        axis.text.x = element_text( colour="black", size=13),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=12),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=12))+ggtitle("Bi-directional ranking
") +  theme(plot.title = element_text(hjust = 0.5,face="bold", colour="black", size=20))->p1

R2<-as.matrix(R2)
R2_sum<-matrixStats::colQuantiles(R2,probs = c(0.025,0.975,0.5)) %>% as.data.frame()
R2_sum$means_R2<-colMeans(R2)

R2_sum<-as.data.frame(R2_sum) %>% arrange(means_R2)
names(R2_sum)[1:3]<-paste0("R2-",colnames(R2_sum[,1:3]))
R2_sum$gene_names<-rownames(R2_sum)

R2_sum<-merge(R2_sum,branks_sum,by="gene_names")
R2_sum %>%
  mutate(name = reorder(gene_names, dplyr::desc(`rank-97.5%`))) %>%
  ggplot( aes(x=name, y=`R2-50%`)) +
    geom_point(col="red") +
  geom_errorbar(aes(ymin = `R2-2.5%`, ymax = `R2-97.5%`,alpha=0.01))+
  theme_classic()+
  theme(axis.text.y = element_blank())+
    coord_flip() +
    xlab("") +
    theme_bw()+ theme(legend.position="none")+ylab("R2")+theme(panel.grid.minor = element_blank(),axis.title.x = element_text(face="bold", colour="black", size=12),
        axis.title.y = element_text(face="bold", colour="black", size=12),
        axis.text.y = element_blank(),
        axis.text.x = element_text( colour="black", size=13),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=12),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=12))+ggtitle("Bi-directional ranking
") +  theme(plot.title = element_text(hjust = 0.5,face="bold", colour="black", size=20))->p2
R2_sum_set_1_set<-R2_sum %<>%  arrange((`rank-50%`))
#save(R2_sum_set_1_set,file="/home/ubuntu/tb_volume/PTB_transcriptomics_project/Analysis/R2_sum_set_1.Rdata")
```

### Step 2: Ranking signal genes with criteria (150-1500) and selecting top-hit genes on set_2 set.

```{r message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
load(file=paste0(path_input,"Data/signal_genes_validation.Rdata"))
set.seed(123)
B <- 2000
m<-ceiling(floor(splitting_set_size * nrow(clinical_RNA))*2/3)
gene_names<-signal_genes
branks<-as.data.frame(matrix(0,ncol=length(gene_names),nrow=B))
colnames(branks) <- gene_names
R2<-branks
```

```{r eval=FALSE, message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
library(matrixStats)
R2_compute<-function(data=data,gene_names=gene_names,variables_names=variables_names){
  #Parallel computing
  f0<-mgcv::gam(log2_CT~1,data,family="gaussian",method="REML")
  l<-mclapply(1:length(gene_names),function(i){
    summary(update(f0,formula(paste0("~ s(",gene_names[i],")"))))$r.sq
    },mc.cores = mcore)
  return(unlist(l))
}

for(k in 1:B) {
  index <- sample(index_set_2, m, replace=F)
  bR2<- R2_compute(data=clinical_RNA[index,],
                    gene_names=gene_names,
                    variables_names=c("log2_CT"))
  names(bR2)<-gene_names
  R2[k,]<-bR2
  branks[k,]<-order(sort(bR2,decreasing = T,index.return=T)$ix)
  print(k)
}
# file_names<-paste0("/home/ubuntu/tb_volume/PTB_transcriptomics_project/Analysis/correlation_set_2_set_",gsub("-","_", Sys.Date()),".Rdata")
# save(branks,R2,file=file_names)
```

#### Re-sampling ranking based on the estimate HR of genes on two-month mortality on validation set.

```{r fig.height=10, fig.width=10, message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
load(file=paste0(path_input,"Data/correlation_set_2_set_2023_07_16.Rdata"))
gene_names<-colnames(branks)
branks<-as.matrix(branks)
branks_sum<-matrixStats::colQuantiles(branks,probs = c(0.025,0.975,0.5)) %>% as.data.frame()
branks_sum$means_ranks<-colMeans(branks)

branks_sum<-as.data.frame(branks_sum) %>% arrange(means_ranks)
names(branks_sum)[1:3]<-paste0("rank-",colnames(branks_sum[,1:3]))
branks_sum$gene_names<-rownames(branks_sum)

branks_sum %>%
  mutate(name = reorder(gene_names, dplyr::desc(`rank-97.5%`))) %>%
  ggplot( aes(x=name, y=`rank-50%`)) +
    geom_point(col="red") +
  geom_errorbar(aes(ymin = `rank-2.5%`, ymax = `rank-97.5%`,alpha=0.01))+
  theme_classic()+
  theme(axis.text.y = element_blank())+
  geom_hline(yintercept =c(50,500), linetype = 'dashed', color= 'red',size=1)+
    coord_flip() +
    xlab("") +
    theme_bw()+ theme(legend.position="none")+ylab("Rank of  genes")+theme(panel.grid.minor = element_blank(),axis.title.x = element_text(face="bold", colour="black", size=12),
        axis.title.y = element_text(face="bold", colour="black", size=12),
        axis.text.y = element_blank(),
        axis.text.x = element_text( colour="black", size=13),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=12),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=12))+ggtitle("Bi-directional ranking
") +  theme(plot.title = element_text(hjust = 0.5,face="bold", colour="black", size=20))->p1

R2<-as.matrix(R2)
R2_sum<-matrixStats::colQuantiles(R2,probs = c(0.025,0.975,0.5)) %>% as.data.frame()
R2_sum$means_R2<-colMeans(R2)

R2_sum<-as.data.frame(R2_sum) %>% arrange(means_R2)
names(R2_sum)[1:3]<-paste0("R2-",colnames(R2_sum[,1:3]))
R2_sum$gene_names<-rownames(R2_sum)

R2_sum<-merge(R2_sum,branks_sum,by="gene_names")
R2_sum %>%
  mutate(name = reorder(gene_names, dplyr::desc(`rank-97.5%`))) %>%
  ggplot( aes(x=name, y=`R2-50%`)) +
    geom_point(col="red") +
  geom_errorbar(aes(ymin = `R2-2.5%`, ymax = `R2-97.5%`,alpha=0.01))+
  theme_classic()+
  theme(axis.text.y = element_blank())+
    coord_flip() +
    xlab("") +
    theme_bw()+ theme(legend.position="none")+ylab("R2")+theme(panel.grid.minor = element_blank(),axis.title.x = element_text(face="bold", colour="black", size=12),
        axis.title.y = element_text(face="bold", colour="black", size=12),
        axis.text.y = element_blank(),
        axis.text.x = element_text( colour="black", size=13),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=12),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=12))+ggtitle("Bi-directional ranking
") +  theme(plot.title = element_text(hjust = 0.5,face="bold", colour="black", size=20))->p2
R2_sum_set_2_set<-R2_sum %<>%  arrange((`rank-50%`))
#save(R2_sum_set_2_set,file="/home/ubuntu/tb_volume/PTB_transcriptomics_project/Analysis/R2_sum_set_2.Rdata")
```

### Combine signals from the two independent sets

```{r message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
load(file=paste0(path_input,"Data/R2_sum_set_1.Rdata"))
load(file=paste0(path_input,"Data/R2_sum_set_2.Rdata"))

R2_sum_set_1_set %<>% mutate(select=ifelse((`rank-2.5%`<=50 & `rank-97.5%`<=500 & `R2-50%`>0.04),"yes","no")) 
R2_sum_set_2_set %<>% mutate(select=ifelse((`rank-2.5%`<=50 & `rank-97.5%`<=500 & `R2-50%`>0.04),"yes","no")) 

top_hit_set_1_signal<-R2_sum_set_1_set$gene_names[R2_sum_set_1_set$select=="yes"]
top_hit_set_2_signal<-R2_sum_set_2_set$gene_names[R2_sum_set_2_set$select=="yes"]

(common_list<-top_hit_set_1_signal[top_hit_set_1_signal%in%top_hit_set_2_signal])
a<-cbind(gene_names=common_list)

#write.table(a,file="/home/ubuntu/tb_volume/PTB_transcriptomics_project/Analysis/top_hit.txt",quote = F,row.names = F)

R2_sum_set_1_set_tophit<-R2_sum_set_1_set %>% filter(select=="yes") %>% mutate(Set="Discovery set") %>% as.data.frame()
R2_sum_set_2_set_tophit<-R2_sum_set_2_set %>% filter(select=="yes") %>% mutate(Set="Validation set")%>% as.data.frame()

R2_top_hit<-rbind(R2_sum_set_1_set_tophit,R2_sum_set_2_set_tophit) %>% mutate(validated_signal=ifelse(gene_names%in%common_list,"yes","no"))
# save(R2_top_hit,file="/home/ubuntu/tb_volume/PTB_transcriptomics_project/Analysis/R2_top_hit.Rdata")
```

#### Forest-plots of top-hit genes

```{r message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
load(file=paste0(path_input,"Data/R2_top_hit.Rdata"))
```

```{r echo=FALSE, fig.height=30, fig.width=30, message=FALSE, warning=FALSE, paged.print=FALSE}
p1<-R2_top_hit %>% mutate(name = reorder(gene_names, dplyr::desc(`rank-50%`))) %>% filter(validated_signal=="yes") %>% 
  ggplot( aes(x=name, y=`rank-50%`))+
    geom_point(aes(col="red"),size=2)+
  geom_errorbar(aes(ymin = `rank-2.5%`, ymax = `rank-97.5%`,alpha=0.01),size=1)+facet_grid(~Set)+
  theme_classic()+
  theme(axis.text.y = element_blank())+
  geom_hline(yintercept =c(50,500), linetype = 'dashed', color= 'red',size=1)+
    coord_flip() +
    xlab("") +ylab("Rank based on R2")+
    theme_bw()+ theme(legend.position="none")+theme(panel.grid.minor = element_blank(),axis.title.x = element_text(face="bold", colour="black", size=13),
        axis.title.y = element_text(face="bold", colour="black", size=13),
        axis.text.y = element_text(face="bold", colour="black", size=13),
        axis.text.x = element_text( colour="black", size=13),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=13),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=18),
        plot.title = element_text(face = "bold",size=18,hjust = 0.5))
p2<-R2_top_hit %>% mutate(name = reorder(gene_names, dplyr::desc(`rank-50%`))) %>% filter(validated_signal=="yes") %>%
  ggplot( aes(x=name, y=`R2-50%`))+
    geom_point(col="red",size=2)+
  geom_errorbar(aes(ymin = `R2-2.5%`, ymax =`R2-97.5%`,alpha=0.01),size=1)+facet_grid(~Set)+
  theme_classic()+
  theme(axis.text.y = element_blank())+
  geom_hline(yintercept =c(0), linetype = 'dashed', color= '#2ca25f',size=1) +
      coord_flip() +
    xlab("") +ylab("R2")+
    theme_bw()+ theme(legend.position="none")+theme(panel.grid.minor = element_blank(),axis.title.x = element_text(face="bold", colour="black", size=13),
        axis.title.y = element_text(face="bold", colour="black", size=13),
        axis.text.y = element_text(face="bold", colour="black", size=13),
        axis.text.x = element_text( colour="black", size=13),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=13),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=18),
        plot.title = element_text(face = "bold",size=18,hjust = 0.5))
cowplot::plot_grid(p1,p2, nrow = 1)
# png('/home/ubuntu/tb_volume/PTB_transcriptomics_project/Analysis/Results/Ranking_R2_signal_genes.png',res=400, units = 'in', width = 20, height =20)
cowplot::plot_grid(p1,p2, nrow = 2)
# dev.off()
```

#### Correlation structure between the top-hit genes

```{r echo=FALSE, fig.height=40, fig.width=120, message=FALSE, warning=FALSE, paged.print=FALSE}
clinical_RNA_signal<-clinical_RNA
corr1 <- round(cor(clinical_RNA_signal[index_set_1,common_list],use = 'pairwise.complete.obs'), 2)
p1<-ggcorrplot(corr1, hc.order = TRUE, outline.col = "white",lab = TRUE,lab_size = 2)+labs(title = "Discovery set")

corr2 <- round(cor(clinical_RNA_signal[index_set_2,common_list],use = 'pairwise.complete.obs'), 2)
p2<-ggcorrplot(corr2, hc.order = TRUE, outline.col = "white",lab = TRUE,lab_size = 2)+labs(title = "Validation set")
cowplot::plot_grid(p1,p2, nrow = 1)

# png('/home/ubuntu/tb_volume/PTB_transcriptomics_project/Analysis/Results/correlation_top_hit.png',res=400, units = 'in', width = 10, height =5)
cowplot::plot_grid(p1,p2, nrow = 1)
# dev.off()
```

```{r}

```

```{r message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
dat_top_hit_set_1<-clinical_RNA[index_set_1,c(common_list,"log2_CT")] %>% mutate(Set="Discovery set") 
dat_top_hit_set_2<-clinical_RNA[index_set_2,c(common_list,"log2_CT")] %>% mutate(Set="Validation set")

dat_top_hit_set_3<-Fludigm_PTB[,c(common_list,"log2_CT")] %>% mutate(Set="Validation set")

dat_top_hit<-rbind(dat_top_hit_set_1,dat_top_hit_set_2)
dat_top_hit$Set<-as.factor(dat_top_hit$Set)
```

We will visualize the non-linear pattern of the signal cross two subset

```{r message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
library(cowplot)
d2_long <- gather(dat_top_hit, gene_names, concentration, `CNIH4`:`AC244453.2`, factor_key=TRUE)

p<-list()
Gene_names<-unique(d2_long$gene_names)
for(i in 1:length(Gene_names)){
  df <-subset(d2_long,gene_names==Gene_names[i])
  
if(i%in%c(1,6)){
  ylab<-"log2 CT"
}else{
  ylab<-""
}

p_tmp<-ggplot(data=df,aes(x=concentration ,y=log2_CT,colour=Set ))+geom_point(size=2.5,alpha=0.5)+geom_smooth(method = "gam",formula = y ~ s(x),se = TRUE,size=2)+
  scale_color_manual(values=c("#313695", "#A50026"))+
  theme_bw()+
  labs(x = paste(Gene_names[i]," expression level [log2 counts] "), y = ylab)+theme_bw()+theme(legend.position="bottom")+ theme(panel.grid.minor = element_blank(),axis.title.x = element_text(face="bold", colour="black", size=12),
        axis.title.y = element_text(face="bold", colour="black", size=20),
        axis.text.y = element_text( colour="black", size=15),
        axis.text.x = element_text( colour="black", size=15),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=15),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=15))

xdens <- axis_canvas(p_tmp, axis = "x")+geom_density(data=df,aes(x=concentration,fill=Set),alpha=0.25)+scale_fill_manual( values = c("#313695", "#A50026"))
 
p1 <- insert_xaxis_grob(p_tmp, xdens, grid::unit(.2, "null"), position = "top")
p[[i]]<-ggdraw(p1)
}

p<-cowplot::plot_grid(plotlist = p, nrow = 2)
png(file=paste0(path_output,"/Figure1_top_hit_Gene_names_plot.png"),res=450, units = 'in', width = 30, height =24)
p
dev.off()
```

**Of note:** One remaining concern in this analysis is to decide the degree of freedom for the trend of each gene in the smoothness selection model. Should we chose the most flexibility model or we chose the parsimonious one? --\> use the REML method to improve this.

#### Checking the degree of freedom of each gene by cohort

```{r eval=FALSE, include=FALSE}
## Step 1: Pres-screening signal genes (gene with signal ) regardless linear or non-linear associations using the whole data set.

signal_genes<-common_list
p.val<-c()
f0<-mgcv::gam(log2_CT~1,data=clinical_RNA[index_set_1,],family="gaussian",method = "REML",select=T)
Rsq_adj<-df<-c()
for(i in signal_genes){
  f<-formula(paste0("~ s(",i,")"))
  sum_f<-summary(update(f0,f))
    signal_genes<-c(signal_genes,i)
    p.val<-c(p.val,sum_f$s.table[,"p-value"])
    Rsq_adj<-c(Rsq_adj,sum_f$r.sq)
    df<-c(df,sum_f$edf)
}

dat_set_1<-data.frame(gene_names=signal_genes,
                Rsq_adj_set_1=Rsq_adj,
                df_set_1=df,p.val_set_1=p.val,
                Set="Discovery set")

p.val<-c()
f0<-mgcv::gam(log2_CT~1,data=clinical_RNA[index_set_2,],family="gaussian",method="REML")
Rsq_adj<-df<-c()
for(i in signal_genes){
  f<-formula(paste0("~ s(",i,")"))
  sum_f<-summary(update(f0,f))
    signal_genes<-c(signal_genes,i)
    p.val<-c(p.val,sum_f$s.table[,"p-value"])
    Rsq_adj<-c(Rsq_adj,sum_f$r.sq)
    df<-c(df,sum_f$edf)
}

dat_set_2<-data.frame(gene_names=signal_genes,
                Rsq_adj_set_2=Rsq_adj,
                df_set_2=df,p.val_set_2=p.val,
                Set="Validation set") 
dat<-cbind(dat_set_1,dat_set_2)
dat$df_parsimony_model<-ceiling(pmin(dat$df_set_1,dat$df_set_2))
# save(dat,file="/home/ubuntu/tb_volume/PTB_transcriptomics_project/Analysis/Data/validated_results.Rdata")
#The association is very weak to validate by FDR.
```

```{r}
library(cowplot)
d2_long <- gather(dat_top_hit, gene_names, concentration, `CNIH4`:`AC244453.2`, factor_key=TRUE)
df<-dat$df_parsimony_model
p<-list()
Gene_names<-unique(d2_long$gene_names)
for(i in 1:length(Gene_names)){
  dat_tmp <-subset(d2_long,gene_names==Gene_names[i])
  j<-df[i]
# if(i%in%c(1,6)){
#   ylab<-"Mortality at two months"
# }else{
#   ylab<-""
# }

p_tmp<-ggplot(data=dat_tmp,aes(x=concentration ,y=log2_CT,colour=Set ))+geom_point(size=2.5,alpha=0.5)+geom_smooth(method = "lm",formula = y ~ splines::ns(x,j),se = TRUE,size=2)+
  scale_color_manual(values=c("#313695", "#A50026"))+
  theme_bw()+
  labs(x = paste(Gene_names[i]," expression level [log2 counts] "), y = ylab)+theme_bw()+theme(legend.position="bottom")+ theme(panel.grid.minor = element_blank(),axis.title.x = element_text(face="bold", colour="black", size=12),
        axis.title.y = element_text(face="bold", colour="black", size=20),
        axis.text.y = element_text( colour="black", size=15),
        axis.text.x = element_text( colour="black", size=15),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=15),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=15))

xdens <- axis_canvas(p_tmp, axis = "x")+geom_density(data=dat_tmp,aes(x=concentration,fill=Set),alpha=0.25)+scale_fill_manual( values = c("#313695", "#A50026"))
 
p1 <- insert_xaxis_grob(p_tmp, xdens, grid::unit(.2, "null"), position = "top")
p[[i]]<-ggdraw(p1)
}

p<-cowplot::plot_grid(plotlist = p, nrow = 2)
png(file=paste0(path_output,"/Figure1_top_hit_Gene_names_plot.png"),res=450, units = 'in', width = 25, height =50)
p
dev.off()
```

```{r}
library(cowplot)
d2_long <- gather(dat_top_hit, gene_names, concentration, `CNIH4`:`AC244453.2`, factor_key=TRUE)
df<-dat$df_parsimony_model
p<-list()
Gene_names<-unique(d2_long$gene_names)
for(i in 1:length(Gene_names)){
  dat_tmp <-subset(d2_long,gene_names==Gene_names[i])
  j<-df[i]
# if(i%in%c(1,6)){
#   ylab<-"Mortality at two months"
# }else{
#   ylab<-""
# }

p_tmp<-ggplot(data=dat_tmp,aes(x=concentration ,y=log2_CT ))+geom_point(size=2.5,alpha=0.5)+geom_smooth(method = "lm",formula = y ~ splines::ns(x,j),se = TRUE,size=2)+
  theme_bw()+
  labs(x = paste(Gene_names[i]," expression level [log2 counts] "), y = ylab)+theme_bw()+theme(legend.position="bottom")+ theme(panel.grid.minor = element_blank(),axis.title.x = element_text(face="bold", colour="black", size=12),
        axis.title.y = element_text(face="bold", colour="black", size=20),
        axis.text.y = element_text( colour="black", size=15),
        axis.text.x = element_text( colour="black", size=15),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=15),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=15))

xdens <- axis_canvas(p_tmp, axis = "x")+geom_density(data=dat_tmp,aes(x=concentration),alpha=0.25)
 
p1 <- insert_xaxis_grob(p_tmp, xdens, grid::unit(.2, "null"), position = "top")
p[[i]]<-ggdraw(p1)
}

p<-cowplot::plot_grid(plotlist = p, nrow = 2)
png(file=paste0(path_output,"/Figure1_top_hit_Gene_names_pooled_plot.png"),res=450, units = 'in', width = 25, height =50)
p
dev.off()
```

#### Association of important genes vs bacterial burden (set_1 set)

```{r echo=FALSE, message=FALSE, warning=FALSE}
R2_sum_set_1_set %>% filter(gene_names%in%common_list) %>% 
  select(gene_names,`R2-50%`,`R2-2.5%`,`R2-97.5%`)->tab1
knitr::kable(tab1)
```

#### Association of important genes vs bacterial burden (set_2 set)

```{r echo=FALSE, message=FALSE, warning=FALSE}
R2_sum_set_2_set  %>% 
  select(gene_names,`R2-50%`,`R2-2.5%`,`R2-97.5%`)->tab1
knitr::kable(tab1)
```

```{r echo=FALSE, fig.height=20, fig.width=25, message=FALSE, warning=FALSE, paged.print=FALSE}
## Step 3. Identify the important genes using sampling stability method on the set_2 set
dat_top_hit<-clinical_RNA[c(common_list,"log2_CT")]
#dat_top_hit<-as.data.frame(cbind(x0[,gene_list],y= y))
mod <- gamboost(log2_CT ~ ., data = dat_top_hit,control = boost_control(risk = "oobag"),dfbase=3)
par(mfrow = c(5, 4))
plot(mod,cex.lab=3, cex.axis=3, cex.main=3, cex.sub=3,font=2)
## First, we check the parameter constellation: Where is the cutoff for a
## PFER of 1 when q = 10 variables are selected per boosting run if we
## assume unimodality:
#stabsel_parameters(mod, q=2, PFER = 1,sampling.type = "SS",B=100)

(stab.boot <- stabs::stabsel(mod, q =2, PFER = 1, mc.cores = mcore,sampling.type = "SS",assumption = "r-concave",B=200))
# save(stab.boot,file=paste0(path_output,"/variable_selection.Rdata"))
```

```{r}
load(file=paste0(path_input,"Data/variable_selection.Rdata"))
library("stringr")
dat_plot<-data.frame(variable=names(stab.boot$max),`Selected probability`=stab.boot$max) 
dat_plot %<>% mutate(variable=c("CNIH4","DPRXP2","TLR5","SLC26A8","AC244453.2")) 

dat_plot %>%
  mutate(name = reorder(variable, Selected.probability)) %>%
  ggplot(aes(y=name,x=Selected.probability))+geom_point(col="#2c7fb8",size=4,shape=15)+
  #geom_vline(xintercept=stab.boot$cutoff,col="red")+
  geom_line(linetype = "dashed")+
  xlab("Selected probability")+theme_bw()+theme(legend.position="bottom")+ theme(panel.grid.minor = element_blank(),axis.title.x = element_text(face="bold", colour="black", size=20),
        axis.title.y = element_blank(),
        axis.text.y = element_text(face = "bold", colour="black", size=18),
        axis.text.x = element_text( colour="black", size=15),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=15),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=15),
        plot.title = element_text(size = 20, face = "bold",hjust = 0.5))+coord_cartesian(xlim=c(0,0.8))->p1


png(file=paste0(path_output,"/Figure_GBX.png"),res=450, units = 'in', width =7, height =5)
p1

dev.off()
```

```{r , echo=FALSE, fig.height=12, fig.width=8, message=FALSE, warning=FALSE, paged.print=FALSE}
library(cowplot)
gene_names <-common_list
dat_top_hit_validate <- subset(clinical_RNA,select=c(gene_names,"log2_CT","Age"))
d2_long <- gather(dat_top_hit_validate, gene_names, expression_level, CNIH4:FCGR1A, factor_key=TRUE)

p<-list()
names<-unique(d2_long$gene_names)
for(i in 1:length(gene_names)){
df <-d2_long %>% filter(gene_names==names[i])
# if(mod(i,4)==1){
#   ylab<-"log2 CT"
# }else{
#   ylab<-""
# }
p_tmp<-ggplot(data=df,aes(x=expression_level ,y=log2_CT))+geom_point(size=2.5,alpha=0.5)+geom_smooth(method = lm,formula = y ~ splines::bs(x, 3), se = T,size=2,lapha=0.5)+
  theme_bw()+
  labs(x = paste(gene_names[i]," abundance [log2 +1sd] "), y = "log2 CT")+theme_bw()+theme(legend.position="bottom")+ theme(panel.grid.minor = element_blank(),axis.title.x = element_text(face="bold", colour="black", size=12),
        axis.title.y = element_text(face="bold", colour="black", size=20),
        axis.text.y = element_text( colour="black", size=15),
        axis.text.x = element_text( colour="black", size=15),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=15),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=15))+coord_cartesian(ylim=c(3,5.5))

xdens <- axis_canvas(p_tmp, axis = "x")+ geom_density(data=df,aes(x=expression_level),alpha=0.25,col="#313695")
 
p1 <- insert_xaxis_grob(p_tmp, xdens, grid::unit(.2, "null"), position = "top")
p[[i]]<-ggdraw(p1)
}


p<-cowplot::plot_grid(plotlist = p, ncol = 4)
# png(file=paste0(path_output,"/FigureS9_combined_genes_plot.png"),res=450, units = 'in', width = 25, height =22)
p
png('/home/ubuntu/tb_volume/PTB_transcriptomics_project/Analysis/Results/association_top_hit.png',res=400, units = 'in', width = 20, height =20)
p
dev.off()
```

#### Additional plot

```{r}
l<-load("/home/ubuntu/tb_volume/PTB_transcriptomics_project/Analysis/Data/Data_26_27_CT_CNIH4.Rdata.RData")

save(d2_long,file="d2_long.Rdata")

d2_long %>% filter(gene_names=="CNIH4") %>% ggplot(aes(x=concentration ,y=log2_CT))+geom_point(size=2.5,alpha=0.5)+geom_smooth(method = "gam",formula = y ~ s(x),se = TRUE,size=2,col="#313695",fill="#313695")+
  theme_bw()+
  labs(x = paste("CNIH4"," expression level [log2 counts] "), y = "Mtb Xpert Ct value")+ theme(panel.grid.minor = element_blank(),axis.title.x = element_text(face="bold", colour="black", size=20),
        axis.title.y = element_text(face="bold", colour="black", size=20),
        axis.text.y = element_text( colour="black", size=15),
        axis.text.x = element_text( colour="black", size=15),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=15),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=15))+scale_y_continuous(trans = log2_trans(),
    breaks = log2(seq(10,40,by=5)),
    labels = seq(10,40,by=5))->p

png(file=paste0(path_output,"/PTB_CNIH4.png"),res=450, units = 'in', width =10, height =10)
p
dev.off()
```

### Validation PCR method (Triet)

```{r}
load('C:/Users/triettm/OneDrive - Oxford University Clinical Research Unit/#OUCRU-TB/#oucru-tb/PTB_project/Raw_data/Fludigm_CNIH4_PTB.RData')

All_Fluidigm_data <- readxl::read_excel(paste0(path_input,"Raw_data/Data Fluidigm_All.xlsx"),sheet = "All") %>% as.data.frame()

tmp<-filter(All_Fluidigm_data,(Study%in% c('28TB','29TB'))&
                              (Sample_Type=='Blood')&
                              (Timepoint=='D0')&
                              (Ct_Call=='Pass')&
                              (EvaGreen_Name%in%common_list))
table(tmp$EvaGreen_Name)
dim(tmp)
```

### Correlation of 24 immune checkpoint gene vs CNIH4

```{r}
impute_df <- read.csv(paste0(path_input,'/Raw_data/CIBERSORTx_Adjusted_CPM_LM22.txt'),sep='\t') %>% filter(P.value<0.05)
rownames(impute_df) <- impute_df$Mixture
impute_df$Mixture <- NULL
```

```{r}
cor.pval.test <- function(x1, x2, ...) {
  r <- ncol(x1)
  n <- ncol(x2)
  p.mat<- matrix(NA, r, n)
  for (i in 1:r) {
    for (j in 1:n) {
      tmp <- cor.test(x1[, i], x2[, j], ...)
      p.mat[i, j] <- tmp$p.value
    }
  }
  rownames(p.mat) <-  colnames(x1)
  colnames(p.mat) <- colnames(x2)
  return(p.mat)
}
```

```{r}
# All
library(tidyverse)
library(stats)
library(corrplot) 
library(ggtext)
library(grid)
library(dplyr)
IC_names<-c('CD200', 'TNFSF4', 'CD200R1', 'IDO1','TNFRSF18', 'TNFSF18',
             'CD28','BTLA', 'CD27', 'HAVCR2', 'CD274', 'TNFSF14', 'TNFRSF4',
             'CD276', 'TNFRSF14', 'PDCD1', 'CD86', 'TNFSF9', 'TIGIT','CD80',
             'ICOS', 'TNFRSF9', 'CTLA4', 'LAG3','CNIH4'

             )

ICvsCNIH4_df <- vst_data %>% select(IC_names)
ICvsCNIH4_df <- merge(ICvsCNIH4_df, impute_df, by=0) %>% 
  select(append(IC_names,c('LGALS9','CD4',
                           'LYMLE',
                           'CT_mean',
                           "Neutrophils",
                            "Monocytes",
                            "T.cells.CD8",
                            "T.cells.CD4.naive",
                            "T.cells.CD4.memory.resting",
                            "T.cells.CD4.memory.activated",
                            "Macrophages.M0",
                            "Macrophages.M1"
                            )))  %>% 
  rename('HAVCR2'='TIM3',
         'CD274'='PDL1',
         'LYMLE'='Lymphocytes',
         'LGALS9' = 'Galactin9',
         'PDCD1'= 'PD1'
         )

cor_method = 'spearman'
ICvsCNIH4_mat <- cor(ICvsCNIH4_df,
                      use='pairwise.complete.obs',
                      method = cor_method)

cor.pval.test <- function(x1, x2, ...) {
  r <- ncol(x1)
  n <- ncol(x2)
  p.mat<- matrix(NA, r, n)
  for (i in 1:r) {
    for (j in 1:n) {
      tmp <- cor.test(x1[, i], x2[, j], ...)
      p.mat[i, j] <- tmp$p.value
    }
  }
  rownames(p.mat) <-  colnames(x1)
  colnames(p.mat) <- colnames(x2)
  return(p.mat)
}

p.ICvsCNIH4_mat <- cor.pval.test(ICvsCNIH4_df, 
                                  ICvsCNIH4_df, 
                                  method=cor_method, 
                                  use='pairwise.complete.obs')
col=c("#001164", 'white', "#FF681E")

# corrplot(ICvsCNIH4_mat, 
#          type="upper", 
#          order="hclust", 
#          tl.cex = 0.5, 
#          col=colorRampPalette(c("blue4", "white", "firebrick1"))(100), 
#          p.mat = p.ICvsCNIH4_mat, 
#          sig.level = 0.05, 
#          insig = "blank") + 
#   theme(axis.text.y = element_text(face="bold", colour="black", size=12),
#             axis.text.x = element_text(face="bold", colour="black", size=12),
#             legend.text = element_text(size = 15),
#             legend.title = element_text( colour="black", size=12))+
#       theme(legend.position="right")
col_list <- list()
for(i in 1:length(names(ICvsCNIH4_df))){
  if(names(ICvsCNIH4_df)[i]=='CNIH4'){
    print(i)
    col_list <- append(col_list, 'red')
  }
  else{
    col_list <- append(col_list, 'black')
  }
}
col_list[28]<-'black'
col_list[30]<-'red'

p2 <- ggcorrplot(ICvsCNIH4_mat,
                 p.mat = p.ICvsCNIH4_mat,
                 hc.order = TRUE,
                 type="full", 
                 color=col,
                 outline.color = 'black',
                 lab = T,
                 lab_size = 2,
                 insig="blank",
                 show.diag = T) +
      scale_fill_gradient2(low=col[1],
                           mid= col[2],
                           high=col[3],
                           na.value = 'gray50',
                           limits=c(-1,1))+
      labs(fill="Corr")+
      theme(#axis.text.y = element_text(face="bold", colour="black", size=12),
            #axis.text.x = element_text(face="bold", colour="black", size=12),
            axis.text.y = element_markdown(face="bold", colour=col_list, size=12),
            axis.text.x = element_markdown(face="bold", colour=col_list, size=12),
            legend.text = element_text(size = 15),
            legend.title = element_text( colour="black", size=12))+
      theme(legend.position="right")


#png(file=paste0(path_output,"/Figure_correlation_CNIH4vsICs_new.png"),res=600, units = 'in', width = 12, height =12)
p2
#dev.off()
# grid.newpage()
```

```{r}
#Re-draw
library(pheatmap)
#library(heatmaply)
library(ggplot2)
#library(tidyverse)
library(dplyr)
library(stats)
library(corrplot) 
library(ggtext)
library(grid)
library(ggplotify)
library(patchwork)
#library(egg)

impute_df <-  read.csv(paste0(path_input,'/Raw_data/CIBERSORTx_Adjusted_CPM_LM22.txt'),sep='\t') %>% filter(P.value<0.05)
rownames(impute_df) <- impute_df$Mixture
impute_df$Mixture <- NULL

IC_names<-c('CD200', 'TNFSF4', 'CD200R1', 'IDO1','TNFRSF18', 'TNFSF18',
             'CD28','BTLA', 'CD27', 'HAVCR2', 'CD274', 'TNFSF14', 'TNFRSF4',
             'CD276', 'TNFRSF14', 'PDCD1', 'CD86', 'TNFSF9', 'TIGIT','CD80',
             'ICOS', 'TNFRSF9', 'CTLA4', 'LAG3','CNIH4')
ICvsCNIH4_df <- vst_data %>% select(IC_names)

ICvsCNIH4_df <- merge(vst_data, impute_df, by=0) %>% 
  select(append(IC_names,c('LGALS9','CD4',
                           'LYMLE',
                           'CT_mean',
                           "Neutrophils",
                            "Monocytes",
                            "T.cells.CD8",
                            "T.cells.CD4.naive",
                            "T.cells.CD4.memory.resting",
                            "T.cells.CD4.memory.activated",
                            "Macrophages.M0",
                            "Macrophages.M1"
                            )))  %>% 
  rename('TIM3'='HAVCR2',
         'PDL1'='CD274',
         'Lymphocytes'='LYMLE',
         'Galactin9'='LGALS9',
         'PD1'='PDCD1'
         )

cor_method = 'spearman'
ICvsCNIH4_mat <- cor(ICvsCNIH4_df,
                      use='pairwise.complete.obs',
                      method = cor_method)
p.ICvsCNIH4_mat <- cor.pval.test(ICvsCNIH4_df, 
                                  ICvsCNIH4_df, 
                                  method=cor_method, 
                                  use='pairwise.complete.obs')

col=c("#001164", 'white', "#FF681E")
#col=c("#5E81AC", 'white', "#D08770")
#col=c("#5E81AC", '#EBCB8B', "#BF616A")
##################################################################################################

# ggcorrplot(ICvsCNIH4_mat,
#            p.mat = p.ICvsCNIH4_mat,
#            hc.order = TRUE,
#            type="full", 
#            color=col,
#            hc.method = "ward.D2",
#            outline.color = 'black',
#            lab = T,
#            lab_size = 2,
#            insig="blank",
#            show.diag = T) +
#   scale_fill_gradient2(low=col[1],
#                            mid= col[2],
#                            high=col[3],
#                            na.value = 'gray50',
#                            limits=c(-1,1))+
#       labs(fill="Corr")+
#       theme(axis.text.y = element_text(face="bold", colour="black", size=12),
#             axis.text.x = element_text(face="bold", colour="black", size=12),
#             legend.text = element_text(size = 15),
#             legend.title = element_text( colour="black", size=12))+
#       theme(legend.position="right")

##################################################################################################
# corrplot(ICvsCNIH4_mat,
#          type="full",
#          order="hclust",
#          tl.cex = 0.5,
#          col=colorRampPalette(col)(100),
#          p.mat = p.ICvsCNIH4_mat,
#          sig.level = 0.05,
#          insig = "blank") +
#   theme(axis.text.y = element_text(face="bold", colour="black", size=12),
#         axis.text.x = element_text(face="bold", colour="black", size=12),
#         legend.text = element_text(size = 15),
#         legend.title = element_text( colour="black", size=12))+
#       theme(legend.position="right") + theme_bw()

##################################################################################################
# draw_colnames_45 <- function (coln, gaps, ...) {
#     coord <- pheatmap:::find_coordinates(length(coln), gaps)
#     x     <-  coord$coord - 0.5 * coord$size
#     res   <- grid::textGrob(
#       coln, x = x, y = unit(1, "npc") - unit(3,"bigpts"),
#       vjust = 0.75, hjust = 1, rot = 45, gp = grid::gpar(...)
#     )
#     return(res)
# }
# # 
# assignInNamespace(
#   x = "draw_colnames",
#   value = "draw_colnames_45",
#   ns = asNamespace("pheatmap")
# )

# CNIH4 vs 24 ICs
mat1 <- ICvsCNIH4_mat[1:25,1:25]
p.mat1 <- p.ICvsCNIH4_mat[1:25,1:25]
labels1 <-round(mat1,2)
vals1 <- labels1
vals1[p.mat1 > 0.05] <- 0
labels1[p.mat1 > 0.05] <- ''

myBreaks <- c(seq(-1, 0, length.out=ceiling(1000/2) + 1),
              seq(1/1000, 1, length.out=floor(1000/2)))

p1 <- pheatmap(vals1,
               color = colorRampPalette(col)(1000),
               breaks=myBreaks,
               display_numbers=labels1,
               show_colnames = F,legend = F)
row.order <- p1$tree_row$order
col.order <- p1$tree_col$order

# CNIH4 vs other
mat2 <- ICvsCNIH4_mat[26:dim(ICvsCNIH4_mat)[1],1:25]
mat2 <- mat2[,col.order]
p.mat2 <- p.ICvsCNIH4_mat[26:dim(ICvsCNIH4_mat)[1],1:25]
p.mat2 <- p.mat2[,col.order]

labels2 <-round(mat2,2)
vals2 <- labels2
vals2[p.mat2 > 0.05] <- 0
labels2[p.mat2 > 0.05] <- ''


myBreaks <- c(seq(-1, 0, length.out=ceiling(1000/2) + 1),
              seq(1/1000, 1, length.out=floor(1000/2)))

p2 <- pheatmap(vals2,
         color = colorRampPalette(col)(1000),
         breaks=myBreaks,
         display_numbers=labels2,
         cluster_rows=T,
         cluster_cols=F,legend = F)

# grid.arrange(grobs = list(p1[[4]],
#                           p1[[4]]),
#              layout_matrix = rbind(c(1,2),c(1,2)))
# 
# plot_grid(as.ggplot(p1), as.ggplot(p2), nrow = 2, 
#           #rel_heights = c(0.1, 0.6),
#           rel_widths = c(0.7,0.7))
# 
# 
# 
# ggarrange(as.ggplot(p1), as.ggplot(p2), ncol=1)
# 
# grid.newpage()
# grid.draw(rbind(ggplotGrob(as.ggplot(p1)),
#                  ggplotGrob(as.ggplot(p2)),
#                  size = "last"))
# 
# pl <- align_plots(as.ggplot(p1), as.ggplot(p2), align="v")
# 
# ggarrange(as.ggplot(p1), as.ggplot(p2), ncol=1)
# 
# ggdraw(pl[[1]])
# ggdraw(pl[[2]])
# 
# plot_grid(as.ggplot(p1), as.ggplot(p1),as.ggplot(p2), align='v',nrow = 2, 
#           rel_heights = c(1/4, 1/4)
#           #rel_widths = c(1/4, 1/4)
#           )
# 
# 
# 
# lm <- rbind(c(0.2,0.2),c(10,10))
# 
# grid.arrange(arrangeGrob(as.ggplot(p1),
#                          as.ggplot(p2),
#                          ncol=2, nrow=1),
#              arrangeGrob(as.ggplot(p1), nrow=1),
#              heights=c(5,1), widths=c(1,1))
# 
# png(file=paste0(path_output,"/tmp.png"),res=600, units = 'in', width = 15, height =6.5)
# grid.arrange(as.ggplot(p1),as.ggplot(p2), ncol=1, nrow =2)
# dev.off()
# 
# 
# 
# 
# 
# grid.arrange(arrangeGrob(as.ggplot(p1),as.ggplot(p2),ncol=2, nrow =1, widths = c(6.1,1)),
#              as.ggplot(p2), nrow = 2,ncol=1)
# 
# 
# 
# 
# heights = c(1, 1)
# as.ggplot(p1)+as.ggplot(p1)/as.ggplot(p2) + plot_layout(widths = c(0.1, 5))
# 
# grid.arrange(as.ggplot(p1),as.ggplot(p2),
#              as.ggplot(p2),ncol=2, widths = unit(20, c("cm")), heights = c(unit(10, c("cm")), unit(10, c("cm"))))
# 
# library(cowplot)
# combined_plot <- plot_grid(as.ggplot(p1), as.ggplot(p2), ncol = 1, align = "v")
# plot(combined_plot)
```

```{r}
temp<-vst_data
temp<- temp[, !duplicated(colnames(temp))]
ggplot(data=temp, aes(CNIH4,CD274))+
  geom_point(size=2.5,alpha=0.5)+
  geom_smooth(method=lm,
              se = TRUE,
              alpha = 0.1,
              size=2,
              fullrange=TRUE) +theme_light()
```

```{r}
dim(tmp)
```

```{r}
library(pheatmap)
myColor <- colorRampPalette(c("#001164", 'white', "#FF681E"))(28)

pheatmap(ICGvsCNIH4_mat,  
         color = myColor, 
         #annotation_row = met_annotation, 
         #annotation_colors = anno_col, 
         #main = "Vietnam",
         show_rownames=T,
         show_colnames=T,
         annotation_names_row=T,
         legend = T,
         annotation_legend = T,annotation=T)
```

```{r}

dist_matrix <- dist(abs(ICGvsCNIH4_mat), method="euclidean")

hc <- hclust(dist_matrix, method="ward.D")
hcd <- as.dendrogram(hc)
# Default plot
library(ggdendro)

p1<-ggdendrogram(hc, rotate =T, theme_dendro = TRUE,size = 3)+ylab("")+xlab("") +
    theme_bw()+ theme(legend.position="bottom")+
  theme(panel.grid.minor = element_blank(),axis.title.x = element_blank(),
        axis.title.y = element_text(face="bold", colour="black", size=12),
        axis.text.y = element_text(face="bold",colour="black", size=13),
        axis.text.x = element_blank(),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=12),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=12)) + 
  theme(plot.title = element_text(hjust = 0.5,face="bold", colour="black", size=20))
p1
```

```{r}
read.csv('C:/Users/triettm/OneDrive - Oxford University Clinical Research Unit/#OUCRU-TB/#oucru-tb/PTB_project/Raw_data/CIBERSORTx_Adjusted_CPM_LM22.txt',sep='\t')
```
