---
title: "Triet's works on PTB"
author: "Triet"
date: "2024-01-08"
output: html_document
---

## Import Libraries and Data

```{r setup, include=FALSE}
#knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(
	echo = TRUE,
	message = FALSE,
	warning = FALSE)
library(parallel)
mcore<-1 # for Window PC

#Load libraries
library(dplyr)
library(readxl)
library(writexl)
library(survival)
library(glmnet)
library(stabs)
library(tidyverse)
library(magrittr)
library(Hmisc)
library(gtsummary)
library(WGCNA)
library(ggcorrplot)
library(scales)
library(gridExtra)
library(ggcorrplot)
library(broom)
library(rms)
library(tidyverse)
library("stabs")
library("mboost")
library(DESeq2)
library(CEMiTool)

#path_output<- paste0(dirname(getwd()),"/Triet_Results") #local oucru
path_output<-paste0(str_replace_all(Sys.getenv("USERPROFILE"), '\\\\', '/'),
                     "/OneDrive - Oxford University Clinical Research Unit/#OUCRU-TB/#oucru-tb/PTB_project/Triet_Results")
path_input <- paste0(str_replace_all(Sys.getenv("USERPROFILE"), '\\\\', '/'),
                     "/OneDrive - Oxford University Clinical Research Unit/#OUCRU-TB/#oucru-tb/PTB_project/")

###------Import clinical data------###
load(paste0(path_input,"/Raw_data/vst_Gensymbol_allPTB_295Sample_18Feb22.RData"))

# clinical <-vst_data%>% select("Patient_ID","Age","SEX","BMI","Timika_score_CXR","Diabetes","Cavity_CXR","Lung_proportion_CXR","CT_mean","Timepoint","GeneXpert","HIV","trial_number","LIMS_ID") %>% rename("SEX"="Sex","trial_number"="Treatment_sensitive")

load(paste0(path_input,"/Raw_data/vst_Gensymbol_allTBM_281Sample_12Jan24.RData"))
rownames(tbm_vst) <- tbm_vst$LIMS_ID
tbm_vst %<>% mutate(Neutrophil_counts=WBC*NEUTLE*0.01,
                     log2_Neutrophil_counts=log2(Neutrophil_counts),
                     Lymphocyte_counts=WBC*LYMLE *0.01,
                     log2_Lymphocyte_counts=log2(Lymphocyte_counts),
                     Monocyte_counts=WBC*MONOLE*0.01,
                     log2_Monocyte_counts=log2(Monocyte_counts),
                     Eosinophil_counts=WBC*EOSLE*0.01,
                     log2_Eosinophil_counts=log2(Eosinophil_counts))


# clinical_tbm <-tbm_vst%>% select("Age","SEX","BMI","Timika_score_CXR","Diabetes","Cavity_CXR","Lung_proportion_CXR","CT_mean","Timepoint","GeneXpert","HIV","trial_number","LIMS_ID") %>% rename("SEX"="Sex","trial_number"="Treatment_sensitive")

###------ImputeCIBERSORT------###
impute_df <-  read.csv(paste0(path_input,'/Raw_data/CIBERSORTx_Adjusted_CPM_LM22.txt'),sep='\t') %>% filter(P.value<0.05)
rownames(impute_df) <- impute_df$Mixture
impute_df$Mixture <- NULL
###------AllData-----###
load(file=paste0(path_input,"Raw_data/Data_RNAseq_RUN1andRUN2_15Feb2023.RData"))
load(file=paste0(path_input,"Raw_data/count_data_adjusted_ComBat.Rdata"))
```

```{r eval=FALSE, include=FALSE}
# Import raw RNA_seq data, do the do the stabilizing variance transformation, and select the top 10000 with high variances genes
path_input_RNA <- paste0(path_input, "Raw_data/Data_RNAseq_RUN1andRUN2_15Feb2023.RData")
lnames <- load(path_input_RNA)

## load clinical data from the meta data
All_sample_RNAseq <- readxl::read_excel(paste0(path_input,"Raw_data/All_sample_RNAseq.xlsx"),sheet = "Run1and2") %>% as.data.frame()
rownames(All_sample_RNAseq) <- All_sample_RNAseq$LIMS_ID

## check clinical data and count data 
all(rownames(All_sample_RNAseq) %in% colnames(count_all))
dim(All_sample_RNAseq)## 1276 samples

## Extract only the data for PTB cohort (28,29TB)
PTB_sample_RNAseq <- All_sample_RNAseq %>% subset(Study %in% c("28TB","29TB"))
count_data <-count_all[,rownames(PTB_sample_RNAseq)]

## create anno data
anno_symbol <- anno[,c("gene_id", "gene_symbol")] %>% filter(gene_id %in% rownames(count_data))
rownames(anno_symbol) <- anno_symbol$gene_id
anno_symbol <- anno_symbol[rownames(count_data),]
## add gene symbol to count data
count_data <- cbind(genenames=anno_symbol$gene_symbol,count_data)
## Some genes have multiple segments that map to one gene. Thus, the counts of that gene is the sum of all the segments counts of that genes
count_deseq2 <- count_data %>%
                group_by(genenames) %>%
                summarise_all(sum) %>%
                data.frame()
rownames(count_deseq2) <- count_deseq2$genenames

count_deseq2 %<>% dplyr::select(-genenames)

## filter low expressed genes
filter_min_group <- ncol(count_deseq2)/5
## filter for gene expression of 10 counts in more than 20% sample
filter_gene <- apply(count_deseq2, 1, function(x) length(x[x>10]) >= filter_min_group) 
count_deseq2 <- count_deseq2[filter_gene,] %>% as.matrix()

## normalize vst by deseq2
PTB_sample_RNAseq %<>% mutate(Study=factor(Study))
dds <- DESeqDataSetFromMatrix(count_deseq2, colData=PTB_sample_RNAseq, ~ Study)
# rownames(dds)
# dds_anno <- anno_gene_symbol(data = dds,data_anno = anno,export = F,name = NULL)
vsd <- assay(vst(dds, blind=T)) # Variance stablized transform with log transform
vsd_t<-vsd %>% t() %>%
 as.data.frame() %>%
 tibble::rownames_to_column(var='LIMS_ID') %>%
 as.data.frame()

RNA_data<-vsd_t %>% select(-LIMS_ID)
rownames(RNA_data)<-vsd_t$LIMS_ID

## Method 1:  Using variance_based filtering method to filter gene most variance

RNA_data_transpose <- as.data.frame(t(RNA_data))
RNA_data_transpose_filtered <- filter_genes(RNA_data_transpose,pct =.75) #.75
# Selected genes
selected <- select_genes(RNA_data_transpose_filtered,filter_pval=0.35) #0.35

# ## Method 2:  Using the top 10000 genes as the most variance
# dv <- NULL
# dv$Symbols <- colnames(RNA_data)
# dv$variance <- matrixStats::colVars(as.matrix(RNA_data))
# dv <- as.data.frame(dv)
# ## sorting gene by var
# dv <- dv[order(dv$variance, decreasing = TRUE),]
# 
# 
# RNA_data$LIMS_ID<-vsd_t$LIMS_ID
# selected <- dv$Symbols[1:10000] #method 2 choose 10000 genes most variant

RNA_data <- RNA_data[,selected]
RNA_data$LIMS_ID<-vsd_t$LIMS_ID

####----Check the level of expression between two batches----###
expr_normalized_df <- data.frame(RNA_data) %>%
  pivot_longer(cols=RN7SL1:ZNF876P)
expr_normalized_df<-plyr::join_all(list(expr_normalized_df,PTB_sample_RNAseq)) %>% 
                    mutate(seq_Batch=factor(seq_Batch))


expr_normalized_df %>% ggplot(., aes(x = seq_Batch, y = value)) +
  geom_violin() +
  geom_jitter(width = 0.1,alpha=0.1,size=1) +
  theme_bw() +
  theme(
    axis.text.x = element_text( angle = 90)
  ) +
  ylim(0, NA) +
  labs(
    title = "Normalized and 95 quantile Expression",
    x = "treatment",
    y = "normalized expression"
  )+ theme(plot.title = element_text(hjust = 0.5))->p

# png(file=paste0(path_output,"/Gene_expression_after_normalized_by_study.png"),
#     res=400, units = 'in', width = 10, height =5)
# p
# dev.off()

####--------------------------------------------------------------------####

gene_names_full<-colnames(RNA_data)
gene_names_full<-gsub("-","_",gene_names_full)
colnames(RNA_data)<-gene_names_full

###---- Import PCR Fluidigm data ----###
path_input_Fluidigm <-  paste0(path_input, "Raw_data/Fluidigm_PTB_wide_2Jan2023.RData")
load(path_input_Fluidigm)
Fluidigm_PTB_wide %<>% as.data.frame() %>% dplyr::rename("Patient_ID"="studycode")
Patient_Fluidigm_PTB<-Fluidigm_PTB_wide$Patient_ID
###----------------------------------###

## Import clinical data for RNA_seq and for PCR data
path_input_clinical<- paste0(path_input, "Raw_data/Clinical full data 913 patients.csv")
clinical <- read_csv(path_input_clinical)%>% as.data.frame()

clinical %>% select("studycode","DiagnosisAge","Gender","bmi","Timika.score","Diabetes","Cavity","Lung.proportion","CT_Xpert_mean","LIMS_ID","HGB","WBC","NEUTLE","LYMLE","MONOLE","EOSLE","PLAT","DUR","study","TBTreatedBefore","GenXpert.f")

####-----Clinical data---####
clinical_2<-clinical %>% select("studycode","DiagnosisAge","Gender","bmi","Timika.score","Diabetes","Cavity","Lung.proportion","CT_Xpert_mean","LIMS_ID","HGB","WBC","NEUTLE","LYMLE","MONOLE","EOSLE","PLAT","DUR","study","TBTreatedBefore","GenXpert.f") %>% 
  mutate("Treatment_sensitive"=ifelse(study=="28TB","MDR","DS")) %>% 
  dplyr::rename("Patient_ID"="studycode",
                "Neutrophil"="NEUTLE",
                "Lymphocyte"="LYMLE",
                "Monocyte"="MONOLE",
                "Eosinophil"="EOSLE",
                "Platelet"="PLAT",
                "Symptom_duration"="DUR",
                "Lung_proportion"="Lung.proportion",
                "Age"="DiagnosisAge",
                "Timika_score"="Timika.score",
                "TB_History"="TBTreatedBefore",
                "GenXpert_catgry"="GenXpert.f"
                )

clinical_2 %<>% mutate(Neutrophil_counts=WBC*Neutrophil*0.01,
                     log2_Neutrophil_counts=log2(Neutrophil_counts),
                     Lymphocyte_counts=WBC*Lymphocyte *0.01,
                     log2_Lymphocyte_counts=log2(Lymphocyte_counts),
                     Monocyte_counts=WBC*Monocyte*0.01,
                     log2_Monocyte_counts=log2(Monocyte_counts),
                     Eosinophil_counts=WBC*Eosinophil*0.01,
                     log2_Eosinophil_counts=log2(Eosinophil_counts),
                     log2_Symptom_duration=log2(Symptom_duration+1),
                     log2_Platelet_counts =log2(Platelet),
                     log2_CT=log2(CT_Xpert_mean))

# Remove one patient with RNA_seq quality low
clinical_2%<>% filter(!is.na(LIMS_ID)&LIMS_ID!="HOA7818A92"|Patient_ID%in%Patient_Fluidigm_PTB)

clinical_RNA_PTB <- clinical_2 %>% filter(!is.na(LIMS_ID)) %>% as.data.frame()

clinical_Fluidigm_PTB <- clinical_2 %>% filter(Patient_ID%in%Patient_Fluidigm_PTB) %>% as.data.frame()

# Here we only combine the 295 patients who have good RNA_seq quality
clinical_RNA<-merge(RNA_data,clinical_RNA_PTB,by="LIMS_ID")
RNA_data<-clinical_RNA[,gene_names_full]
rownames(RNA_data)<-rownames(clinical_RNA)<-clinical_RNA$LIMS_ID

# Here we only combine the 190 patients who have good PCR Fluidigm quality
clinical_Fluidigm<-merge(Fluidigm_PTB_wide,clinical_Fluidigm_PTB,by="Patient_ID")

clinical_2 %<>% mutate(cohort=ifelse(Patient_ID%in%Patient_Fluidigm_PTB,"Validation","Discovery"))

clinical_2_discovery <- filter(clinical_2,cohort=='Discovery')
rownames(clinical_2_discovery) <- clinical_2_discovery$LIMS_ID

clinical_2_pcr <- filter(clinical_2,cohort=='Validation')

# save(Fluidigm_PTB_wide,clinical_2,clinical_Fluidigm,file="Data/Fluidigm_PTB_data.Rdata")
# save(RNA_data,clinical_RNA,clinical_2,file="Data/Most_variance_RNA_genes.Rdata")
```

## Split TBM and PTB and PCR

```{r}
library(brms)
library(cmdstanr)
library(posterior)
library(bayesplot)
library(tidybayes)
library(broom)
library(dplyr)
library(ggplot2)
library(broom.mixed)

options(mc.cores = 1, brms.backend = "cmdstanr")

CHAINS <- 4
ITER <- 2000
WARMUP <- 1000
BAYES_SEED <- 1234

# Create a custom family that is logit if y = 0, normal/gaussian if not
hurdle_gaussian <- custom_family("hurdle_gaussian", 
                dpars = c("mu", "sigma", "hu"),
                links = c("identity", "log", "logit"),
                lb = c(NA, 0, NA),
                type = "real")

# Stan code
stan_funs <- "
  real hurdle_gaussian_lpdf(real y, real mu, real sigma, real hu) { 
    if (y == 0) { 
      return bernoulli_lpmf(1 | hu); 
    } else { 
      return bernoulli_lpmf(0 | hu) +  
             normal_lpdf(y | mu, sigma); 
    } 
  }
"
# Prepare Stan code for use in brm()
stanvars <- stanvar(scode = stan_funs, block = "functions")
set_cmdstan_path('C:/Users/triettm/AppData/Local/R/win-library/4.3/.cmdstan/cmdstan-2.33.1')
```

```{r}
Fludigm_CNIH4_PTB <- get(load(paste0(path_input,'Raw_data/Fludigm_CNIH4_PTB.Rdata')))
dim(Fludigm_CNIH4_PTB)# 171  12
hist(Fludigm_CNIH4_PTB$CT_Xpert_mean)

Fludigm_CNIH4_PTB %<>% mutate(
  bacterial_load_1 = log2(40/CT_Xpert_mean)
)

ggplot( Fludigm_CNIH4_PTB,aes(x = bacterial_load_1, y = CT_Xpert_mean))+
  geom_point()

hist(Fludigm_CNIH4_PTB$bacterial_load_1)

Hmisc::describe(Fludigm_CNIH4_PTB)


hist(Fludigm_CNIH4_PTB$log2_CT)
colnames(Fludigm_CNIH4_PTB)


model_hurdle_CNIH4 <- brm(
  bf(bacterial_load_1 ~ CNIH4_Norm,
     hu ~ CNIH4_Norm),
  data = Fludigm_CNIH4_PTB,
  family = hurdle_gaussian,  # <--- This is new
  stanvars = stanvars,   # <--- This is new
  chains = CHAINS, iter = ITER, warmup = WARMUP, seed = BAYES_SEED,
  silent = 2
)

tidy(model_hurdle_CNIH4)
# non-hurdle: 0.103 [0.0249,0.183], p = 0.0090
# hurdle: -0.130, [-1.23, 0.955], p = 0.8085
p_value(model_hurdle_CNIH4) # 

pp_check(model_hurdle_CNIH4, ndraws = 500)

c_CNIH4 <- cbind.data.frame(
  data.frame(tidy(model_hurdle_CNIH4))[1:4,4:8],
  p.value=p_value(model_hurdle_CNIH4)[,2])

colnames(c_CNIH4)<-c("Parameter","Estimate","std.error","conf.low","conf.high","p_value")

summary(Fludigm_CNIH4_PTB$CNIH4_Norm)

newdata=data.frame(CNIH4_Norm=seq(2,6.5,len=50))

pred_CNIH4<-fitted(model_hurdle_CNIH4, newdata = newdata)

d_pred_CNIH4<-cbind(newdata,pred_CNIH4)


d_pred_CNIH4<-d_pred_CNIH4 %>% mutate(CT_pred=40/2^(Estimate),
              lower_pred=40/2^(Q2.5),
              upper_pred=40/2^(Q97.5))
dim(d_pred_CNIH4)# 50  8
summary(d_pred_CNIH4$CNIH4_Norm)
summary(d_pred_CNIH4$CT_pred)

p_CNIH4<-ggplot(data=d_pred_CNIH4,aes(x=CNIH4_Norm,y=CT_pred))+
  geom_line(size=2,col="#313695")+
  geom_ribbon(aes(ymin=lower_pred,ymax=upper_pred,x=CNIH4_Norm),alpha=0.4,fill="#313695")+
  scale_y_continuous(trans = log2_trans(),
    breaks = (seq(10,40,by=5)),
    labels = seq(10,40,by=5))+
  
  geom_jitter(data=Fludigm_CNIH4_PTB,aes(x = CNIH4_Norm, y = CT_Xpert_mean  ), size=2.5,alpha=0.5)+
  theme_bw()+
  #labs(x = paste("CNIH4 Norm"," expression level [log2 counts] "), y = "Ct value")+
  labs(x = paste("CNIH4 Norm"), y = "Ct value")+
  theme(panel.grid.minor = element_blank(),axis.title.x = element_text(face="bold", colour="black", size=20),
        axis.title.y = element_text(face="bold", colour="black", size=20),
        axis.text.y = element_text( colour="black", size=15),
        axis.text.x = element_text( colour="black", size=15),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=15),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=15))+coord_cartesian(ylim=c(10,40))
p_CNIH4
```

```{r}
library(rstatix)
library(brms)
library(parameters)

posterior_predict_hurdle_gaussian <- function(i, prep, ...) {
  mu <- brms::get_dpar(prep, "mu", i = i)
  sigma <- brms::get_dpar(prep, "sigma", i = i)
  theta <- brms::get_dpar(prep, "hu", i = i)
     
  hu <- runif(prep$ndraws, 0, 1)
  ifelse(hu < theta, 0, rnorm(prep$ndraws, mu,sigma))
}
   
posterior_epred_hurdle_gaussian <- function(prep) {
  with(prep$dpars, mu * (1 - hu))
}

load(file=paste0(path_input,"Raw_data/Data_26_27_CT_CNIH4.Rdata.RData"))
dim(Data_26_27_CT)# 254,  15

model_hurdle <- brm(bf(bacterial_load_1 ~ CNIH4,
                       hu ~ CNIH4),
                    data = Data_26_27_CT,
                    family = hurdle_gaussian,  # <--- This is new
                    stanvars = stanvars,   # <--- This is new
                    chains = CHAINS, 
                    iter = ITER, 
                    warmup = WARMUP, 
                    seed = BAYES_SEED,
                    silent = 2
)

tidy(model_hurdle)
pp_check(model_hurdle, ndraws = 500)
c <- cbind.data.frame(data.frame(tidy(model_hurdle))[1:4,4:8],p.value=p_value(model_hurdle)[,2])
colnames(c)<-c("Parameter","Estimate","std.error","conf.low","conf.high","p_value")

summary(Data_26_27_CT$CNIH4)
newdata=data.frame(CNIH4=seq(8.753,11.771,len=50))
pred<-fitted(model_hurdle, newdata = newdata)
d<-cbind(newdata,pred)
d<-d %>% mutate(CT_pred=40/2^(Estimate),
            lower_pred=40/2^(Q2.5),
            upper_pred=40/2^(Q97.5))

p1<-ggplot(data=d,aes(x=CNIH4,y=CT_pred))+
  geom_line(size=2,col="#313695")+
  geom_ribbon(aes(ymin=lower_pred,ymax=upper_pred,x=CNIH4),alpha=0.4,fill="#313695")+
  geom_jitter(data=Data_26_27_CT,aes(x = CNIH4, y = NAT), size=2.5,alpha=0.5)+
  theme_bw()+
  labs(x = paste("CNIH4"," expression level [log2 counts] "), y = "Mtb Xpert Ct value")+ 
  theme(panel.grid.minor = element_blank(), axis.title.x = element_text(face="bold", colour="black", size=20),
        axis.title.y = element_text(face="bold", colour="black", size=20),
        axis.text.y = element_text( colour="black", size=15),
        axis.text.x = element_text( colour="black", size=15),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=15),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=15))+
  scale_y_continuous(trans = log2_trans(),
    breaks = log2(seq(10,40,by=5)),
    labels = seq(10,40,by=5))
p1
png(file=paste0(path_output,"/TBM_CNIH4.png"),res=450, units = 'in', width =10, height =10)
p1
dev.off()

#############################################################################################################################
# PTB

load(paste0(path_input,'Raw_data/d2_long.Rdata'))
d2_long <- d2_long %>% filter(gene_names=="CNIH4")
p2<-ggplot(data=d2_long,aes(x=concentration ,y=log2_CT))+
  geom_point(size=2.5,alpha=0.5)+
  geom_smooth(method = "gam",formula = y ~ s(x),se = TRUE,size=2,col="#313695",fill="#313695")+
  theme_bw()+
  labs(x = paste("CNIH4"," expression level [log2 counts] "), y = "Mtb Xpert Ct value")+ 
  theme(panel.grid.minor = element_blank(),axis.title.x = element_text(face="bold", colour="black", size=20),
        axis.title.y = element_text(face="bold", colour="black", size=20),
        axis.text.y = element_text( colour="black", size=15),
        axis.text.x = element_text( colour="black", size=15),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=15),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=15))+
  scale_y_continuous(trans = log2_trans(),
                     breaks = log2(seq(10,40,by=5)),
                     labels = seq(10,40,by=5))
p2
png(file=paste0(path_output,"/PTB_CNIH4.png"),res=450, units = 'in', width =10, height =10)
p2
dev.off()
```

```{r}
View(clinical_2_pcr)
View(Data_26_27_CT)
```

```{r}
rownames(clinical_2_pcr) <- clinical_2_pcr$Patient_ID
rownames(clinical_Fluidigm) <- clinical_Fluidigm$Patient_ID
Fludigm_CNIH4_PTB <- merge(clinical_Fluidigm,clinical_2_pcr,by=0)
View(Fludigm_CNIH4_PTB)
```

```{r}
# PCR
# Fludigm_CNIH4_PTB <- get(load(paste0(path_input,"Raw_data/Fludigm_CNIH4_PTB.Rdata")))
# dim(Fludigm_CNIH4_PTB)# 171  12
# hist(Fludigm_CNIH4_PTB$CT_Xpert_mean)
# 
# Fludigm_CNIH4_PTB %<>% mutate(
#   bacterial_load_1 = log2(40/CT_Xpert_mean)
# )

Fludigm_CNIH4_PTB <- clinical_Fluidigm
#Fludigm_CNIH4_PTB <- filter(Fludigm_CNIH4_PTB,CT_Xpert_mean!=40)
Fludigm_CNIH4_PTB %<>% mutate(
  bacterial_load_1 = log2(40/CT_Xpert_mean)
)

ggplot( Fludigm_CNIH4_PTB,aes(x = bacterial_load_1, y = CT_Xpert_mean))+
  geom_point()

hist(Fludigm_CNIH4_PTB$bacterial_load_1)

Hmisc::describe(Fludigm_CNIH4_PTB)


hist(Fludigm_CNIH4_PTB$log2_CT)
colnames(Fludigm_CNIH4_PTB)

model_hurdle_CNIH4 <- brm(
  bf(bacterial_load_1 ~ CNIH4,
     hu ~ CNIH4),
  data = Fludigm_CNIH4_PTB,
  family = hurdle_gaussian,  # <--- This is new
  stanvars = stanvars,   # <--- This is new
  chains = CHAINS, iter = ITER, warmup = WARMUP, seed = BAYES_SEED,
  silent = 2
)

tidy(model_hurdle_CNIH4)
# non-hurdle: 0.103 [0.0249,0.183], p = 0.0090
# hurdle: -0.130, [-1.23, 0.955], p = 0.8085
p_value(model_hurdle_CNIH4) # 

pp_check(model_hurdle_CNIH4, ndraws = 500)

c_CNIH4 <- cbind.data.frame(
  data.frame(tidy(model_hurdle_CNIH4))[1:4,4:8],
  p.value=p_value(model_hurdle_CNIH4)[,2])

colnames(c_CNIH4)<-c("Parameter","Estimate","std.error","conf.low","conf.high","p_value")

summary(Fludigm_CNIH4_PTB$CNIH4)

newdata=data.frame(CNIH4=seq(9.943,19.462,len=50))

pred_CNIH4<-fitted(model_hurdle_CNIH4, newdata = newdata)

d_pred_CNIH4<-cbind(newdata,pred_CNIH4)


d_pred_CNIH4<-d_pred_CNIH4 %>% mutate(CT_pred=40/2^(Estimate),
              lower_pred=40/2^(Q2.5),
              upper_pred=40/2^(Q97.5))

dim(d_pred_CNIH4)# 50  8
summary(d_pred_CNIH4$CNIH4)
summary(d_pred_CNIH4$CT_pred)

p3<-ggplot(data=d_pred_CNIH4,aes(x=CNIH4,y=CT_pred))+
  geom_line(size=2,col="#313695")+
  geom_ribbon(aes(ymin=lower_pred,ymax=upper_pred,x=CNIH4),alpha=0.4,fill="#313695")+
  geom_jitter(data=Fludigm_CNIH4_PTB,aes(x = CNIH4, y = CT_Xpert_mean), size=2.5,alpha=0.5)+
  theme_bw()+
  labs(x = paste("CNIH4"," expression level [log2 counts] "), y = "Mtb Xpert Ct value")+
  theme(panel.grid.minor = element_blank(),axis.title.x = element_text(face="bold", colour="black", size=20),
        axis.title.y = element_text(face="bold", colour="black", size=20),
        axis.text.y = element_text( colour="black", size=15),
        axis.text.x = element_text( colour="black", size=15),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=15),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=15))+
  coord_cartesian(ylim=c(10,40)) +  
  scale_y_continuous(trans = log2_trans(),
    breaks = (seq(10,40,by=5)),
    labels = seq(10,40,by=5))

p3
png(file=paste0(path_output,"/PCR_CNIH4.png"),res=450, units = 'in', width =10, height =10)
p3
dev.off()

png(paste0(path_output,"/PTB_TBM_PCR_CNIH4.png"),res=400, units = 'in', width =21, height =7)
cowplot::plot_grid(p1,p2,p3, nrow = 1,labels = c('PTB', 'TBM','PCR'),align = "v",axis = "tblr")
dev.off()

```

```{r}
Fludigm_CNIH4_PTB <- get(load(file="R:/nguyenlehoaibao/anh_Nhat/Hurdle_26_27/1_data/Fludigm_CNIH4_PTB/Fludigm_CNIH4_PTB.Rdata"))
dim(Fludigm_CNIH4_PTB)# 171  12
hist(Fludigm_CNIH4_PTB$CT_Xpert_mean)

Fludigm_CNIH4_PTB %<>% mutate(
  bacterial_load_1 = log2(40/CT_Xpert_mean)
)

ggplot( Fludigm_CNIH4_PTB,aes(x = bacterial_load_1, y = CT_Xpert_mean))+
  geom_point()

hist(Fludigm_CNIH4_PTB$bacterial_load_1)

Hmisc::describe(Fludigm_CNIH4_PTB)


hist(Fludigm_CNIH4_PTB$log2_CT)
colnames(Fludigm_CNIH4_PTB)


model_hurdle_CNIH4 <- brm(
  bf(bacterial_load_1 ~ CNIH4_Norm,
     hu ~ CNIH4_Norm),
  data = Fludigm_CNIH4_PTB,
  family = hurdle_gaussian,  # <--- This is new
  stanvars = stanvars,   # <--- This is new
  chains = CHAINS, iter = ITER, warmup = WARMUP, seed = BAYES_SEED,
  silent = 2
)

tidy(model_hurdle_CNIH4)
# non-hurdle: 0.103 [0.0249,0.183], p = 0.0090
# hurdle: -0.130, [-1.23, 0.955], p = 0.8085
p_value(model_hurdle_CNIH4) # 
```

## PTB

```{r}

dim(Data_26_27_CT)# 254,  15

model_hurdle <- brm(bf(bacterial_load_1 ~ CNIH4,
                       hu ~ CNIH4),
                    data = Data_26_27_CT,
                    family = hurdle_gaussian,  # <--- This is new
                    stanvars = stanvars,   # <--- This is new
                    chains = CHAINS, 
                    iter = ITER, 
                    warmup = WARMUP, 
                    seed = BAYES_SEED,
                    silent = 1
)

tidy(model_hurdle)
pp_check(model_hurdle, ndraws = 500)
c <- cbind.data.frame(data.frame(tidy(model_hurdle))[1:4,4:8],p.value=p_value(model_hurdle)[,2])
colnames(c)<-c("Parameter","Estimate","std.error","conf.low","conf.high","p_value")


summary(Data_26_27_CT$CNIH4)
newdata=data.frame(CNIH4=seq(8.753,11.771,len=50))
pred<-fitted(model_hurdle, newdata = newdata)
d<-cbind(newdata,pred)
d<-d %>% mutate(CT_pred=40/2^(Estimate),
            lower_pred=40/2^(Q2.5),
            upper_pred=40/2^(Q97.5))

p<-ggplot(data=d,aes(x=CNIH4,y=CT_pred))+
  geom_line(size=2,col="#313695")+
  geom_ribbon(aes(ymin=lower_pred,ymax=upper_pred,x=CNIH4),alpha=0.4,fill="#313695")+
  scale_y_continuous(trans = log2_trans(),
    breaks = log2(seq(10,40,by=5)),
    labels = seq(10,40,by=5))+
  geom_jitter(data=Data_26_27_CT,aes(x = CNIH4, y = NAT), size=2.5,alpha=0.5)+
  theme_bw()+
  labs(x = paste("CNIH4"," expression level [log2 counts] "), y = "Mtb Xpert Ct value")+ 
  theme(panel.grid.minor = element_blank(),axis.title.x = element_text(face="bold", colour="black", size=20),
        axis.title.y = element_text(face="bold", colour="black", size=20),
        axis.text.y = element_text( colour="black", size=15),
        axis.text.x = element_text( colour="black", size=15),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=15),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=15))+scale_y_continuous(trans = log2_trans(),
    breaks = log2(seq(10,40,by=5)),
    labels = seq(10,40,by=5))
p
```

```{r}
load(file=paste0(path_input,"Raw_data/Data_26_27_CT_CNIH4.Rdata.RData")) #TBM
Data_26_27_CT %>% ggplot(aes(x=CNIH4 ,y=Xpert_mean))+
  geom_point(size=2.5,alpha=0.5)+
  geom_smooth(method = "gam",formula = y ~ s(x),se = TRUE,size=2,col="#313695",fill="#313695")+
  theme_bw()+
  labs(x = paste("CNIH4"," expression level [log2 counts] "), y = "Mtb Xpert Ct value")+ 
  theme(panel.grid.minor = element_blank(),
        axis.title.x = element_text(face="bold", colour="black", size=20),
        axis.title.y = element_text(face="bold", colour="black", size=20),
        axis.text.y = element_text( colour="black", size=15),
        axis.text.x = element_text( colour="black", size=15),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=15),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=15)) +xlim(8,13)+ylim(0,40)
+scale_y_continuous(trans = log2_trans())
  scale_y_continuous(trans = log2_trans(),
    breaks = log2(seq(10,40,by=5)),
    labels = seq(10,40,by=5))->p
p
```

```{r}
load(file=paste0(path_input,"Data/Most_variance_RNA_genes.Rdata"))
load(file=paste0(path_input,"Raw_data/Data_26_27_CT_CNIH4.Rdata.RData")) #TBM
load(file=paste0(path_input,"Raw_data/d2_long.Rdata")) #PTB

d2_long %>% filter(gene_names=="CNIH4") %>% ggplot(aes(x=concentration ,y=log2_CT))+
  geom_point(size=2.5,alpha=0.5)+
  geom_smooth(method = "gam",formula = y ~ s(x),se = TRUE,size=2,col="#313695",fill="#313695")+
  theme_bw()+
  labs(x = paste("CNIH4"," expression level [log2 counts] "), y = "Mtb Xpert Ct value")+ 
  theme(panel.grid.minor = element_blank(),axis.title.x = element_text(face="bold", colour="black", size=20),
        axis.title.y = element_text(face="bold", colour="black", size=20),
        axis.text.y = element_text( colour="black", size=15),
        axis.text.x = element_text( colour="black", size=15),
        axis.ticks.x=element_blank(),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=15),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=15))+
  scale_y_continuous(trans = log2_trans(),
    breaks = log2(seq(10,40,by=5)),
    labels = seq(10,40,by=5))->p
png(file=paste0(path_output,"/TBM_CNIH4.png"),res=450, units = 'in', width =10, height =10)
p
dev.off()
```

### 1. Correlation heatmap CNIH4 vs 27 ICs

```{r}
library(pheatmap)
library(ggplot2)
library(ggplotify)
#library(NatParksPalettes)

cor.pval.test <- function(x1, x2, ...) {
  r <- ncol(x1)
  n <- ncol(x2)
  p.mat<- matrix(NA, r, n)
  for (i in 1:r) {
    for (j in 1:n) {
      tmp <- cor.test(x1[, i], x2[, j], ...)
      p.mat[i, j] <- tmp$p.value
    }
  }
  rownames(p.mat) <-  colnames(x1)
  colnames(p.mat) <- colnames(x2)
  return(p.mat)
}

IC_names<-c('CD200', 'CD2','CD58','TNFSF4', 'CD200R1', 'IDO1','TNFRSF18', 'TNFSF18',
             'CD28','BTLA', 'CD27', 'HAVCR2', 'CD274', 'TNFSF14', 'TNFRSF4',
             'CD276', 'TNFRSF14', 'PDCD1', 'CD86', 'TNFSF9', 'TIGIT','CD80',
             'ICOS', 'TNFRSF9', 'CTLA4', 'LAG3','LGALS9','CNIH4')

ICvsCNIH4_df <- vst_data %>% select(IC_names)
#ICvsCNIH4_df <- merge(ICvsCNIH4_df, impute_df, by=0) %>% tibble::column_to_rownames('Row.names') %>% select("CD200" : "Neutrophils" )
ICvsCNIH4_df <- ICvsCNIH4_df[vapply(ICvsCNIH4_df, 
                                    function(x) length(unique(x)) > 1, 
                                    logical(1L))] %>% dplyr::rename('HAVCR2/TIM-3'='HAVCR2',
                                                                    'CD274/PDL1'='CD274',
                                                                    'LGALS9/Galactin-9'='LGALS9',
                                                                    'PDCD1/PD1'='PDCD1')

cor_method = 'spearman'
ICvsCNIH4_mat <- cor(ICvsCNIH4_df,
                      use='pairwise.complete.obs',
                      method = cor_method)

p.ICvsCNIH4_mat <- cor.pval.test(ICvsCNIH4_df, 
                                  ICvsCNIH4_df, 
                                  method=cor_method, 
                                  use='pairwise.complete.obs')

#col=c("#001164", 'white', "#FF681E")
#col=c("#5E81AC", '#EBCB8B', "#BF616A")
col=c("#395b85", 'white', "#FF681E")

draw_colnames_45 <- function (coln, gaps, ...) {
    coord <- pheatmap:::find_coordinates(length(coln), gaps)
    x     <-  coord$coord - 0.5 * coord$size
    res   <- grid::textGrob(
      coln, x = x, y = unit(1, "npc") - unit(3,"bigpts"),
      vjust = 0.75, hjust = 1, rot = 45, gp = grid::gpar(...)
    )
    return(res)
}

assignInNamespace(
  x = "draw_colnames",
  value = "draw_colnames_45",
  ns = asNamespace("pheatmap")
)

# CNIH4 vs 25 ICs
mat1 <- ICvsCNIH4_mat
p.mat1 <- p.ICvsCNIH4_mat
labels1 <-round(mat1,2)
vals1 <- labels1
vals1[p.mat1 > 0.05] <- 0
labels1[p.mat1 > 0.05] <- ''

myBreaks <- c(seq(-1, 0, length.out=ceiling(1000/2) + 1),
              seq(1/1000, 1, length.out=floor(1000/2)))

p01 <- pheatmap(vals1,
               color = colorRampPalette(col)(1000),
               breaks=myBreaks,
               display_numbers=labels1,
               show_colnames = F,legend = F)
p01
row.order <- p01$tree_row$order
col.order <- p01$tree_col$order
p1 <- as.ggplot(p01)
```

```{r}
Cell_names <- c("Neutrophil_counts","Lymphocyte_counts", "Monocyte_counts", "Eosinophil_counts")
ICvsCNIH4_df <- merge(vst_data,clinical_2_discovery,by=0) %>% select(append(IC_names,Cell_names))
ICvsCNIH4_df <- ICvsCNIH4_df[vapply(ICvsCNIH4_df, 
                                    function(x) length(unique(x)) > 1, 
                                    logical(1L))] %>% dplyr::rename('HAVCR2/TIM-3'='HAVCR2',
                                                                    'CD274/PDL1'='CD274',
                                                                    'LGALS9/Galactin-9'='LGALS9',
                                                                    'PDCD1/PD1'='PDCD1')
cor_method = 'spearman'
ICvsCNIH4_mat <- cor(ICvsCNIH4_df,
                      use='pairwise.complete.obs',
                      method = cor_method)

p.ICvsCNIH4_mat <- cor.pval.test(ICvsCNIH4_df, 
                                  ICvsCNIH4_df, 
                                  method=cor_method, 
                                  use='pairwise.complete.obs')

mat1 <- ICvsCNIH4_mat[29:32,1:28]
mat1 <- mat1[,col.order]
p.mat1 <- p.ICvsCNIH4_mat[29:32,1:28]
p.mat1 <- p.mat1[,col.order]

labels1 <-round(mat1,2)
vals1 <- labels1
vals1[p.mat1 > 0.05] <- 0
labels1[p.mat1 > 0.05] <- ''

myBreaks <- c(seq(-1, 0, length.out=ceiling(1000/2) + 1),
              seq(1/1000, 1, length.out=floor(1000/2)))

#get legend as plot
p_tmp<-ggcorrplot(vals1,
           colors = col,
           lab = F, 
           show.legend = T,
           lab_size = 2,
           outline.color = 'black') + 
  guides(fill = guide_colourbar(title = "Corrr",
                                title.position = "right",size=15,
                                barheight = 45))

leg <- ggpubr::get_legend(p_tmp)


p02 <- pheatmap(vals1,
               color = colorRampPalette(col)(1000),
               breaks=myBreaks,
               display_numbers=labels1,
               cluster_cols=F,
               show_colnames = T,legend = F)

p1 <- as.ggplot(p01)
p2 <- as.ggplot(p02)
p3 <- ggplot() + theme_void()
# CellCounts

png(file=paste0(path_output,"/Figure4A_CorrelationHeatmap_CNIH4vsICs_PTB.png"),res=900, units = 'in',
    width = 10,
    height =10,)
p1
dev.off()

png(file=paste0(path_output,"/Figure4A_CorrelationHeatmap_CNIH4vsICsvsCellCounts_PTB.png"),res=900, units = 'in',
    width = 12,
    height =4,)
p2
dev.off()

# CellCounts combined
k = 0.996
png(file=paste0(path_output,"/Figure4A_CorrelationHeatmap_CNIH4vsICsvsCellCounts_PTB_combine.png"),res=600, units = 'in',
    width = 12,
    height =12,)
grid.arrange(arrangeGrob(arrangeGrob(p1, p3,
                         widths = c(k,1-k)),
                         p2,
                         nrow=2,
                         heights = c(0.75,0.2)),
             ggpubr::as_ggplot(leg),
             
             widths = c(4,0.5),
             ncol=2)
dev.off()
```

### 2. Scatter plot CNIH4 vs 27 ICs

```{r}
library(ggpmisc)
library(ggpubr)

df <- vst_data %>%select(IC_names) %>% dplyr::rename('HAVCR2/TIM-3'='HAVCR2',
         'CD274/PDL1'='CD274',
         'LGALS9/Galactin-9'='LGALS9',
         'PDCD1/PD1'='PDCD1')

p1<-ggplot(data=df, aes_string(x='CNIH4',y=as.name('CD58')))+
  geom_point(alpha=0.2,col='red')+
  theme_bw()+
  stat_poly_line(se=T) +
  stat_cor(cor.coef.name = "r",size = 7,method = 'spearman')+
  theme_bw()+ 
  ylab(paste0(as.name('CD58')," expression"))+
  xlab("CNIH4 expression")+
  theme(plot.title = element_text(size = 15),
        legend.position = "none",
        axis.text = element_text(size = 15),
        axis.text.x=element_text(color="black",size=13,face="bold"),
        axis.title.x = element_text(size = 15,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=15),
        axis.title.y = element_text(size = 15,color="black",face="bold"),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=12),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=12))

png(file=paste0(path_output,"/Figure4B_scatterplot_CNIH4vsCD58.png"),res=600, units = 'in', width=7, height =7)
p1
dev.off()
```

```{r}
library(ggpmisc)
library(ggpubr)
library(cowplot)

df <- vst_data %>%select(IC_names) %>% dplyr::rename('HAVCR2/TIM-3'='HAVCR2',
         'CD274/PDL1'='CD274',
         'LGALS9/Galactin-9'='LGALS9',
         'PDCD1/PD1'='PDCD1'
         ) #295 patients

p<-list()

for(i in 1:27){
p1<-ggplot(data=df, aes_string(x='CNIH4',y=as.name(colnames(df)[i])))+
  geom_point(alpha=0.2,col='red')+
  theme_bw()+
  stat_poly_line(se=T) +
  stat_cor(cor.coef.name = "r",size = 7,method = 'spearman')+
  theme_bw()+ 
  ylab(paste0(as.name(colnames(df)[i])," expression"))+
  xlab("CNIH4 expression")+
  theme(plot.title = element_text(size = 15),
        legend.position = "none",
        axis.text = element_text(size = 15),
        axis.text.x=element_text(color="black",size=13,face="bold"),
        axis.title.x = element_text(size = 15,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=15),
        axis.title.y = element_text(size = 15,color="black",face="bold"),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=12),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=12))
  p[[i]]<-ggdraw(p1)
  ggdraw(p1)
}

png(file=paste0(path_output,"/Figure4B_scatterplot_CNIH4vsICs_all_PTB.png"),res=600, units = 'in', width=45, height =30)#30:18
ggpubr::ggarrange(
p[[1]],p[[2]],
p[[3]],p[[4]],
p[[5]],p[[6]],
p[[7]],p[[8]],
p[[9]],p[[10]],
p[[11]],p[[12]],
p[[13]],p[[14]],
p[[15]],p[[16]],
p[[17]],p[[18]],
p[[19]],p[[20]],
p[[21]],p[[22]],
p[[23]],p[[24]],
p[[25]],p[[26]],
p[[27]],
ncol =6,nrow=5)
dev.off()
```

### 3. Differential Co-expression for low and high CNIH4

```{r}
healthy_df <- count_data_adjusted_ComBat %>% filter(LIMS_ID %in% filter(Cli_All,trial_number=='HV'|trial_number=='43TB')$LIMS_ID)
qt<-quantile(healthy_df$CNIH4,c(0.75)) #Q3 CNIH4 of healthy control

PTB_df <- count_data_adjusted_ComBat %>% filter(LIMS_ID%in%filter(Cli_All,trial_number=='MDR_PTB'|trial_number=='DS_PTB')$LIMS_ID)

# tmp <- PTB_df %>% select(CNIH4) %>% mutate(level=ifelse((CNIH4<qt[1]),"LOW",'HIGH'))

tmp <- PTB_df %>% select(CNIH4) %>% mutate(level=ifelse((CNIH4<qt[1]),1,2))

rownames(tmp) <- PTB_df$LIMS_ID
table(tmp$level) # 244 52
vst_data2<-merge(vst_data,tmp[,2, drop=FALSE],by=0)
rownames(vst_data2) <- vst_data2$LIMS_ID
dim(vst_data2) # 295 31427

```

```{r}
# lowCNIH4vsIC_df <- vst_data2 %>% select(append(IC_names[1:27],'level')) %>% filter(level=='LOW') %>% select(-level)
# highCNIH4vsIC_df <- vst_data2 %>% select(append(IC_names[1:27],'level')) %>% filter(level=='HIGH') %>% select(-level)

lowCNIH4vsIC_df <- vst_data2 %>% select(append(IC_names[1:27],'level')) %>% filter(level==1) %>% select(-level)
highCNIH4vsIC_df <- vst_data2 %>% select(append(IC_names[1:27],'level')) %>% filter(level==2) %>% select(-level)

```

```{r}

library(dcanr)

cor_method = 'spearman'
ICvslowCNIH4_mat <- cor(lowCNIH4vsIC_df,
                      use='pairwise.complete.obs',
                      method = cor_method)

p.ICvslowCNIH4_mat <- cor.pval.test(lowCNIH4vsIC_df, 
                                  lowCNIH4vsIC_df, 
                                  method=cor_method, 
                                  use='pairwise.complete.obs')

ICvshighCNIH4_mat <- cor(highCNIH4vsIC_df,
                      use='pairwise.complete.obs',
                      method = cor_method)
p.ICvshighCNIH4_mat <- cor.pval.test(highCNIH4vsIC_df, 
                                  highCNIH4vsIC_df, 
                                  method=cor_method, 
                                  use='pairwise.complete.obs')

emat <- t(vst_data2%>%select(IC_names[1:27])) #tranpose
condition <- vst_data2$level # high and low
scores <- dcScore(emat, 
                  condition,
                  dc.method = "zscore",
                  cor.method = 'spearman')
raw_p <- dcTest(scores, emat, condition)
adj_p <- dcAdjust(raw_p, f = p.adjust, method = 'fdr')
```

#### 3.1. Differential Coexpression Heatmap

```{r}
library(cowplot)
library(grid)
library(NatParksPalettes)

col1 =c("#395b85", 'white', "#FF681E")
col2 = c("#00b8ff","black","magenta")

p1<-ggcorrplot(ICvslowCNIH4_mat[,27:1],
           p.mat = p.ICvslowCNIH4_mat[,27:1], 
           colors = col1,
           lab = F, 
           show.legend = F,
           lab_size = 2,
           insig="blank",
           outline.color = 'black',)+
  theme(legend.position="")+
  labs(fill="Corr")+
  theme(axis.text.x = element_text(face="bold", colour="black", size=12),
        axis.text.y = element_text(face="bold", colour="black", size=12),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=12))+
  scale_fill_gradientn(colors=natparks.pals("Acadia"),limit = c(-1,1))

p2<-ggcorrplot(ICvshighCNIH4_mat[,27:1],
           p.mat = p.ICvshighCNIH4_mat[,27:1], 
           colors = col1,
           lab = F,
           lab_size = 2,
           insig="blank",
           outline.color = 'black')+theme(legend.position="right")+
  labs(fill="Corr")+
  theme(axis.text.x = element_text(face="bold", colour="black", size=12),
        axis.text.y = element_text(face="bold", colour="white", size=6),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=12))+
  guides(fill = guide_colourbar(title = "Corr",
                                size=15,
                                barheight = 40)) + 
  scale_fill_gradientn(colors=natparks.pals("Acadia"),limit = c(-1,1))
  
p3 <- ggcorrplot(scores[,27:1],
           p.mat = raw_p[,27:1], 
           colors = col2,
           show.diag = TRUE,
           lab = F, 
           lab_size = 2,
           insig="blank",ggtheme =ggplot2::theme_void,
           outline.color = 'black',
           ) +
  theme(legend.position="right")+
  #theme_void()+
  labs(fill="Differential Z-scores")+
  scale_fill_gradient2(limit = c(-5,5), 
                       low = col2[1], 
                       high =  col2[3], 
                       mid = col2[2], midpoint = 0)+
  
  theme(axis.text.x = element_text(face="bold", colour="black", size=12,angle = 45),
        axis.text.y = element_text(face="bold", colour="black", size=12,hjust = 1),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=12,angle = -90,
                                     hjust = 0.5, vjust = 0.5),
        panel.border = element_rect(colour = "black", fill=NA, size=1,linewidth = 1)) +   
  guides(fill = guide_colourbar(title = "Differential Z-scores",
                                title.position = "right",size=15,
                                barheight = 40))
p3
# 
# 
# png(file=paste0(path_output,"/Figure4C_differential_coexpression_heatmap.png"),res=600, units = 'in', width=30, height =10)
# plot_grid(p1, p2, p3, align = "h", ncol = 3, rel_widths = c(1/3.19, 1/3, 1/3))
# dev.off()
```

```{r}
#Combined p1 and p2 into single heatmap
library(NatParksPalettes)
library(cowplot)
library(grid)

highlow_mat <- ICvshighCNIH4_mat
highlow_mat[upper.tri(ICvshighCNIH4_mat,diag=F)] <- ICvslowCNIH4_mat[upper.tri(ICvslowCNIH4_mat,diag=F)]
p.highlow_mat <- p.ICvshighCNIH4_mat
p.highlow_mat[upper.tri(p.ICvshighCNIH4_mat,diag=F)] <- p.ICvslowCNIH4_mat[upper.tri(p.ICvslowCNIH4_mat,diag=F)]

p12<-ggcorrplot(highlow_mat[,27:1],
           p.mat = p.highlow_mat[,27:1], 
           colors = col1,
           lab = F,
           lab_size = 2,
           insig="blank",
           outline.color = 'black')+theme(legend.position="right")+
  labs(fill="Corr")+
  theme(axis.text.x = element_text(face="bold", colour="black", size=12),
        axis.text.y = element_text(face="bold", colour="black", size=12),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=12))+
  guides(fill = guide_colourbar(title = "Corr",
                                size=15,
                                barheight = 40)) + 
  scale_fill_gradientn(colors=natparks.pals("Acadia"),limit = c(-1,1)) + annotate('segment', 
               x = 27.5, 
               xend = 0.5, 
               y = 0.5, 
               yend = 27.5, color = 'black', size = 0.25)

tmp <- highlow_mat
tmp[is.na(raw_p)] <- NA
tmp[raw_p >= 0.05] <- NA
tmp<-which(!is.na(tmp[27:1,]), arr.ind=T)

for(row in 1:nrow(tmp)) {
  i = tmp[row,]
  p12 <- p12 + annotate('rect', 
               xmin = i[2]-0.5, 
               xmax = i[2]+0.5, 
               ymin = i[1]-0.5, 
               ymax = i[1]+0.5,
           fill = NA, color = 'black', size = 0.5)
}

p3 <- ggcorrplot(scores[,27:1],
           p.mat = raw_p[,27:1], 
           colors = col2,
           show.diag = TRUE,
           lab = F, 
           lab_size = 2,
           insig="blank",ggtheme =ggplot2::theme_void,
           outline.color = 'black',
           ) +
  theme(legend.position="right")+
  #theme_void()+
  labs(fill="Differential Z-scores")+
  scale_fill_gradient2(limit = c(-5,5), 
                       low = col2[1], 
                       high =  col2[3], 
                       mid = col2[2], midpoint = 0)+
  
  theme(axis.text.x = element_text(face="bold", colour="black", size=12,angle = 45),
        axis.text.y = element_text(face="bold", colour="black", size=12,hjust = 1),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=12,angle = -90,
                                     hjust = 0.5, vjust = 0.5),
        panel.border = element_rect(colour = "black", fill=NA, size=1,linewidth = 1)) +   
  guides(fill = guide_colourbar(title = "Differential Z-scores",
                                title.position = "right",size=15,
                                barheight = 40))

png(file=paste0(path_output,"/Figure4C_differential_coexpression_heatmap_PTB.png"),res=600, units = 'in', width=20, height =10)
plot_grid(p12,p3, align = "h", ncol = 2, rel_widths = c(1/5, 1/5))
dev.off()
```

#### 3.2. Differential Coexpression Network

```{r}
library(igraph)
library(qgraph)

#set.seed(7)
dcnet <- dcNetwork(scores,
                   raw_p,
                   thresh=0.05)

e <- get.edgelist(dcnet,names=FALSE)
l <- qgraph.layout.fruchtermanreingold(e,
                                       vcount=vcount(dcnet),
                                       area=13*(vcount(dcnet)^2),
                                       repulse.rad=(vcount(dcnet)^3.3)
                                       )
png(file=paste0(path_output,"/Figure4C_differential_coexpression_network_PTB.png"),res=900, units = 'in', width=6.5, height = 6.5)

plot(dcnet,layout=l,
     vertex.size=2,
     vertex.label.cex = 0.5,
     vertex.label.dist=0.5,
     edge.width=1,
     main = 'Differential co-expression network',
     width=2,
     layout.rotate = 100
     )
dev.off()
```

#### 3.3. Boxplot of pairwise coexpression in low and high CNIH4 group

```{r}
library(dcanr)
library(ggpubr)

emat <- t(vst_data2%>%select(IC_names[1:27])) #tranpose
condition <- vst_data2$level # high and low
scores <- dcScore(emat, 
                  condition,
                  dc.method = "zscore",
                  cor.method = 'spearman')
raw_p <- dcTest(scores, emat, condition)
adj_p <- dcAdjust(raw_p, f = p.adjust, method = 'fdr')

sigf_ICvslowCNIH4_mat <- ICvslowCNIH4_mat
sigf_ICvslowCNIH4_mat[raw_p>=0.05] <- NA

sigf_ICvshighCNIH4_mat <- ICvshighCNIH4_mat
sigf_ICvshighCNIH4_mat[raw_p>=0.05] <- NA

low_corr <- abs(sigf_ICvslowCNIH4_mat[upper.tri(sigf_ICvslowCNIH4_mat,diag=F)])
high_corr <- abs(sigf_ICvshighCNIH4_mat[upper.tri(sigf_ICvshighCNIH4_mat,diag=F)])

low_high <- data.frame(low_corr,high_corr)

names(low_high) <- c('Low CNIH4','High CNIH4')
df <- pivot_longer(low_high,colnames(low_high))
names(df) <- c('CNIH4level','Coexpression')

p<-ggplot(data=df,
       aes(x=reorder(CNIH4level, Coexpression, na.rm = TRUE),
           y=Coexpression)) +
  theme_bw() +
  geom_violin(aes(color=CNIH4level,fill=CNIH4level),
              width=0.7,alpha=0.5,
              trim=FALSE)+
  scale_fill_manual(values=c("#B48EAD","#81A1C1"))+
  scale_color_manual(values=c("#B48EAD","#81A1C1"))+
  geom_boxplot(aes(color=CNIH4level),
               size=1,
               position=position_dodge(1),
               width=0.2,
               outlier.colour="red",
               outlier.shape=NA,outlier.size=3)+
    geom_jitter(aes(color=CNIH4level,fill=CNIH4level),
              width=0.05,
              height=0,
              size=2,
              alpha = 0.7)+
  #geom_quasirandom(aes(color=CNIH4level),dodge.width=.8)+
  #scale_color_manual(values=c("#BF616A","#81A1C1"))+
  
  labs(x="",y = "Absolute pairwise coexpression of \n27 immune checkpoint genes (p <0.05)", title = "")+
  theme(plot.title = element_text(size = 25),
        legend.position = "none",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_text(color="black",size=13,face="bold"),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold"))+
  stat_compare_means( #aes(label = paste0("p = ", ..p.format..)),
                      size = 5,
                      label.x = 1.3,
                      label.y = 1.3)

# png(file=paste0(path_output,"/Figure4C_differential_coexpression_boxplot_PTB.png"),res=900, units = 'in', width=8, height = 7)
p
# dev.off()
```

### 4.0. PTB vs Health vs (TBM)

```{r}
library(ggpubr)
library(rstatix)

healthy_df <- count_data_adjusted_ComBat %>% filter(LIMS_ID %in% filter(Cli_All,trial_number=='HV'|trial_number=='43TB')$LIMS_ID)
rownames(healthy_df) <- healthy_df$LIMS_ID

PTB_df <- count_data_adjusted_ComBat %>% filter(LIMS_ID%in%filter(Cli_All,trial_number=='MDR_PTB'|trial_number=='DS_PTB')$LIMS_ID)
rownames(PTB_df) <- PTB_df$LIMS_ID


TBM_df <- count_data_adjusted_ComBat %>% filter(LIMS_ID%in%filter(Cli_All,(trial_number=='26TB'|trial_number=='27TB')&(Timepoint=='D0'))$LIMS_ID)
rownames(TBM_df) <- TBM_df$LIMS_ID

tmp1 <- merge(impute_df,healthy_df,by=0) %>% select(B.cells.naive:Neutrophils)
dim(tmp1)
tmp1 <- pivot_longer(tmp1,B.cells.naive:Neutrophils)
tmp1$group <- rep('Healthy',each=dim(tmp1)[1])

tmp2 <- merge(impute_df,PTB_df,by=0) %>% select(B.cells.naive:Neutrophils)
dim(tmp2)
tmp2 <- pivot_longer(tmp2,B.cells.naive:Neutrophils)
tmp2$group <- rep('PTB',each=dim(tmp2)[1])

tmp3 <- merge(impute_df,TBM_df,by=0) %>% select(B.cells.naive:Neutrophils)
dim(tmp3)
tmp3 <- pivot_longer(tmp3,B.cells.naive:Neutrophils)
tmp3$group <- rep('TBM',each=dim(tmp3)[1])



tmp<-rbind(tmp1,tmp2,tmp3)

stat.test <- tmp %>%
  group_by(name) %>% 
  t_test(value ~ group) %>% 
  add_significance("p") %>%
  add_xy_position(x = "name")

p <- ggplot(data=tmp,
       aes(x=name,y=value, color=factor(group, level = c('Healthy','PTB','TBM'))))+
  geom_boxplot() +
  geom_point(position=position_jitterdodge(jitter.width =0.3),
             alpha=0.1)+
  scale_color_manual(values=c("#81A1C1","#B48EAD","#A3BE8C"))+
  labs(color='Group')+ 
  xlab('Cell type')+
  ylab('% Cell population')+
  theme_bw()+
  theme(plot.title = element_text(size = 25),
        legend.position = "right",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_text(color="black",size=13,face="bold",angle = 45,vjust = 1, hjust=1),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold")) +
  stat_pvalue_manual(stat.test, label = 'p.signif', tip.length = 0)

png(file=paste0(path_output,"/Figure_CellPop_HealthyvsPTBvsTBM.png"),res=900, units = 'in', width=17, height = 9)
p
dev.off()
```

### 4. Boxplot for cell population, compare in low vs high CNIH4 groups

```{r}
tmp <- merge(impute_df,vst_data2,by=0) %>% select(c(B.cells.naive:Neutrophils,level))
tmp <- pivot_longer(tmp,B.cells.naive:Neutrophils)

p<-ggplot(data=tmp,
       aes(x=name,y=value))+
  geom_boxplot(aes(color=name)) +
  geom_jitter(height=0,aes(color=name),alpha=0.1)+
  xlab('Cell type')+
  ylab('% Cell population')+
  theme_bw()+
  theme(plot.title = element_text(size = 25),
        legend.position = "",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_text(color="black",size=13,face="bold",angle = 45,vjust = 1, hjust=1),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold"))
  
png(file=paste0(path_output,"/Figure_CellPop_All.png"),res=900, units = 'in', width=14, height = 9)
p
dev.off()

stat.test <- tmp %>%
  group_by(name) %>% 
  t_test(value ~ level) %>% 
  add_significance("p") %>%
  add_xy_position(x = "name")

p <- ggplot(data=tmp,
       aes(x=name,y=value, color=factor(level, level = c('LOW','HIGH'))))+
  geom_boxplot() +
  geom_point(position=position_jitterdodge(jitter.width =0.3),
             alpha=0.1)+
  scale_color_manual(values=c("#81A1C1","#B48EAD"))+
  labs(color='CNIH4 Level')+ 
  xlab('Cell type')+
  ylab('% Cell population')+
  theme_bw()+
  theme(plot.title = element_text(size = 25),
        legend.position = "right",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_text(color="black",size=13,face="bold",angle = 45,vjust = 1, hjust=1),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold")) +
  stat_pvalue_manual(stat.test, label = 'p.signif', tip.length = 0)

png(file=paste0(path_output,"/Figure_CellPop_LowvsHighCNIH4_PTB.png"),res=900, units = 'in', width=14, height = 9)
p
dev.off()
```

### 5. Box plot of CNIH4 and PDL1 expression in Neutrophils in high and low CNIH4 groups

```{r}
library(ggpubr)
require(plyr)
library(ggbeeswarm)

tmp0 <- read.csv(paste0(path_input,'Raw_data/LM22_ref_CNIH4/ICs/CIBERSORTxHiRes_NA_Neutrophils_Window40.txt'),sep='\t')
rownames(tmp0) <- tmp0$GeneSymbol
tmp0$GeneSymbol <- NULL
tmp0 <- t(tmp0) %>% as.data.frame() %>% dplyr::rename(CNIH4InNeutrophils = CNIH4)

lowCNIH4_neutrophils <- merge(lowCNIH4vsIC_df,tmp0,by=0)
highCNIH4_neutrophils <- merge(highCNIH4vsIC_df,tmp0,by=0)
CNIH4lv_neutrophils <- merge(vst_data2,tmp0,by=0)[,c('CNIH4InNeutrophils','level')]
CNIH4lv_neutrophils$level <- mapvalues(CNIH4lv_neutrophils$level, 
                                        from=c('LOW',"HIGH"), 
                                        to=c("Low CNIH4","High CNIH4"))

p<-ggplot(data=CNIH4lv_neutrophils,
       aes(x=reorder(level, CNIH4InNeutrophils,na.rm = TRUE),
           y=CNIH4InNeutrophils)) +
  theme_bw() +
  scale_fill_manual(values=c("#B48EAD","#81A1C1"))+
  scale_color_manual(values=c("#B48EAD","#81A1C1"))+
  geom_boxplot(aes(color=level),
               size=1,
               position=position_dodge(1),
               width=0.5,
               outlier.colour="red",
               outlier.shape=NA,outlier.size=3,
               varwidth=TRUE
               )+
  geom_beeswarm(aes(color=level),priority='density',size=1,alpha=0.5,cex = 1,
                corral.width = 0.1)+
  labs(x="",y = "CNIH4 expression in Neutrophils", title = "")+
  theme(plot.title = element_text(size = 25),
        legend.position = "none",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_text(color="black",size=13,face="bold"),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold"))+
  stat_compare_means( #aes(label = paste0("p = ", ..p.format..)),
                      size = 5,
                      label.x = 1.3,
                      label.y = 220)

#png(file=paste0(path_output,"/Figure_HighvsLowCNIH4InNeutrophils_CNIH4.png"),res=900, units = 'in', width=5, height = 7)
p
#dev.off()
```

## TBM

### 1. Correlation heatmap CNIH4 vs 25 ICs

```{r}
library(pheatmap)
library(ggplot2)
library(ggplotify)
#library(NatParksPalettes)

cor.pval.test <- function(x1, x2, ...) {
  r <- ncol(x1)
  n <- ncol(x2)
  p.mat<- matrix(NA, r, n)
  for (i in 1:r) {
    for (j in 1:n) {
      tmp <- cor.test(x1[, i], x2[, j], ...)
      p.mat[i, j] <- tmp$p.value
    }
  }
  rownames(p.mat) <-  colnames(x1)
  colnames(p.mat) <- colnames(x2)
  return(p.mat)
}

IC_names<-c('CD200', 'CD2','CD58','TNFSF4', 'CD200R1', 'IDO1','TNFRSF18', 'TNFSF18',
             'CD28','BTLA', 'CD27', 'HAVCR2', 'CD274', 'TNFSF14', 'TNFRSF4',
             'CD276', 'TNFRSF14', 'PDCD1', 'CD86', 'TNFSF9', 'TIGIT','CD80',
             'ICOS', 'TNFRSF9', 'CTLA4', 'LAG3','LGALS9','CNIH4')

ICvsCNIH4_df <- tbm_vst %>% select(IC_names)
#ICvsCNIH4_df <- merge(ICvsCNIH4_df, impute_df, by=0) %>% tibble::column_to_rownames('Row.names') %>% select("CD200" : "Neutrophils" )
ICvsCNIH4_df <- ICvsCNIH4_df[vapply(ICvsCNIH4_df, 
                                    function(x) length(unique(x)) > 1, 
                                    logical(1L))] %>% dplyr::rename('HAVCR2/TIM-3'='HAVCR2',
                                                                    'CD274/PDL1'='CD274',
                                                                    'LGALS9/Galactin-9'='LGALS9',
                                                                    'PDCD1/PD1'='PDCD1')

cor_method = 'spearman'
ICvsCNIH4_mat <- cor(ICvsCNIH4_df,
                      use='pairwise.complete.obs',
                      method = cor_method)

p.ICvsCNIH4_mat <- cor.pval.test(ICvsCNIH4_df, 
                                  ICvsCNIH4_df, 
                                  method=cor_method, 
                                  use='pairwise.complete.obs')

#col=c("#001164", 'white', "#FF681E")
#col=c("#5E81AC", '#EBCB8B', "#BF616A")
col=c("#395b85", 'white', "#FF681E")

draw_colnames_45 <- function (coln, gaps, ...) {
    coord <- pheatmap:::find_coordinates(length(coln), gaps)
    x     <-  coord$coord - 0.5 * coord$size
    res   <- grid::textGrob(
      coln, x = x, y = unit(1, "npc") - unit(3,"bigpts"),
      vjust = 0.75, hjust = 1, rot = 45, gp = grid::gpar(...)
    )
    return(res)
}

assignInNamespace(
  x = "draw_colnames",
  value = "draw_colnames_45",
  ns = asNamespace("pheatmap")
)

# CNIH4 vs 25 ICs
mat1 <- ICvsCNIH4_mat
p.mat1 <- p.ICvsCNIH4_mat
labels1 <-round(mat1,2)
vals1 <- labels1
vals1[p.mat1 > 0.05] <- 0
labels1[p.mat1 > 0.05] <- ''

myBreaks <- c(seq(-1, 0, length.out=ceiling(1000/2) + 1),
              seq(1/1000, 1, length.out=floor(1000/2)))

p01 <- pheatmap(vals1,
               color = colorRampPalette(col)(1000),
               breaks=myBreaks,
               display_numbers=labels1,
               show_colnames = F,legend = F)
p01
row.order <- p01$tree_row$order
col.order <- p01$tree_col$order
p1 <- as.ggplot(p01)
```

```{r}
Cell_names <- c("Neutrophil_counts","Lymphocyte_counts", "Monocyte_counts", "Eosinophil_counts")

ICvsCNIH4_df <- tbm_vst %>% select(append(IC_names,Cell_names))
ICvsCNIH4_df <- ICvsCNIH4_df[vapply(ICvsCNIH4_df, 
                                    function(x) length(unique(x)) > 1, 
                                    logical(1L))] %>% dplyr::rename('HAVCR2/TIM-3'='HAVCR2',
                                                                    'CD274/PDL1'='CD274',
                                                                    'LGALS9/Galactin-9'='LGALS9',
                                                                    'PDCD1/PD1'='PDCD1')
cor_method = 'spearman'
ICvsCNIH4_mat <- cor(ICvsCNIH4_df,
                      use='pairwise.complete.obs',
                      method = cor_method)

p.ICvsCNIH4_mat <- cor.pval.test(ICvsCNIH4_df, 
                                  ICvsCNIH4_df, 
                                  method=cor_method, 
                                  use='pairwise.complete.obs')

mat1 <- ICvsCNIH4_mat[29:32,1:28]
mat1 <- mat1[,col.order]
p.mat1 <- p.ICvsCNIH4_mat[29:32,1:28]
p.mat1 <- p.mat1[,col.order]

labels1 <-round(mat1,2)
vals1 <- labels1
vals1[p.mat1 > 0.05] <- 0
labels1[p.mat1 > 0.05] <- ''

myBreaks <- c(seq(-1, 0, length.out=ceiling(1000/2) + 1),
              seq(1/1000, 1, length.out=floor(1000/2)))

#get legend as plot
p_tmp<-ggcorrplot(vals1,
           colors = col,
           lab = F, 
           show.legend = T,
           lab_size = 2,
           outline.color = 'black') + 
  guides(fill = guide_colourbar(title = "Corrr",
                                title.position = "right",size=15,
                                barheight = 45))

leg <- ggpubr::get_legend(p_tmp)


p02 <- pheatmap(vals1,
               color = colorRampPalette(col)(1000),
               breaks=myBreaks,
               display_numbers=labels1,
               cluster_cols=F,
               show_colnames = T,legend = F)

p1 <- as.ggplot(p01)
p2 <- as.ggplot(p02)
p3 <- ggplot() + theme_void()
# CellCounts

png(file=paste0(path_output,"/Figure4A_CorrelationHeatmap_CNIH4vsICs_TBM.png"),res=900, units = 'in',
    width = 10,
    height =10,)
p1
dev.off()

png(file=paste0(path_output,"/Figure4A_CorrelationHeatmap_CNIH4vsICsvsCellCounts_TBM.png"),res=900, units = 'in',
    width = 12,
    height =4,)
p2
dev.off()

# CellCounts combined
k = 0.996
png(file=paste0(path_output,"/Figure4A_CorrelationHeatmap_CNIH4vsICsvsCellCounts_TBM_combine.png"),res=600, units = 'in',
    width = 12,
    height =12,)
grid.arrange(arrangeGrob(arrangeGrob(p1, p3,
                         widths = c(k,1-k)),
                         p2,
                         nrow=2,
                         heights = c(0.75,0.2)),
             ggpubr::as_ggplot(leg),
             
             widths = c(4,0.5),
             ncol=2)

dev.off()
```

### 2. Scatter plot

```{r}
library(ggpmisc)
library(ggpubr)

df <- tbm_vst %>%select(IC_names) %>% dplyr::rename('HAVCR2/TIM-3'='HAVCR2',
         'CD274/PDL1'='CD274',
         'LGALS9/Galactin-9'='LGALS9',
         'PDCD1/PD1'='PDCD1')

l <- c('CD274/PDL1','CD58')
i = 2
ic <- l[i]
p1<-ggplot(data=df, aes_string(x='CNIH4',y=as.name(ic)))+
geom_point(alpha=0.2,col='red')+
theme_bw()+
stat_poly_line(se=T) +
stat_cor(cor.coef.name = "r",size = 7,method = 'spearman')+
theme_bw()+ 
ylab(paste0(as.name(ic)," expression"))+
xlab("CNIH4 expression")+
theme(plot.title = element_text(size = 15),
      legend.position = "none",
      axis.text = element_text(size = 15),
      axis.text.x=element_text(color="black",size=13,face="bold"),
      axis.title.x = element_text(size = 15,color="black",face="bold"),
      axis.text.y=element_text(color="black",size=15),
      axis.title.y = element_text(size = 15,color="black",face="bold"),
      legend.text = element_text(size = 15),
      legend.title = element_text( colour="black", size=12),
      legend.key = element_rect(colour = NA),
      strip.text=element_text(face = "bold",size=12))
png(file=paste0(path_output,"/Figure4B_scatterplot_CNIH4vsCD58_TBM.png"),res=600, units = 'in', width=7, height=7)
p1
dev.off()

```

```{r}
library(ggpmisc)
library(ggpubr)
library(cowplot)

df <- tbm_vst %>%select(IC_names) %>% dplyr::rename('HAVCR2/TIM-3'='HAVCR2',
         'CD274/PDL1'='CD274',
         'LGALS9/Galactin-9'='LGALS9',
         'PDCD1/PD1'='PDCD1'
         ) #281 patients

p<-list()

for(i in 1:27){
p1<-ggplot(data=df, aes_string(x='CNIH4',y=as.name(colnames(df)[i])))+
  geom_point(alpha=0.2,col='red')+
  theme_bw()+
  stat_poly_line(se=T) +
  stat_cor(cor.coef.name = "r",size = 7,method = 'spearman')+
  theme_bw()+ 
  ylab(paste0(as.name(colnames(df)[i])," expression"))+
  xlab("CNIH4 expression")+
  theme(plot.title = element_text(size = 15),
        legend.position = "none",
        axis.text = element_text(size = 15),
        axis.text.x=element_text(color="black",size=13,face="bold"),
        axis.title.x = element_text(size = 15,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=15),
        axis.title.y = element_text(size = 15,color="black",face="bold"),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=12),
        legend.key = element_rect(colour = NA),
        strip.text=element_text(face = "bold",size=12))
  p[[i]]<-ggdraw(p1)
  ggdraw(p1)
}

png(file=paste0(path_output,"/Figure4B_scatterplot_CNIH4vsICs_all_TBM.png"),res=600, units = 'in', width=45, height =30)
ggpubr::ggarrange(
p[[1]],p[[2]],
p[[3]],p[[4]],
p[[5]],p[[6]],
p[[7]],p[[8]],
p[[9]],p[[10]],
p[[11]],p[[12]],
p[[13]],p[[14]],
p[[15]],p[[16]],
p[[17]],p[[18]],
p[[19]],p[[20]],
p[[21]],p[[22]],
p[[23]],p[[24]],
p[[25]],p[[26]],p[[27]],
ncol =6,nrow=5)
dev.off()
```

### 3. Differential Coexpression

```{r}
healthy_df <- count_data_adjusted_ComBat %>% filter(LIMS_ID %in% filter(Cli_All,trial_number=='HV'|trial_number=='43TB')$LIMS_ID)
qt<-quantile(healthy_df$CNIH4,c(0.75)) #Q3 CNIH4 of healthy control

TBM_df <- count_data_adjusted_ComBat %>% filter(LIMS_ID%in%filter(Cli_All,(trial_number=='26TB'|trial_number=='27TB')&(Timepoint=='D0'))$LIMS_ID)

# tmp <- TBM_df %>% select(CNIH4) %>% mutate(level=ifelse((CNIH4<qt[1]),"LOW",'HIGH'))

tmp <- TBM_df %>% select(CNIH4) %>% mutate(level=ifelse((CNIH4<qt[1]),1,2))

rownames(tmp) <- TBM_df$LIMS_ID
table(tmp$level) # 241 40
tbm_vst2<-merge(tbm_vst,tmp[,2, drop=FALSE],by=0)
rownames(tbm_vst2) <- tbm_vst2$LIMS_ID
tbm_vst2$Row.names <- NULL

tmp <-  read.csv(paste0(path_input,'Raw_data/data_3genes_Xray_select_17_JAN_2024.csv')) # check if TBM contain PTB
tmp$PTB_group[is.na(tmp$PTB_group)] <- 'NA'
tmp <- filter(tmp, grepl('26TB|27TB', ID)) %>% select(c('LIMS_ID','PTB_group'))
rownames(tmp)<-tmp$LIMS_ID
tmp$LIMS_ID <- NULL
tbm_vst2 <- merge(tbm_vst2,tmp,by=0)
rownames(tbm_vst2) <- tbm_vst2$LIMS_ID
tbm_vst2$Row.names <- NULL
dim(tbm_vst2) # 281 31391
```

```{r}
library(dcanr)
tmp0 <- tbm_vst2# ALL
#tmp0 <- filter(tbm_vst2,PTB_group=='NA') # NA group
#tmp0 <- filter(tbm_vst2,PTB_group=='TBM without PTB') # NA group
#tmp0 <- filter(tbm_vst2,PTB_group=='TBM with PTB') # NA group

# lowCNIH4vsIC_df <- tmp0 %>% select(append(IC_names[1:27],'level')) %>% filter(level=='LOW') %>% select(-level)
# highCNIH4vsIC_df <- tmp0 %>% select(append(IC_names[1:27],'level')) %>% filter(level=='HIGH') %>% select(-level)

lowCNIH4vsIC_df <- tmp0 %>% select(append(IC_names[1:27],'level')) %>% filter(level==1) %>% select(-level)
highCNIH4vsIC_df <- tmp0 %>% select(append(IC_names[1:27],'level')) %>% filter(level==2) %>% select(-level)

cor_method = 'spearman'
ICvslowCNIH4_mat <- cor(lowCNIH4vsIC_df,
                      use='pairwise.complete.obs',
                      method = cor_method)

p.ICvslowCNIH4_mat <- cor.pval.test(lowCNIH4vsIC_df, 
                                  lowCNIH4vsIC_df, 
                                  method=cor_method, 
                                  use='pairwise.complete.obs')

ICvshighCNIH4_mat <- cor(highCNIH4vsIC_df,
                      use='pairwise.complete.obs',
                      method = cor_method)

p.ICvshighCNIH4_mat <- cor.pval.test(highCNIH4vsIC_df, 
                                  highCNIH4vsIC_df, 
                                  method=cor_method, 
                                  use='pairwise.complete.obs')

emat <- t(tmp0%>%select(IC_names[1:27])) #tranpose, 27
condition <- tmp0$level # high and low
scores <- dcScore(emat,
                  condition,
                  dc.method = "zscore",
                  cor.method = 'spearman')

raw_p <- dcTest(scores, emat, condition)
adj_p <- dcAdjust(raw_p, f = p.adjust, method = 'fdr')


```

#### 3.1. Differential Coexpression Heatmap

```{r}
library(NatParksPalettes)
library(cowplot)
library(grid)

col1 =c("#395b85", 'white', "#FF681E")
col2 = c("#00b8ff","black","magenta")

highlow_mat <- ICvshighCNIH4_mat
highlow_mat[upper.tri(ICvshighCNIH4_mat,diag=F)] <- ICvslowCNIH4_mat[upper.tri(ICvslowCNIH4_mat,diag=F)]
p.highlow_mat <- p.ICvshighCNIH4_mat
p.highlow_mat[upper.tri(p.ICvshighCNIH4_mat,diag=F)] <- p.ICvslowCNIH4_mat[upper.tri(p.ICvslowCNIH4_mat,diag=F)]

p12<-ggcorrplot(highlow_mat[,27:1],
           p.mat = p.highlow_mat[,27:1], 
           colors = col1,
           lab = F,
           lab_size = 2,
           insig="blank",
           outline.color = 'black')+
  theme(legend.position="right")+
  labs(fill="Corr")+
  theme(axis.text.x = element_text(face="bold", colour="black", size=12),
        axis.text.y = element_text(face="bold", colour="black", size=12),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=12))+
  guides(fill = guide_colourbar(title = "Corr",
                                size=15,
                                barheight = 40)) + 
  scale_fill_gradientn(colors=natparks.pals("Acadia"),limit = c(-1,1)) + 
  annotate('segment', 
               x = 27.5, 
               xend = 0.5, 
               y = 0.5, 
               yend = 27.5, color = 'black', size = 0.25)

tmp <- highlow_mat
tmp[is.na(raw_p)] <- NA
tmp[raw_p >= 0.05] <- NA

tmp<-which(!is.na(tmp[27:1,]), arr.ind=T)

for(row in 1:nrow(tmp)) {
  i = tmp[row,]
  p12 <- p12 + annotate('rect', 
               xmin = i[2]-0.5, 
               xmax = i[2]+0.5, 
               ymin = i[1]-0.5, 
               ymax = i[1]+0.5,
           fill = NA, 
           color = 'black',
           size = 0.5)
}

p3 <- ggcorrplot(scores[,27:1],
           p.mat = raw_p[,27:1], 
           colors = col2,
           show.diag = TRUE,
           lab = F, 
           lab_size = 2,
           insig="blank",ggtheme =ggplot2::theme_void,
           outline.color = 'black',
           ) +
  theme(legend.position="right")+
  #theme_void()+
  labs(fill="Differential Z-scores")+
  scale_fill_gradient2(limit = c(-5,5), 
                       low = col2[1], 
                       high =  col2[3], 
                       mid = col2[2], midpoint = 0)+
  
  theme(axis.text.x = element_text(face="bold", colour="black", size=12,angle = 45),
        axis.text.y = element_text(face="bold", colour="black", size=12,hjust = 1),
        legend.text = element_text(size = 15),
        legend.title = element_text( colour="black", size=12,angle = -90,
                                     hjust = 0.5, vjust = 0.5),
        panel.border = element_rect(colour = "black", fill=NA, size=1,linewidth = 1)) +   
  guides(fill = guide_colourbar(title = "Differential Z-scores",
                                title.position = "right",size=15,
                                barheight = 40))

png(file=paste0(path_output,"/Figure4C_differential_coexpression_heatmap_TBM_NA_PTB.png"),res=600, units = 'in', width=20, height =10)
plot_grid(p12,p3, align = "h", ncol = 2, rel_widths = c(1/5, 1/5))
dev.off()
```

#### 3.2. Differential Coexpression Network

```{r}
library(igraph)
library(qgraph)

set.seed(123)
dcnet <- dcNetwork(scores, raw_p,0.05)
plot(dcnet, vertex.label = '')

e <- get.edgelist(dcnet,names=FALSE)
l <- qgraph.layout.fruchtermanreingold(e,
                                       vcount=vcount(dcnet),
                                       area=7.5*(vcount(dcnet)^2),
                                       repulse.rad=(vcount(dcnet)^3.3)
                                       )
png(file=paste0(path_output,"/Figure4C_differential_coexpression_network_TBM.png"),res=900, units = 'in', width=6.5, height = 6.5)

plot(dcnet,layout=l,
     vertex.size=2,
     vertex.label.cex = 0.5,
     vertex.label.dist=0.5,
     edge.width=1,
     main = 'Differential co-expression network',
     width=2,
     layout.rotate = 100
     )
dev.off()
```

#### 3.4 Boxplot

```{r}
library(dcanr)
library(ggpubr)

# emat <- t(tmp0%>%select(IC_names[1:27])) #tranpose
# condition <- tmp0$level # high and low
# scores <- dcScore(emat, 
#                   condition,
#                   dc.method = "zscore",
#                   cor.method = 'spearman')
# raw_p <- dcTest(scores, emat, condition)
# adj_p <- dcAdjust(raw_p, f = p.adjust, method = 'fdr')

sigf_ICvslowCNIH4_mat <- ICvslowCNIH4_mat
sigf_ICvslowCNIH4_mat[raw_p>=0.05] <- NA

sigf_ICvshighCNIH4_mat <- ICvshighCNIH4_mat
sigf_ICvshighCNIH4_mat[raw_p>=0.05] <- NA

low_corr <- abs(sigf_ICvslowCNIH4_mat[upper.tri(sigf_ICvslowCNIH4_mat,diag=F)])
high_corr <- abs(sigf_ICvshighCNIH4_mat[upper.tri(sigf_ICvshighCNIH4_mat,diag=F)])

low_high <- data.frame(low_corr,high_corr)

names(low_high) <- c('Low CNIH4','High CNIH4')
df <- pivot_longer(low_high,colnames(low_high))
names(df) <- c('CNIH4level','Coexpression')

p<-ggplot(data=df,
       aes(x=fct_reorder(CNIH4level, Coexpression, .na_rm = TRUE,.desc =TRUE),
           y=Coexpression)) +
  geom_violin(aes(color=CNIH4level,
                  fill=CNIH4level),
              width=0.7,alpha=0.5,
              trim=FALSE)+
  scale_fill_manual(values=c("#B48EAD","#81A1C1"))+
  scale_color_manual(values=c("#B48EAD","#81A1C1"))+
  geom_boxplot(aes(color=CNIH4level),
               size=1,
               position=position_dodge(1),
               width=0.03,
               outlier.colour="red",
               outlier.shape=NA,outlier.size=3)+
    geom_jitter(aes(color=CNIH4level,fill=CNIH4level),
              width=0.05,
              height=0,
              size=2,
              alpha = 0.7)+
  theme_bw() +
  #geom_quasirandom(aes(color=CNIH4level),dodge.width=.8)+
  #scale_color_manual(values=c("#BF616A","#81A1C1"))+
  
  labs(x="",y = "Absolute pairwise coexpression of \n27 immune checkpoint genes (p <0.05)", title = "")+
  theme(plot.title = element_text(size = 25),
        legend.position = "none",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_text(color="black",size=13,face="bold"),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold"))+
  stat_compare_means( #aes(label = paste0("p = ", ..p.format..)),
                      size = 5,
                      label.x = 1.3,
                      label.y = 1.3)
p
# png(file=paste0(path_output,"/Figure4C_differential_coexpression_boxplot_TBM_with_PTB.png"),res=900, units = 'in', width=8, height = 7)
# p
# dev.off()
```

### 4. Boxplot Low vs high TBM

```{r}
tmp <- merge(impute_df,tbm_vst2,by=0) %>% select(c(B.cells.naive:Neutrophils,level))
tmp <- pivot_longer(tmp,B.cells.naive:Neutrophils)

p<-ggplot(data=tmp,
       aes(x=name,y=value))+
  geom_boxplot(aes(color=name)) +
  geom_jitter(height=0,aes(color=name),alpha=0.1)+
  xlab('Cell type')+
  ylab('% Cell population')+
  theme_bw()+
  theme(plot.title = element_text(size = 25),
        legend.position = "",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_text(color="black",size=13,face="bold",angle = 45,vjust = 1, hjust=1),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold"))
p
png(file=paste0(path_output,"/Figure_CellPop_All.png"),res=900, units = 'in', width=14, height = 9)
p
dev.off()

stat.test <- tmp %>%
  group_by(name) %>% 
  t_test(value ~ level) %>% 
  add_significance("p") %>%
  add_xy_position(x = "name")

p <- ggplot(data=tmp,
       aes(x=name,y=value, color=factor(level, level = c('LOW','HIGH'))))+
  geom_boxplot() +
  geom_point(position=position_jitterdodge(jitter.width =0.3),
             alpha=0.1)+
  scale_color_manual(values=c("#81A1C1","#B48EAD"))+
  labs(color='CNIH4 Level')+ 
  xlab('Cell type')+
  ylab('% Cell population')+
  theme_bw()+
  theme(plot.title = element_text(size = 25),
        legend.position = "right",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_text(color="black",size=13,face="bold",angle = 45,vjust = 1, hjust=1),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold"))+
  stat_pvalue_manual(stat.test, label = 'p.signif', tip.length = 0)

png(file=paste0(path_output,"/Figure_CellPop_LowvsHighCNIH4_TBM.png"),res=900, units = 'in', width=14, height = 9)
p
dev.off()
```

### 5. Boxplot CIBERSORT

```{r}
library(ggpubr)
require(plyr)
library(ggbeeswarm)

tmp0 <- read.csv(paste0(path_input,'Raw_data/LM22_ref_CNIH4/ICs/CIBERSORTxHiRes_NA_Neutrophils_Window40.txt'),sep='\t')
rownames(tmp0) <- tmp0$GeneSymbol
tmp0$GeneSymbol <- NULL
tmp0 <- t(tmp0) %>% as.data.frame() %>% dplyr::rename(CNIH4InNeutrophils = CNIH4)

lowCNIH4_neutrophils <- merge(lowCNIH4vsIC_df,tmp0,by=0)
highCNIH4_neutrophils <- merge(highCNIH4vsIC_df,tmp0,by=0)
CNIH4lv_neutrophils <- merge(tbm_vst2,tmp0,by=0)[,c('CNIH4InNeutrophils','level')]
CNIH4lv_neutrophils$level <- mapvalues(CNIH4lv_neutrophils$level, 
                                        from=c('LOW',"HIGH"), 
                                        to=c("Low CNIH4","High CNIH4"))

p<-ggplot(data=CNIH4lv_neutrophils,
       aes(x=reorder(level, CNIH4InNeutrophils,na.rm = TRUE),
           y=CNIH4InNeutrophils)) +
  theme_bw() +
  scale_fill_manual(values=c("#B48EAD","#81A1C1"))+
  scale_color_manual(values=c("#B48EAD","#81A1C1"))+
  geom_boxplot(aes(color=level),
               size=1,
               position=position_dodge(1),
               width=0.5,
               outlier.colour="red",
               outlier.shape=NA,outlier.size=3,
               varwidth=TRUE
               )+
  geom_beeswarm(aes(color=level),priority='density',size=1,alpha=0.5,cex = 1,
                corral.width = 0.1)+
  labs(x="",y = "CNIH4 expression in Neutrophils", title = "")+
  theme(plot.title = element_text(size = 25),
        legend.position = "none",
        legend.text = element_text(size = 20) ,
        legend.title  = element_text(size = 20,color="black",face="bold"),
        axis.text = element_text(size = 25),
        axis.text.x=element_text(color="black",size=13,face="bold"),
        axis.title.x = element_text(size = 20,color="black",face="bold"),
        axis.text.y=element_text(color="black",size=20),
        axis.title.y = element_text(size = 20,color="black",face="bold"),
        axis.title= element_text(size = 40,color="black",face="bold"))+
  stat_compare_means( #aes(label = paste0("p = ", ..p.format..)),
                      size = 5,
                      label.x = 1.3,
                      label.y = 250)#220

png(file=paste0(path_output,"/Figure_HighvsLowCNIH4InNeutrophils_CNIH4_TBM.png"),res=900, units = 'in', width=5, height = 7)
p
dev.off()
```
